---
title: "R Notebook"
output: html_notebook
---

Figure 6 scRNAseq from FACS sorted populations

1. processing filtering and creating data objects
2. Compare expression of 13 Antibody panel (use NK for O4) and create correlation matrix

```{r}
## library loading

library(lattice) # for correlation heat map
library(Seurat)
library(dplyr)
library('reshape')
library("ggplot2")

# data processing 

```


Correlation between FC and scRNAseq (protien an RNA) use NKX6.2 as proxy for O4

```{r}

output_path = "/Users/rhalenathomas/Documents/Data/FlowCytometry/PhenoID/Analysis/PaperFigures/PostSortGatingPops/"
seu.fc <- readRDS(paste(output_path,"june10postSort.seu.RDS", sep = ""))
unique(seu.fc$GatedPop)

# read in the transcriptomics
pathway <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PhenoID/scRNAseqSorted/objs/"
seu.sc <- readRDS(paste(pathway,"All4popssub07102022.RDS"))

marker.ab <- c("CD24","CD56","CD29","CD15","CD184","CD133","CD71","CD44","GLAST","AQP4","HepaCAM", "CD140a","O4")

marker.genes <-c("CD24","NCAM1","ITGB1","FUT4","CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4","HEPACAM", "PDGFRA","NKX6-2")



```


Make expression matrixes

```{r}

Idents(seu.sc) <- 'orig.ident'
DefaultAssay(seu.sc) <- 'RNA'

seu.sc <- ScaleData(seu.sc)

scRNAseq.mean <- as.data.frame(AverageExpression(seu.sc, assays = 'RNA',features = marker.genes,
                                   group.by = 'orig.ident', slot= 'scale.data'))


# change gene names to protein/AB names
rownames(scRNAseq.mean) <- marker.ab

# get the mean expression from FACS 
DefaultAssay(seu.fc) <- 'RNA'
Idents(seu.fc) <- 'GatedPop'

seu.fc <- ScaleData(seu.fc)

FC.mean <- as.data.frame(AverageExpression(seu.fc, assays = 'RNA',features = marker.ab,
                                   group.by = 'GatedPop', slot= 'scale.data'))




```


```{r}

# check some heatmaps
mat <- as.matrix(scRNAseq.mean)

heatmap(mat)

mat2 <- as.matrix(FC.mean)

heatmap(mat2)


```


Calculate correlation between two matrixes or Dataframes


```{r}

# rename and reorder dataframes to match
colnames(FC.mean)
colnames(FC.mean) <- c("Astrocytes","RadialGlia","Neurons1","Neurons2")
colnames(FC.mean)

colnames(scRNAseq.mean)
colnames(scRNAseq.mean) <- c("Astrocytes","Neurons1","Neurons2","RadialGlia")
colnames(scRNAseq.mean)
rna.mean <- scRNAseq.mean %>% select("Astrocytes","RadialGlia","Neurons1","Neurons2")

df.cor <- cor(FC.mean, rna.mean)
# first one is the rows and second is the columns, FC measures on the y axis

df.cor

```


Plot a heat map of the correlations

```{r}

# need to melt the matrix
longData<- melt(df.cor)

head(longData)
colnames(longData) <- c("FC","scRNA","R2")

ggplot(longData, aes(x=scRNA,y=FC, fill =R2)) + geom_tile() +
  scale_fill_gradient2(low = "#075AFF",
                       mid = "#FFFFCC",
                       high = "#FF0000") +
   guides(fill = guide_colourbar(label = TRUE,
                                ticks = FALSE)) + theme_bw() +
  coord_fixed()  +scale_x_discrete(expand=c(0,0))+
  scale_y_discrete(expand=c(0,0))+ theme(text = element_text(size=16, colour = "black"),
        axis.text.x = element_text(size = 18, colour = "black", angle = 90), axis.text.y = element_text(size= 18, colour = "black"))  + xlab('scRNAseq') + ylab('Flow Cytometry') + guides(fill=guide_legend(title="Correlation Coefficient"), element_text(size = 16, colour = "black"))


# save the plot of the correlation
out_plots <- "/Users/rhalenathomas/Documents/Projects_Papers/PhenoID/ForFigures/"
#pdf(paste(out_plots,"FC_scRNA_correlation.pdf"))
ggplot(longData, aes(x=scRNA,y=FC, fill =R2)) + geom_tile() +
  scale_fill_gradient2(low = "#075AFF",
                       mid = "#FFFFCC",
                       high = "#FF0000", midpoint = 0, limit = c(-1,1)) +
   guides(fill = guide_colourbar(label = TRUE,
                                ticks = FALSE)) + theme_bw() +
  coord_fixed()  +scale_x_discrete(expand=c(0,0))+
  scale_y_discrete(expand=c(0,0))+ theme(text = element_text(size=16, colour = "black"),
        axis.text.x = element_text(size = 18, colour = "black", angle = 90), axis.text.y = element_text(size= 18, colour = "black"))  + xlab('scRNAseq') + ylab('Flow Cytometry') + guides(fill=guide_legend(title="Correlation Coefficient"), element_text(size = 16, colour = "black"))  
#dev.off()

output_path <- "/Users/rhalenathomas/Documents/Projects_Papers/PhenoID/ForFigures/scRNA/"

# something is odd about the scale 
# I'll take stuff away to try and fix it
pdf(paste(output_path,"FC_scRNA_correlation.pdf"), width = 8, height = 5)
ggplot(longData, aes(x=scRNA,y=FC, fill =R2)) + geom_tile() +
  scale_fill_gradient2(low = "#075AFF",
                       mid = "#FFFFCC",
                       high = "#FF0000", midpoint = 0, limit = c(-0.7,0.7)) +
   guides(fill = guide_colourbar(label = TRUE,
                                ticks = FALSE)) + theme_bw() +
  coord_fixed()  +scale_x_discrete(expand=c(0,0)) +
  scale_y_discrete(expand=c(0,0)) + 
  theme(text = element_text(size=16, colour = "black"),
        axis.text.x = element_text(size = 16, colour = "black", angle = 90), axis.text.y = element_text(size= 16, colour = "black"))  + xlab('scRNAseq') + ylab('Flow Cytometry') 

dev.off()


+ guides(fill=guide_legend(title="Correlation Coefficient"), element_text(size = 16, colour = "black"))  
#dev.off()




```


Have a look at the expression of the given markers RNA by 


```{r}

#DoHeatmap(seu.sc, group.by = 'orig.ident', features = marker.genes)

RidgePlot(seu.sc, group.by = 'orig.ident', features = marker.genes, slot = 'scale.data', log = TRUE)

DotPlot(seu.sc, group.by = 'orig.ident', features = marker.genes)


```






