---
title: "R Notebook Create Reference matrix"
output: github_document
---

This workbook contains:
1. Instructions on how to make a reference matrix manually or using in house and public data.
  a) Manual input values
  b) Read values from a csv (made in excel)
  c) Create a data frame from Seurat objects.
  d) Combine different data sets to make one reference matrix.
  e) Save the reference matrix to read later.
2. Creation of the reference matrix used throughout the manuscript.
  a) Data inputs used and their sources.
  b) Combination and normalization of the inputs.
  c) Final reference matrix used in this manuscript for the first antibody panel.
3. Creation of a second reference matrix with an adjusted antibody panel used for the time course analysis. 
  a) Data inputs used and their sources.
  b) Combination and normalization of the inputs.
  c) Final reference matrix used for the time course analysis.


To create the reference matrix you must first obtain expression data from different sources.

Not all users will be able to easily process these public data sources. However it is feasible to use excel and save one tab as a csv file then read this into R (examples below). The csv file (from excel) should formatted with column names:  
Marker, celltype1, celltype2, celltype3 ... etc
In the Marker column is the names of the antibody.  
In the other columns (cell types) are expression values.  
Values can be entered into directly into the spreadsheet by estimations from the literature or copied from browsers.  These would be qualitative estimations but will still work as a reference matrix in the correlation matrix. 

A third option is to enter values into R.  
An example is shown here.

To make your own reference matrix by estimating from publications you can score expression from 0-5 (or whatever scale you like).  Enter your own antibodies under markers.  Enter expression values for each cell type as shown here. 
```{r}
# example of directly entering values into R for a reference matrix.  Relative values estimated from literature search. 
# these values are not used in the manuscript and are placed here as an example and are not actually specific values from a literature search. 

# first we create a vector for each set of values
Marker <- c("CD24","CD56","CD29","CD15","CD184","CD133","CD71","CD44","GLAST","AQP4","HepaCAM", "CD140a","O4")

# create a vector for each cell type with an expression value for each antibody.  
# you must follow the order of the antibodies in your marker list.
StemCell <- c(5,2,5,5,5,5,1,1,1,1,1,3,0)
NPC <- c(3,3,5,5,3,3,2,1,1,1,1,1,0)
Neuron <- c(5,5,3,3,1,1,4,1,1,1,1,1,0)
Astrocyte <- c(1,1,2,1,1,1,3,4,4,5,5,2,0)
OPC <- c(2,2,2,2,2,2,2,3,3,1,1,5,2)
Oligo <- c(1,1,1,1,2,2,1,2,2,1,1,3,5)
RG <- c(3,3,4,4,5,6,4,4,4,3,1,3,0)
Endothelial <- c(2,2,2,4,4,4,3,4,1,1,1,1,0)
Epithelial <- c(1,1,2,5,1,3,2,3,3,1,2,2,0)

ref_dataframe <- data.frame(Marker, Astrocyte, Endothelial, Epithelial, Neuron, 
                            NPC, Oligo, OPC, RG, StemCell)
dim(ref_dataframe)
head(ref_dataframe)


```
Now we z-score the data frame

```{r}
df.value <- ref_dataframe[1:13, 2:10]  # Select the columns you want to normalize
zero.one <- t(apply(df.value, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
colnames(zero.one) <- colnames(df.value)
rownames(zero.one) <- ref_dataframe$Marker
head(zero.one)

# replace with normalized dataframe
ref_dataframe <- zero.one

# we can now save the reference matrix 
output_path <- "where/to/save/file/"   # example not a real file path

output_path <- "/GITHUB/CelltypeR/ExampleOuts/"
write.csv(ref_dataframe, paste(output_path,"ManualReferenceMatrix.csv",sep = ""))


```

Load libraries for data wrangling and plotting

```{r}
# load libraries
library(tidyverse)
library(ggplot2)
library(Seurat)
#library(pryr) # for memory check

```

Now we can plot the reference matrix

```{r}

# we can use the normalized matrix without the column of antibodies names (these are the row names) to plot a heatmap

heatmap(ref_dataframe, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

# remember these are rough/made up values not used in the manuscript


```


# Making the reference matrix with different input data 
For antibody panel: "CD24","CD56","CD29","CD15","CD184","CD133","CD71","CD44","GLAST","AQP4","HepaCAM", "CD140a","O4"

1. 2D Flow (experimentally acquired this manuscript)
2. BenBarres bulk RNA from human isolated cells
3. scRNAseq public data from human brain
4. scRNAseq from previously acquired (and previously published) scRNAseq from midbrain organoids (165 days in final differentiation)
5. scRNAseq from cerebral organoids (public data)


Flow cytometry data from 2D cell cultures

```{r}
# read in the flow cytometry data 
# this is a csv that was created with the preprocessing workflow

input_path <- "where/you/have/the/data/file/"

# this is the complete data 
# this data is available in the github

fc <- read.csv(paste(input_path,"2Dcells_surface_flow.csv", sep = ""))

# we now want to get the mean expression values for each antibody in each cell type
# check what columns we have
colnames(fc)

# here "batch" is the cell type
fcf <- fc[4:17]
head(fcf)

colnames(fcf)
dim(fcf)
unique(fcf$Batch)

# now we want to make a new dataframe with mean expression values
# Calculate the mean value per column and per cell type
meanfc <- aggregate(. ~ fcf$Batch, data = fcf[, -14], FUN = mean)

# our table is transposed, we need to switch it

# set the Batch values (cell types) as rownames
rownames(meanfc) <- meanfc$`fcf$Batch`
fcmeans <- data.frame(t(meanfc[, -1]), stringsAsFactors = FALSE)
head(fcmeans)

# now we see that there are two astrocyte samples, two NPC samples, two oligodendrocytes samples.

# We will use oligo.e (early) as OPC and oligo.m (mature) as oligodendrocytes (these are still premyelinating)

# For NPC and Astrocytes we will take an average



```

```{r}
# Take the mean of the repeat samples
df_flow <- fcmeans

# Calculate the average for NPCs and create the new column
df_flow$NPCs <- rowMeans(df_flow[, c("NPC1", "NPC2")])

# Remove the individual NPC columns
df_flow <- df_flow %>% select(-c("NPC1", "NPC2"))

df_flow$Astrocytes <- rowMeans(df_flow[, c("Astrocytes2", "Astrocyes1")])
df_flow <- df_flow %>% select(-c("Astrocytes2", "Astrocyes1"))



#rename columns
colnames(df_flow) <- c("IPSC","Neuron","OPC","Oligo","NPC","Astrocyte")

head(df_flow)

```
Now z-score the flow data

```{r}
df.value <- df_flow 
zero.one <- t(apply(df.value, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
fcz <- zero.one


# we can use the normalized matrix without the column of antibodies names (these are the row names) to plot a heatmap
heatmap(fcz, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

# later we will see that the cell types not included in the flow cytometry data will have NA (empty spaces) for the missing cell types. 


# save the flow cytometery in 2D cells normalized matrix
output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(fcz, paste(output_path,"FC_ReferenceMatrix.csv",sep = ""))


```

Use RNA values from sorted human brain tissue. 
These values are from an excel spreadsheet downloaded as the supplemental data from https://www.brainrnaseq.org/
The data is in the supplemental data from doi: 10.1523/JNEUROSCI.1860-14.2014
I adjusted the column headings in excel so there is only one column name row at the top. 

```{r}

input_path <- "/Users/rhalenathomas/Documents/Data/RNA/BenBarres/"
RNA_adult_brain <- read.csv(paste(input_path,"HumanBrainRNASortedCells.csv",sep = ""))

colnames(RNA_adult_brain)
# there are multiple samples for each cell type.
# we will take averages for each cell type
# we will not included brain tumors
# first we need to filter for the genes that match our proteins targeted by the antibody panel.

```
Now we need to reformat the data frame so we can normalize that data
```{r}

# first we need to select the genes that match our proteins 
# O4 is a glycoprotein with unknown gene transcript. NKX6-2 and NKX2-2 are involved in oligodendrocyte differentiation and correlate with O4 staining

# antibody/proteins "CD24","CD56","CD29","CD15","CD184","CD133","CD71","CD44","GLAST","AQP4","HepaCAM", "CD140a","O4"
marker_genes = marker_genes = c("CD24","NCAM1","ITGB1","FUT4",
                            "CXCR4","PROM1","TFRC","CD44",
                            "SLC1A3","AQP4","HEPACAM", "PDGFRA",
                            "NKX2-2")
  

df <- subset(RNA_adult_brain, X %in% marker_genes)

# now we want to add the gene names as the row names
rownames(df) <- df$X
colnames(df)

# remove the X column of gene names, the Tumor cells and the Whole cortex
df <- df[,6:38]
colnames(df)

# now take the average of the different cell types

df$Astrocytes.Fetal <- rowMeans(df[, c("AstrocytesFetal",
                                           "AstrocytesFetal.1","AstrocytesFetal.2",
                                           "AstrocytesFetal.3","AstrocytesFetal.4",
                                           "AstrocytesFetal.5","AstrocytesFetal.6",
                                           "AstrocytesFetal.7","AstrocytesFetal.8",
                                           "AstrocytesFetal.9")])
head(df)
# Remove the individual columns
df <- df %>% select(-c("AstrocytesFetal",
                                           "AstrocytesFetal.1","AstrocytesFetal.2",
                                           "AstrocytesFetal.3","AstrocytesFetal.4",
                                           "AstrocytesFetal.5","AstrocytesFetal.6",
                                           "AstrocytesFetal.7","AstrocytesFetal.8",
                                           "AstrocytesFetal.9"))
head(df)

df$Astrocytes.Mature <- rowMeans(df[, c("AstrocytesMature","AstrocytesMature.1",
                                        "AstrocytesMature.2","AstrocytesMature.3",
                                        "AstrocytesMature.4","AstrocytesMature.5",
                                        "AstrocytesMature.6","AstrocytesMature.7",
                                        "AstrocytesMature.8","AstrocytesMature.9",
                                        "AstrocytesMature.10",
                                        "AstrocytesMature.11")])

# remove individual columns
df <- df %>% select(-c("AstrocytesMature","AstrocytesMature.1",
                                        "AstrocytesMature.2","AstrocytesMature.3",
                                        "AstrocytesMature.4","AstrocytesMature.5",
                                        "AstrocytesMature.6","AstrocytesMature.7",
                                        "AstrocytesMature.8","AstrocytesMature.9",
                                        "AstrocytesMature.10",
                                        "AstrocytesMature.11"))
colnames(df)

#means of oligo
df$Oligodendrocyte <- rowMeans(df[, c("Oligo","Oligo.1","Oligo.2",
                                      "Oligo.3","Oligo.4")])

df <- df %>% select(-c("Oligo","Oligo.1","Oligo.2", 
                       "Oligo.3","Oligo.4"))

colnames(df)

# means of macrophage/microglia - these will be removed later as we do not expect these cells in midbrain organoids
df$marcrophage.microglia <- rowMeans(df[, c("MacrophageMicroglia",
                                            "MacrophageMicroglia.1",
                                            "MacrophageMicroglia.2")])

df <- df %>% select(-c("MacrophageMicroglia","MacrophageMicroglia.1",
                                            "MacrophageMicroglia.2"))

colnames(df)

# mean from endothelial cells
df$marcrophage.microglia <- rowMeans(df[, c("Endothelial","Endothelial.1")])

df <- df %>% select(-c("Endothelial","Endothelial.1"))

colnames(df)
class(df)
head(df)

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
fc_rnaseq <- zero.one

# quick visualization
heatmap(fc_rnaseq, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

# the normalization is across each row (gene) and is effected by the cell types included.
# we can see that mature astrocytes match expected expression (Glast/SLC1A3, HEPACAM, AQP4 and CD79/TFRC and low in CD44) This matches the flow data except for Glast.  The fetal astroctyes match expection of glia stem cells/progenitors, higher in CD44 and precursor markers CD133/PROM1, CD184/CXCR4, unexpectedly the fetal astrocytes are high in CD24, reported high in neurons and stem cells. 

```
```{r}
# z-score the total RNAseq brain data without microglia

df <- df %>% select(-c("marcrophage.microglia"))
colnames(df)

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
fc_rnaseq <- zero.one

# quick visualization
heatmap(fc_rnaseq, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")


```
Adult brain data from single cell sequencing
All data sets were processed into Seurat objects.

1. Adult midbrain: https://doi.org/10.1038/s41593-022-01061-1.  (Kamath/Macosko) 
2. Developing brain: https://cells.ucsc.edu/?ds=cortex-dev DOI: 10.1126/science.aap8809 (Nowakowski/Kriegstein)
3. Developing forebrain
4. Fetal Midbrain La Manno


```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"

adult_midbrain <- readRDS(paste(input_path,"KamathTotal_downsample.RDS",sep = ""))

colnames(adult_midbrain@meta.data)
unique(adult_midbrain$Cell_Subtype)
unique(adult_midbrain$Cell_Type)

# we will use the cell type (main cell types)

```

```{r}
# get the mean expression values for each gene matching the protein targets in our antibody panel
# use seurat function mydatafetch

# get expression for each cell type
Idents(adult_midbrain) <- "Cell_Type"
unique(adult_midbrain$Cell_Type)
# we need to remove the NaN cells
seu <- subset(adult_midbrain, idents = c("astro", "endo",  "mg",
                                                   "olig", "nonda", "da", "opc"))
unique(seu$Cell_Type)
# scale the data
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))

expression <- FetchData(seu, vars = c("ident",marker_genes))


# get means for each cell type
ex <- as.tbl(expression)
meanEx <- ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean"))
head(meanEx)
# remove the gene name column and add these as row names

df <- as.data.frame(meanEx)
rownames(df) <- df$ident
df <- df[,2:14]
df <- as.data.frame(t(df))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")

# now we will normalize the data z-score the same as the other data
# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scAmidbrain <- zero.one

# quick visualization
heatmap(scAmidbrain, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

# as we won't separate DA neurons from other neurons I will take the mean here
# I will remove microglia


```
Clean up the dataframe

```{r}
# as we won't separate DA neurons from other neurons I will take the mean here
# I will remove microglia

df$Neuron <- rowMeans(df[, c("da","nonda")])
df <- df %>% select(-c("mg","da","nonda"))
# change to same cell type names
colnames(df) <- c("Astrocyte","Endothelial","Oligodendrocyte","OPC","Neuron")

# now we will normalize the data z-score the same as the other data
# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scAmidbrain <- zero.one

# quick visualization
heatmap(scAmidbrain, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

# there is no expression of CD24 in this data

```

Developing cortex
```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"

dev_cortex <- readRDS(paste(input_path,"Nowakowski_dev_cortext.RDS",sep = ""))

colnames(dev_cortex@meta.data)
unique(dev_cortex$Celltypes)
# I'll take only the cell types of interest
# IPC = intermediate progenitors
Idents(dev_cortex) <- "Celltypes"
seu <- subset(dev_cortex, idents = c("RG","EN","IN","IPC","OPC",
                                        "Astrocyte","Endothelial"))
# normalize and scale
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))


marker_genes = marker_genes = c("CD24","NCAM1","ITGB1","FUT4",
                            "CXCR4","PROM1","TFRC","CD44",
                            "SLC1A3","AQP4","HEPACAM", "PDGFRA",
                            "NKX2-2","NKX2.2")
# the genes are named differently in this data set

# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))

# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names

rownames(meanEx) <- meanEx$ident
head(meanEx)
meanEx <- meanEx[,2:14]
df <- as.data.frame(t(meanEx))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")

head(df)

# EN and IN are excitation and inhibitory neurons
# I'll take the average

df$Neuron <- rowMeans(df[, c("IN","EN")])
df <- df %>% select(-c("IN","EN"))
# change to same cell type names
head(df)
colnames(df) 
# we might consider IPC as IPSC/StemCell or as NPC

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scDevCor <- zero.one

# quick visualization
heatmap(scDevCor, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")




```

Midbrain and striatum two ages.
Bhaduri https://doi.org/10.1038/s41586-021-03910-8
https://cells.ucsc.edu/?ds=dev-brain-regions

```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"

midbrain_str <- readRDS(paste(input_path,"Bhaduri_midbrain_striatum.RDS",sep = ""))
colnames(midbrain_str@meta.data)
unique(midbrain_str$cell_type)
unique(midbrain_str$age)
unique(midbrain_str$cell_class) # cell class same as cell type
unique(midbrain_str$cell_cluster)
# I'll take only the cell types of interest
# IPC = intermediate progenitors
Idents(midbrain_str) <- "cell_type"
seu <- subset(midbrain_str, idents = c("dividing" ,"neuron","radial glia",
                                     "endothelial","oligodendrocyte",
                                        "astrocyte","interneuron"))
      
# normalize and scale
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))


marker_genes = marker_genes = c("CD24","NCAM1","ITGB1","FUT4",
                            "CXCR4","PROM1","TFRC","CD44",
                            "SLC1A3","AQP4","HEPACAM", "PDGFRA",
                            "NKX2-2","NKX2.2")
# the genes are named differently in this data set

# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))

# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names

rownames(meanEx) <- meanEx$ident
head(meanEx)
meanEx <- meanEx[,2:14]
df <- as.data.frame(t(meanEx))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")

head(df)
colnames(df)

# Neuron and Interneuron are excitation and inhibitory neurons
# I'll take the average

df$Neuron <- rowMeans(df[, c("neuron","interneuron")])
df <- df %>% select(-c("neuron","interneuron"))
# change to same cell type names
head(df)
colnames(df) 
# we might consider IPC as IPSC/StemCell or as NPC

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scMidStr <- zero.one

# quick visualization
heatmap(scMidStr, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")



```

Developing Forebrain
https://cells.ucsc.edu/?ds=human-forebraindev+sc-rna-seq
Van Bruggen et al, Developmental cell 2022

```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"

forebrain <- readRDS(paste(input_path,"Karolinski_DevForebrain_downsample_Level1.RDS",sep = ""))
colnames(forebrain@meta.data)
unique(forebrain$Age)
unique(forebrain$Level1)
unique(forebrain$Celltypes)
# I'll take only the cell types of interest
# IPC = intermediate progenitors
Idents(forebrain) <- "Celltypes"

seu <- subset(forebrain, idents = c("RG" ,"IN","EN",
                                     "Endothelial","NPC","Neurons","OPC"))
      
# normalize and scale
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))


marker_genes = marker_genes = c("CD24","NCAM1","ITGB1","FUT4",
                            "CXCR4","PROM1","TFRC","CD44",
                            "SLC1A3","AQP4","HEPACAM", "PDGFRA",
                            "NKX2-2","NKX2.2")
# the genes are named differently in this data set

# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))

# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names

rownames(meanEx) <- meanEx$ident
head(meanEx)
meanEx <- meanEx[,2:14]
df <- as.data.frame(t(meanEx))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")

head(df)
colnames(df)

# Neuron and Interneuron are excitation and inhibitory neurons
# I'll take the average

df$Neuron <- rowMeans(df[, c("Neurons","IN","EN")])
df <- df %>% select(-c("Neurons","IN","EN"))
# change to same cell type names
head(df)
colnames(df) 
# we might consider IPC as IPSC/StemCell or as NPC

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scForebrain <- zero.one

# quick visualization
heatmap(scForebrain, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")
```

Fetal midbrain https://doi.org/10.1016/j.cell.2016.09.027
La Manno

```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"
# this is a singlecellexperiment object
fetalmid <- readRDS(paste(input_path,"manno_human.rds",sep = ""))
# convert to seurat
seu <- as.Seurat(fetalmid, counts = "counts", data = "logcounts")
colnames(seu@meta.data)
unique(seu$cell_type1)
unique(seu$Source)
Idents(seu) <- "cell_type1"
levels(seu)
# cell type main groups
celltypes <- c("NB","NB","NB","NB","Prog","Prog",
               "Prog","Prog","RG","RG","RG","RG",
               "RG","RG","ESC","ESC","ESC","DAN",
               "DAN","DAN","Endothelial","IN","MGL","IN",
               "NB","NB","NB","Prog","OMTN","OPC",
               "Pericyte","Prog","Prog","Prog","Prog","RG",
               "RG","RG","RG","RG","RN","Neurons",
               "DAN","DAN","DAN","MN","MN","NB",
               "NB","Prog","Prog","Prog","RG","RG",
               "RG","Unk")

names(celltypes) <- levels(seu)
seu <- RenameIdents(seu, celltypes)
seu <- AddMetaData(seu, metadata = Idents(seu), col.name = "Celltypes")
# save the Main cell types for later
saveRDS(seu,"/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/manno_human.rds" )

table(seu$Celltypes,seu$cell_type1)
unique(seu$Celltypes)
# select only the cell types I want to use
Idents(seu) <- "Celltypes"
seu <- subset(seu, idents = c("RG","Prog","DAN","Endothelial","Neurons","OPC","IN"))

seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))


marker_genes = marker_genes = c("CD24","NCAM1","ITGB1","FUT4",
                            "CXCR4","PROM1","TFRC","CD44",
                            "SLC1A3","AQP4","HEPACAM", "PDGFRA",
                            "NKX2-2","NKX2.2")
# the genes are named differently in this data set

# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))

# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names

rownames(meanEx) <- meanEx$ident
head(meanEx)
meanEx <- meanEx[,2:14]
df <- as.data.frame(t(meanEx))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")

head(df)
colnames(df)

# Neurons, DAN and IN are all neurons 
# I'll take the average

df$Neuron <- rowMeans(df[, c("Neurons","IN","DAN")])
df <- df %>% select(-c("Neurons","IN","DAN"))
# change to same cell type names
head(df)
colnames(df) 
# we might consider IPC as IPSC/StemCell or as NPC

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scManno <- zero.one

# quick visualization
heatmap(scManno, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

```



Now we will merge the brain scRNAseq data sets and the total RNA data set

```{r}
# we need to make sure the marker names and the cell type names match

sc1 <- scAmidbrain
sc2 <- scDevCor
sc3 <- scForebrain
sc4 <- scMidStr
sc5 <- scManno
total <- fc_rnaseq

rownames(sc1)
rownames(sc2)
rownames(sc3)
rownames(sc4)
rownames(sc5)
rownames(total)

# the total RNA has a different order to the rows
# reorder the rows

# Get the row order of sc1 in total
row_order <- match(rownames(sc1), rownames(total))

# Reorder rows of total based on sc1's order
t1 <- total[row_order, ]

rownames(t1)
rownames(total)
rownames(sc1)

```
Merge and take the average of the two data sets and then take the mean value when there is overlap

```{r}
# I want to merge by gene/protein 

# Join dataframes using merge
df_brain <- merge(sc1, sc2, by = "row.names")
# add the row names to the merged data frame
rownames(df_brain) <- df_brain$Row.names
# remove the newly created Row.names column
df_brain <- df_brain %>% select(-"Row.names")

df_brain <- merge(df_brain, sc3, by = "row.names")
rownames(df_brain) <- df_brain$Row.names
df_brain <- df_brain %>% select(-"Row.names")

df_brain <- merge(df_brain, sc4, by = "row.names")
rownames(df_brain) <- df_brain$Row.names
df_brain <- df_brain[,2:23]

df_brain <- merge(df_brain, sc5, by = "row.names")
rownames(df_brain) <- df_brain$Row.names
dim(df_brain)
df_brain <- df_brain[,2:28]

df_brain <- merge(df_brain, t1, by = "row.names")
rownames(df_brain) <- df_brain$Row.names
dim(df_brain)
df_brain <- df_brain[,2:32]

head(df_brain)
dim(df_brain)
colnames(df_brain)

# take the means
# copy the variable in case of an error to not have to run the above code again
scBrain <- df_brain

scBrain$neuron <- rowMeans(scBrain[, c("Neuron.x", "Neuron.y","Neuron.x.1",
                                       "Neuron.y.1", "Neurons","Neuron")], na.rm = TRUE)
scBrain$Astrocyte <- rowMeans(scBrain[, c("Astrocyte.x", "Astrocyte.y",
                                          "astrocyte","Astrocytes.Mature")], na.rm = TRUE)
scBrain$OPC <- rowMeans(scBrain[, c("OPC.x", "OPC.y","OPC.y.1","OPC.x.1")], na.rm = TRUE)
scBrain$Endothelial <- rowMeans(scBrain[, c("Endothelial.x", "Endothelial.y",
                                            "Endothelial.x.1","Endothelial.y.1",
                                            "endothelial")], na.rm = TRUE)
scBrain$Oligo <- rowMeans(scBrain[, c("Oligodendrocyte.x","oligodendrocyte",
                                     "Oligodendrocyte.y"), na.rm = TRUE])

scBrain$RadialGlia <- rowMeans(scBrain[, c("RG.x","RG.y","radial glia","RG")], na.rm = TRUE)
scBrain$Stem <- rowMeans(scBrain[, c("IPC","dividing","Prog")], na.rm = TRUE)

scBrain <- scBrain %>% select(c("Astrocyte","Endothelial","OPC",
                                "Oligo","neuron","RadialGlia","Stem"))
head(scBrain)

# z-score
zero.one <- t(apply(scBrain, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
RNAbrain <- zero.one

# quick visualization
heatmap(RNAbrain, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

# save the average RNA brain expression matrix
output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(RNAbrain, paste(output_path,"BrainRNAReferenceMatrix.csv",sep = ""))


```

scRNAseq from organoids
1. Midbrain organoids 165 days AST23 and isogenic control https://doi.org/10.1093/braincomms/fcab223

2.Cerebral organoids


AST23 and isogenic control 165 days in final differentiation media

```{r}

# read in the data from the 165 days midbrain organoids
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/AST23_BrainComm/"
MO <- readRDS(paste(input_path,"SNCAandiso_labels140522.rds",sep = ""))
colnames(MO@meta.data)
unique(MO$celltypes)
unique(MO$cluster.ids)
# I'll take only the cell types of interest

Idents(MO) <- "celltypes"

seu <- subset(MO, idents = c("RG" ,"Astrocytes",
                                     "Epithelial","NPC","Neurons","DA-Neurons"))
      
# normalize and scale
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))


marker_genes = marker_genes = c("CD24","NCAM1","ITGB1","FUT4",
                            "CXCR4","PROM1","TFRC","CD44",
                            "SLC1A3","AQP4","HEPACAM", "PDGFRA",
                            "NKX2-2")

# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))

# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names

rownames(meanEx) <- meanEx$ident
head(meanEx)
meanEx <- meanEx[,2:14]
df <- as.data.frame(t(meanEx))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")

head(df)
colnames(df)

# Neuron and DA neurons
# I'll take the average

df$Neuron <- rowMeans(df[, c("Neurons","DA-Neurons")])
df <- df %>% select(-c("Neurons","DA-Neurons"))
# change to same cell type names
head(df)
colnames(df) 
# we might consider IPC as IPSC/StemCell or as NPC

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scMO <- zero.one

# quick visualization
heatmap(scMO, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(scMO, paste(output_path,"scMOReferenceMatrix.csv",sep = ""))


```

Cerebral organoids

Tanaka et al. 2020. Cell Rep. https://cells.ucsc.edu/?ds=organoidatlas&meta=Cluster


```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"
so <- readRDS(paste(input_path,"UCSCorg30000cellsSeurat.RDS"))

colnames(so@meta.data)
unique(so$Type)
unique(so$Subtype)
Idents(so) <- "Type"
# select the cell types 
#[1] "ExcitatoryNeuron" "RadialGlia"       "Astrocyte"       
#[4] "IPC"              "InhibitoryNeuron" "Outlier"         
#[7] "Unknown"
seu <- subset(so, idents = c("ExcitatoryNeuron","Astrocyte","RadialGlia","IPC",
                             "InhibitoryNeuron"))
      
# normalize and scale
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))


marker_genes = marker_genes = c("CD24","NCAM1","ITGB1","FUT4",
                            "CXCR4","PROM1","TFRC","CD44",
                            "SLC1A3","AQP4","HEPACAM", "PDGFRA",
                            "NKX2-2")

# check for genes
DoHeatmap(seu, features = marker_genes)

# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))

# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names

rownames(meanEx) <- meanEx$ident
head(meanEx)
dim(meanEx)
meanEx <- meanEx[,2:8]
df <- as.data.frame(t(meanEx))
# some genes are missing
rownames(df)

rownames(df) <- c("CD24","NCAM1","ITGB1","CXCR4","PROM1","TRFC",
                  "SLC1A3")
# present: "CD24_mean"   "NCAM1_mean"  "ITGB1_mean"  "CXCR4_mean" 
# "PROM1_mean"  "TFRC_mean"   "SLC1A3_mean"
#absent NKX2-2, PDGFRA, HEPACAM, AQP4, CD44, FUT4
row.names <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")

# we need to add in rows of NaN values for the missing genes
# Create a new dataframe with NaN rows for missing genes
# Create a new dataframe with NaN rows and desired row names and column names
new_df <- data.frame(matrix(NA, nrow = 6, ncol = ncol(df)))
rownames(new_df) <- c("NKX2-2", "PDGFRA", "HEPACAM",
                      "AQP4", "CD44", "FUT4")
colnames(new_df) <- colnames(df)

# Merge the new dataframe with the existing dataframe
# Combine the existing dataframe and new dataframe
combined_df <- rbind(df, new_df)

# reorder
# Reorder the rows based on the desired row names
reordered_df <- combined_df[row.names, ]
head(reordered_df)
# check the cell type names
colnames(reordered_df)

# [1] combine "ExcitatoryNeuron"  "InhibitoryNeuron"
# I'll take the average
df <- reordered_df

df$Neuron <- rowMeans(df[, c("ExcitatoryNeuron","InhibitoryNeuron")])
df <- df %>% select(-c("ExcitatoryNeuron","InhibitoryNeuron"))
# change to same cell type names
head(df)
colnames(df) 
# we might consider IPC as IPSC/StemCell or as NPC

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scCO <- zero.one

# quick visualization
heatmap(scCO, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")


```

Take the mean of the two organoid datasets

```{r}

scMO <- read.csv("/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/scMOReferenceMatrix.csv")
rownames(scMO) <- scMO$X
scMO <- scMO %>% select(-"X")
sc1 <- scMO
rownames(sc1)
sc2 <- scCO
rownames(sc2)
rownames(sc2) <- c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")
# Join dataframes using merge
df_Org <- merge(sc1, sc2, by = "row.names")
# add the row names to the merged data frame
rownames(df_Org) <- df_Org$Row.names
head(df_Org)
colnames(df_Org)
# remove the newly created Row.names column
df_Org <- df_Org %>% select(-"Row.names")

# take the means
# copy the variable in case of an error to not have to run the above code again
dfOrganoid <- df_Org
#df_Org <- dfOrganoid
colnames(df_Org)

df_Org$Neuron <- rowMeans(df_Org[, c("Neuron.x", "Neuron.y")], na.rm = TRUE)
df_Org$Astro <- rowMeans(df_Org[, c("Astrocytes", "Astrocyte")], na.rm = TRUE)

df_Org$Radial_Glia <- rowMeans(df_Org[, c("RG", "RadialGlia")], na.rm = TRUE)

df_Org <- df_Org %>% select(c("Astro","Neuron","Radial_Glia","NPC",
                              "IPC","Epithelial"))
zero.one <- t(apply(df_Org, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
organoid <- zero.one

# quick visualization
heatmap(organoid, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

# save the organoid dataframe
write.csv(organoid,"/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/OrganoidReferenceMatrix.csv")


```


Finally I will combine the different inputs.
1. FC protein
2. Brain RNAseq
3. Organoid RNAseq

I will first take the mean of the transcriptomics then take the mean of the result with the FC levels to weight the protein higher in the correlation matrix.

```{r}
# I will read in the saved csv files and will have 3 dataframes with a column for the rownames

input_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"

fc <- read.csv(paste(input_path,"FC_ReferenceMatrix.csv",sep = ""))
brain <- read.csv(paste(input_path,"BrainRNAReferenceMatrix.csv",sep = ""))
organoid <- read.csv(paste(input_path,"OrganoidReferenceMatrix.csv",sep = ""))


```

Merge the transcriptomics and take means of the same cell types

```{r}

rna_merge <- merge(brain,organoid, by = "X")
colnames(rna_merge)

# take means
rna_merge$Neurons <- rowMeans(rna_merge[, c("Neuron", "neuron")], na.rm = TRUE)
rna_merge$Astrocytes <- rowMeans(rna_merge[, c("Astrocyte", "Astro")], na.rm = TRUE)
rna_merge$StemCell <- rowMeans(rna_merge[, c("IPC", "Stem")], na.rm = TRUE)
rna_merge$RG <- rowMeans(rna_merge[, c("RadialGlia", "Radial_Glia")], na.rm = TRUE)
colnames(rna_merge)

df <- rna_merge %>% select(c("Endothelial","Epithelial", "Neurons",
                             "Astrocytes","StemCell","RG","OPC","NPC",
                             "Oligo"))

rownames(df) <- rna_merge$X

zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
trans_df <- zero.one

# quick visualization
heatmap(trans_df, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")



```

Now take the mean of the transcription and the FC protein levels

```{r}

# We need to change the gene names into the protein names
protein.names <- fc$X

rownames(fc) <- fc$X

gene.names <- rownames(trans_df)
# see the two lists
gene.names
protein.names
# rename the rownames of the transcription df with protein names
# the order needs to be the same as the gene names, which is not the same as the proteins because they are sorted alphabetically
gene.protein <- c("AQP4","CD24","CD44","CD184","CD15",   
                  "HepaCAM","CD29", "CD56",  "O4", "CD140a",
                  "CD133","GLAST", "CD71")

rownames(trans_df) <- gene.protein

# merge the dataframes by rownames
df_merge <- merge(fc,trans_df, by = 0)
# add in the rownames
rownames(df_merge) <- df_merge$Row.names

df_merge <- df_merge %>% select(-c("X","Row.names"))
# see the column names to know which to take the average
colnames(df_merge)

df <- df_merge
# take means
df$neurons <- rowMeans(df[, c("Neuron", "Neurons")], na.rm = TRUE)
df$astrocytes <- rowMeans(df[, c("Astrocyte", "Astrocytes")], na.rm = TRUE)
df$stemlike <- rowMeans(df[, c("IPSC", "StemCell")], na.rm = TRUE)
df$NPC <- rowMeans(df[, c("NPC.x", "NPC.y")], na.rm = TRUE)
df$OPC <- rowMeans(df[, c("OPC.x", "OPC.y")], na.rm = TRUE)
df$NPC <- rowMeans(df[, c("NPC.x", "NPC.y")], na.rm = TRUE)
df$oligodendrocyte <- rowMeans(df[, c("Oligo.x", "Oligo.y")], na.rm = TRUE)

# check the new column names
colnames(df)

df_final <- df %>% select(c("astrocytes","Endothelial","Epithelial",
                            "neurons","NPC","OPC","oligodendrocyte",
                            "RG","stemlike"))


zero.one <- t(apply(df_final, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
ref_mat <- zero.one

# quick visualization
heatmap(ref_mat, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Refernce Matrix")



```

Save the reference matrix to be used in the correlation function

```{r}

write.csv(ref_mat,"/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/FinalReferenceMatrix.csv" )

```


Make a second reference matrix for the adjusted time course panel
1. FC data for overlapping proteins
2. Brain transcriptomics
3. organoid transcriptomics

New Panel:
TH,CD24,CD56,CD29,CD15,CD184,CD133,SSEA4,CD44,CD49f,CD140a

Antibodies overlapping with the original panel
CD133,CD140a,CD15,CD184,CD24,CD29,CD44,CD56

New additions
SSEA4,TH,CD49f

Flow cytometry data

```{r}

# read in the flow cytometry data 
# this is the same flow data as before an will only contain the overlapping antibodies

input_path <- "/Users/rhalenathomas/Documents/Data/FlowCytometry/PhenoID/Analysis/2Dcells_surface/preprocessing/Allsamples/"

# this is the complete data
fc <- read.csv(paste(input_path,"2Dcells_surfacealigned_transformed_flowset.csv", sep = ""))

# we now want to get the mean expression values for each antibody in each cell type
# check what columns we have
colnames(fc)


# here "batch" is the cell type
# we will select just the antibodies in our new panel
fcf <- fc %>% select(c("CD133","CD140a","CD15","CD184","CD24",
                       "CD29","CD44","CD56","Batch"))
head(fcf)

colnames(fcf)
dim(fcf)
unique(fcf$Batch)

# now we want to make a new dataframe with mean expression values
# Calculate the mean value per column and per cell type
meanfc <- aggregate(. ~ fcf$Batch, data = fcf[, -9], FUN = mean)

# our table is transposed, we need to switch it

# set the Batch values (cell types) as rownames
rownames(meanfc) <- meanfc$`fcf$Batch`
fcmeans <- data.frame(t(meanfc[, -1]), stringsAsFactors = FALSE)
head(fcmeans)

# Take the mean of the repeat samples
df_flow <- fcmeans

# Calculate the average for NPCs and create the new column
df_flow$NPCs <- rowMeans(df_flow[, c("NPC1", "NPC2")])

# Remove the individual NPC columns
df_flow <- df_flow %>% select(-c("NPC1", "NPC2"))

df_flow$Astrocytes <- rowMeans(df_flow[, c("Astrocytes2", "Astrocyes1")])
df_flow <- df_flow %>% select(-c("Astrocytes2", "Astrocyes1"))

#rename columns
colnames(df_flow) <- c("IPSC","Neuron","OPC","Oligo","NPC","Astrocyte")
# make a DAneurons cell type
df_flow$DANeuron <- df_flow$Neuron
# make an NPC cell type 
df_flow$DANPC <- df_flow$NPC

head(df_flow)
#I'm going to make expression predictions from the literature for each cell type for the 3 new antibodies
# must be in the cell type order
TH <- c(-2,-2,-1,-2,2,0,10,10)
SSEA4 <- c(10,-5,3,-5,3,0,-3,0)
CD49f <- c(-3,-5,2,0,2,10,-5,2)
# add  rows for the missing antibodies

df_temp <- t(data.frame(TH,SSEA4,CD49f))
# add column names
colnames(df_temp) <- colnames(df_flow)
combined_df <- rbind(df_flow,df_temp)

# now z score the each antibody

zero.one <- t(apply(combined_df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
fcz <- zero.one

heatmap(fcz, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "FC Refernce Matrix")


# save the flow cytometery in 2D cells normalized matrix
output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(fcz, paste(output_path,"FC_ReferenceMatrix_panel2.csv",sep = ""))

```

Expression from Total RNAseq and Brain RNAseq then combine
```{r}
# Total Brain seq
input_path <- "/Users/rhalenathomas/Documents/Data/RNA/BenBarres/"
RNA_adult_brain <- read.csv(paste(input_path,"HumanBrainRNASortedCells.csv",sep = ""))
# new panel
# Protein names:"CD133" = PROM1, "CD140a" = PDGFRA, "CD15" = FUT4,   "CD184" = CXCR4,"CD24" = CD24,"CD29" = ITGB1,"CD44" = CD44, "CD56"=NCAM1,"TH"= TH, "SSEA4" = ST3GAL2,  "CD49f"=ITGA6  
marker_genes = c("CD24","NCAM1","ITGB1","FUT4","CXCR4",
                 "PROM1","CD44","PDGFRA","TH","ST3GAL2","ITGA6")
# get the expression for the genes of interest
df <- subset(RNA_adult_brain, X %in% marker_genes)
# now we want to add the gene names as the row names
rownames(df) <- df$X
colnames(df)
# remove the X column of gene names, the Tumor cells and the Whole cortex
df <- df[,6:38]
colnames(df)

# now take the average of the different cell types
df$Astrocytes.Fetal <- rowMeans(df[, c("AstrocytesFetal","AstrocytesFetal.1",
                                       "AstrocytesFetal.2","AstrocytesFetal.3",
                                       "AstrocytesFetal.4",
                                           "AstrocytesFetal.5","AstrocytesFetal.6",
                                           "AstrocytesFetal.7","AstrocytesFetal.8",
                                           "AstrocytesFetal.9")])
df$Astrocytes.Mature <- rowMeans(df[, c("AstrocytesMature","AstrocytesMature.1",
                                        "AstrocytesMature.2","AstrocytesMature.3",
                                        "AstrocytesMature.4","AstrocytesMature.5",
                                        "AstrocytesMature.6","AstrocytesMature.7",
                                        "AstrocytesMature.8","AstrocytesMature.9",
                                        "AstrocytesMature.10",
                                        "AstrocytesMature.11")])
#means of oligo
df$Oligodendrocyte <- rowMeans(df[, c("Oligo","Oligo.1","Oligo.2",
                                      "Oligo.3","Oligo.4")])
# mean from endothelial cells
df$Endothelials <- rowMeans(df[, c("Endothelial","Endothelial.1")])
# select mean cell types and all desired cell types
dfRNA <- df %>% select(c("Astrocytes.Fetal","Astrocytes.Mature","Endothelials",
                       "Oligodendrocyte","Neurons"))
# z-score
zero.one <- t(apply(dfRNA, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
fc_rnaseq <- zero.one

# quick visualization
heatmap(fc_rnaseq, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Total RNA Brain")

output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(fc_rnaseq, paste(output_path,"BrainRNABBPanel2.csv",sep = ""))

fc_rnaseq <- read.csv("/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/BrainRNABBPanel2.csv")
rownames(fc_rnaseq) <- fc_rnaseq$X
fc_rnaseq <- fc_rnaseq %>% select(-"X")

```
scRNAseq brain

```{r}
# Adult midbrain
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"
adult_midbrain <- readRDS(paste(input_path,"KamathTotal_downsample.RDS",sep = ""))
unique(adult_midbrain$Cell_Type)
# we will use the cell type (main cell types)
# get expression for each cell type
Idents(adult_midbrain) <- "Cell_Type"
# we need to remove the NaN cells
seu <- subset(adult_midbrain, idents = c("astro", "endo","olig", "nonda", "da", "opc"))
unique(seu$Cell_Type)
# scale the data
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))
expression <- FetchData(seu, vars = c("ident",marker_genes))
# get means for each cell type
ex <- as.tbl(expression)
meanEx <- ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean"))
head(meanEx)
# remove the gene name column and add these as row names
df <- as.data.frame(meanEx)
rownames(df) <- df$ident
df <- df[,2:12]
df <- as.data.frame(t(df))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","CD44","PDGFRA","TH",
                   "ST3GAL2","ITGA6")

# now we will normalize the data z-score the same as the other data
# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
midbrain <- zero.one

# quick visualization
heatmap(midbrain, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Adult Midbrain")

output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(midbrain, paste(output_path,"AdultMidbrainPanel2.csv",sep = ""))
midbrain <- read.csv("/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/AdultMidbrainPanel2.csv")
rownames(midbrain) <- midbrain$X
midbrain <- midbrain %>% select(-"X")

```

Human Fetal midbrain
```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"
# this is a singlecellexperiment object
fetalmid <- readRDS(paste(input_path,"manno_human.rds",sep = ""))
Idents(fetalmid) <- "Celltypes"
seu <- subset(fetalmid, idents = c("RG","Prog","DAN","Endothelial","Neurons","OPC","IN"))
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))
# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))
# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names
rownames(meanEx) <- meanEx$ident
head(meanEx)
df <- as.data.frame(meanEx)
rownames(df) <- df$ident
df <- df[,2:12]
df <- as.data.frame(t(df))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","CD44","PDGFRA","TH",
                   "ST3GAL2","ITGA6")
# some cell types still need to be combined
colnames(df)
# just the Neurons and IN (interneurons)
df$Neuron <- rowMeans(df[, c("Neurons","IN")])
df <- df %>% select(-c("Neurons","IN"))

# now we will normalize the data z-score the same as the other data
# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
fetalmidbrain <- zero.one

# quick visualization
heatmap(fetalmidbrain , 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Fetal Midbrain")
output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(fetalmidbrain, paste(output_path,"FeatalMidbrainPanel2.csv",sep = ""))

fetalmidbrain <- read.csv("/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/FeatalMidbrainPanel2.csv")
rownames(fetalmidbrain) <- fetalmidbrain$X
fetalmidbrain <- fetalmidbrain %>% select(-"X")


```

Developing forebrain
```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"
forebrain <- readRDS(paste(input_path,"Karolinski_DevForebrain_downsample_Level1.RDS",sep = ""))
# again there is TH in the total neurons I will see if I can find the subgroup
colnames(forebrain@meta.data)
unique(forebrain$Clusters)
DotPlot(forebrain, features = "TH", group.by = "Clusters")
# TH is expressed in: 
#Foerbrain inhibitory neuroblast, 
#Excitatory neurons possibly midbrain,
#mid/hindbrain neuroblast, 
#inhibtor neurons midbrain, possibly GABAergic

# I'll take only the cell types of interest
# IPC = intermediate progenitors

# will need to rename to capture DA neurons
Idents(forebrain) <- "Clusters"

neworder <- c("OPCs","Microglia","NPC","RG-div","Glioblast","IN","Excitatory neurons midbrain","Excitatory neurons possibly midbrain","Excitatory neurons cortex","Forebrain early neuroblast possibly GABAergic","GABAergic forebrain","Cortical Pyramidal","Radial Glia","Radial Glia potential glioblast","Radial Glia VLMC primed?","Glioblast/Pre-OPC","Forebrain neural progenitor EMX1","Endothelial","Striatum/Cortical neurons","Radial glia/Glioblast/Forebrain progenitor EMX1","Forebrain inhibitory neuroblast","VLMCs","Inhibitory neurons Midbrain, possibly GABAergic","Hindbrain neuroblast","Radial Glia/Glioblast","GABAergic or interneuron neuroblast probably midbrain","Midbrain inhibitory neuroblast","U","Mid/Hindbrain neuroblast","Midbrain/Hindbrain inhibitory neuroblast" )

Idents(forebrain) <- "Clusters"
levels(forebrain)
Idents(forebrain) <- factor(Idents(forebrain), levels = neworder)
levels(forebrain)
newcells <- c("OPCs","Microglia","NPC","RG-div","Glioblast","Cortical Interneurons","EN","DAneurons","EN","NPC","IN","EN","RG","RG","RG","Pre-OPC","NPC","Endothelial","Neurons","Glioblast","NPC","VLMCs","DAneurons","NPC","RG","IN","NPC","U","NPC-DA","NPC")

# add to seurat object
names(newcells) <- levels(forebrain)
forebrain <- RenameIdents(forebrain, newcells)
forebrain <- AddMetaData(forebrain, metadata = Idents(forebrain), col.name = "Celltypes2")
# check
table(forebrain$Celltypes2,forebrain$Clusters) # not the order is wrong
table(forebrain$Celltypes2)
unique(forebrain$Celltypes2)
# save new labels
saveRDS(forebrain, "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/DevForbrain.RDS")

# select the desired cell types
seu <- subset(forebrain, idents = c("RG","RG-div","IN","EN","DAneurons",
                                     "Endothelial","NPC","NPC-DA","Neurons","OPCs"))

seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))
# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))
# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names
rownames(meanEx) <- meanEx$ident
head(meanEx)
df <- as.data.frame(meanEx)
rownames(df) <- df$ident
df <- df[,2:12]
df <- as.data.frame(t(df))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","CD44","PDGFRA","TH",
                   "ST3GAL2","ITGA6")
# Check if some cell types still need to be combined
colnames(df)
# merge EN and IN (excitatory and inhibitor neurons)
df$Neuron <- rowMeans(df[, c("EN","IN","Neurons")])
df <- df %>% select(-c("EN","IN","Neurons"))

# now we will normalize the data z-score the same as the other data
# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
devforebrain <- zero.one

# quick visualization
heatmap(devforebrain, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Developing Forebrain")
      
output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(devforebrain, paste(output_path,"DevForebrainPanel2.csv",sep = ""))

```

Developing cortex
```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"
dev_cortex <- readRDS(paste(input_path,"Nowakowski_dev_cortext.RDS",sep = ""))
# I see there is TH expression in the "neurons" I might need to see the subtypes 
unique(dev_cortex$WGCNAcluster)
unique(dev_cortex$Celltypes)
# see which cell types when into each main group
table(dev_cortex$Celltypes,dev_cortex$WGCNAcluster)
# there are not subtypes but ages of neurons 


Idents(dev_cortex) <- "Celltypes"
seu_neurons <- subset(seu, idents = c("EN","IN"))
# nEN-early1 has all the TH expression and non  of the others do
# I need to make a new cell type meta data
DotPlot(seu_neurons, features = "TH", group.by = "WGCNAcluster")

Idents(dev_cortex) <- "WGCNAcluster"
levels(dev_cortex)
newcells <- c( "RG","RG-div","EN","EN","EN",
               "RG","RG-div","IN","IN",
               "IPC","RG" ,"IPC","IPC","OPC",
               "IN","IPC","IN","IPC",
               "EN","IN","EN","IN","IN","IN",
               "Astrocyte","IN","EN","IN","EN",
               "NeuronsDA","EN","RG","Endothelial")
# add to seurat object
names(newcells) <- levels(dev_cortex)
dev_cortex <- RenameIdents(dev_cortex, newcells)
dev_cortex <- AddMetaData(dev_cortex, metadata = Idents(dev_cortex), col.name = "Celltypes2")
# check it is correct
table(dev_cortex$Celltypes2,dev_cortex$WGCNAcluster)
# check that the correct Neurons are set as DA
DotPlot(dev_cortex, features = c("TH"), group.by = "Celltypes2")
# yes it is correct
# save the new cell types for later
saveRDS(dev_cortex,"/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/Nowakowski_dev_cortext.RDS")

# now I'll take the cell types of interest
Idents(dev_cortex) <- "Celltypes2"
seu <- subset(dev_cortex, idents = c("RG","RG-div","EN","IN","NeuronsDA","IPC","OPC",
                                        "Astrocyte","Endothelial"))
# normalize and scale
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))
# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))
# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names
rownames(meanEx) <- meanEx$ident
head(meanEx)
df <- as.data.frame(meanEx)
rownames(df) <- df$ident
df <- df[,2:12]
df <- as.data.frame(t(df))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","CD44","PDGFRA","TH",
                   "ST3GAL2","ITGA6")
# Check if some cell types still need to be combined
colnames(df)
# merge EN and IN (excitatory and inhibitor neurons)
df$Neuron <- rowMeans(df[, c("EN","IN")])
df <- df %>% select(-c("EN","IN"))
# now we will normalize the data z-score the same as the other data
# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))
# Assign the column names from the reference data frame to the normalized matrix
devcortex <- zero.one
# quick visualization
heatmap(devcortex, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Developing Cortex")

```


Merge transcriptomics from brain

```{r}
total <- fc_rnaseq
sc1 <- midbrain
sc2 <- fetalmidbrain
sc3 <- devforebrain
sc4 <- devcortex 

# Join data frames using merge
df_brain <- merge(sc1, sc2, by = "row.names")
# add the row names to the merged data frame
rownames(df_brain) <- df_brain$Row.names
# remove the newly created Row.names column
df_brain <- df_brain %>% select(-"Row.names")

df_brain <- merge(df_brain, sc3, by = "row.names")
rownames(df_brain) <- df_brain$Row.names
df_brain <- df_brain %>% select(-"Row.names")

df_brain <- merge(df_brain, sc4, by = "row.names")
rownames(df_brain) <- df_brain$Row.names
dim(df_brain)
df_brain <- df_brain[,2:29]

df_brain <- merge(df_brain, total, by = "row.names")
rownames(df_brain) <- df_brain$Row.names
dim(df_brain)
df_brain <- df_brain[,2:34]
head(df_brain)
dim(df_brain)
colnames(df_brain)
# take the means
scBrain <- df_brain
 
scBrain$neuron <- rowMeans(scBrain[, c("nonda","Neuron.x",
                                       "Neuron.y","Neuron","Neurons")],
                           na.rm = TRUE)
scBrain$neuronDA <- rowMeans(scBrain[, c("da","DAN","DAneurons","NeuronsDA")],
                           na.rm = TRUE)
scBrain$Astrocytes <- rowMeans(scBrain[, c("astro","Astrocyte",
                                          "Astrocytes.Mature")], na.rm = TRUE)
scBrain$OPC <- rowMeans(scBrain[, c("opc","OPC.x",
                                    "OPC.y","OPCs")], na.rm = TRUE)
scBrain$endothelial <- rowMeans(scBrain[, c("endo","Endothelial.x",
                                      "Endothelial.y","Endothelial",
                                      "Endothelials")], na.rm = TRUE)
scBrain$Oligo <- rowMeans(scBrain[, c("olig","Oligodendrocyte")], na.rm = TRUE)
scBrain$RadialGlia <- rowMeans(scBrain[, c("RG.x","RG.y","RG")], na.rm = TRUE)
scBrain$RadialGliaDiv <- rowMeans(scBrain[, c("RG-div.x","RG-div.y")], na.rm = TRUE)
scBrain$Stem <- rowMeans(scBrain[, c("IPC","Prog")], na.rm = TRUE)

scBrain <- scBrain %>% select(c("Astrocytes","Endothelial",
                                "neuron","neuronDA","NPC-DA","NPC",
                              "OPCs","Oligo","RadialGlia",
                              "RadialGliaDiv","Astrocytes.Fetal",
                                "Stem"))
head(scBrain)

# z-score
zero.one <- t(apply(scBrain, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
RNAbrain <- zero.one

# quick visualization
heatmap(RNAbrain, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Reference Matrix")

# save the average RNA brain expression matrix
output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(RNAbrain, paste(output_path,"BrainRNAReferenceMatrixPanel2.csv",sep = ""))


```

Get the organoid expression data

```{r}
# read in the data from the 165 days midbrain organoids
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/AST23_BrainComm/"
MO <- readRDS(paste(input_path,"SNCAandiso_labels140522.rds",sep = ""))
unique(MO$celltypes)
# I'll take only the cell types of interest
Idents(MO) <- "celltypes"
seu <- subset(MO, idents = c("RG" ,"Astrocytes",
                             "Epithelial","NPC","Neurons","DA-Neurons"))
# normalize and scale
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))

# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))

# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names

rownames(meanEx) <- meanEx$ident
head(meanEx)
dim(meanEx)
meanEx <- meanEx[,2:12]
df <- as.data.frame(t(meanEx))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","CD44","PDGFRA","TH",
                   "ST3GAL2","ITGA6")

#SSEA4 is a carbohydrate epitope of glycoproteins expressed during early development and is a marker of human embryonic stem cells. 17. The presence of SSEA4 synthase is an indicator of SSEA4 synthesis.
# ST3GAL2 is the synthase gene

head(df)
colnames(df)

# no need to merge cell types
# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scMO <- zero.one

colnames(scMO) <- c("Neurons", "Astrocytes","RG",
                    "Epithelial","DA-Neurons","DA-NPC")
# quick visualization
heatmap(scMO, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Midbrain Organoids")

output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(scMO, paste(output_path,"scMOReferenceMatrixPanel2.csv",sep = ""))

```

Cerebral orgnoids

```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"
so <- readRDS(paste(input_path,"UCSCorg30000cellsSeurat.RDS",sep = ""))

colnames(so@meta.data)
unique(so$Type)
unique(so$Subtype)
Idents(so) <- "Type"

seu <- subset(so, idents = c("ExcitatoryNeuron","Astrocyte","RadialGlia","IPC",
                             "InhibitoryNeuron"))
      
# normalize and scale
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))

# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))

# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names

rownames(meanEx) <- meanEx$ident
head(meanEx)
dim(meanEx)
meanEx <- meanEx[,2:7]
df <- as.data.frame(t(meanEx))
# some genes are missing
rownames(df)
# there are only about 7500 genes so many are missing
rownames(df) <- c("CD24","NCAM1","ITGB1","CXCR4","PROM1",
                  "ITGA6")

#missing <- c("FUT4","CD44","PDGFRA","TH","ST3GAL2")
# we need to add in rows of NaN values for the missing genes
# Create a new dataframe with NaN rows for missing genes
# Create a new dataframe with NaN rows and desired row names and column names
new_df <- data.frame(matrix(NA, nrow = 5, ncol = ncol(df)))
rownames(new_df) <- c("FUT4","CD44","PDGFRA","TH","ST3GAL2")
colnames(new_df) <- colnames(df)

# Merge the new dataframe with the existing dataframe
# Combine the existing dataframe and new dataframe
combined_df <- rbind(df, new_df)

row.names <- c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","CD44","PDGFRA","TH",
                   "ST3GAL2","ITGA6")
# reorder
# Reorder the rows based on the desired row names
reordered_df <- combined_df[row.names, ]
head(reordered_df)
# check the cell type names
colnames(reordered_df)

# [1] combine "ExcitatoryNeuron"  "InhibitoryNeuron"
# I'll take the average
df <- reordered_df

df$Neuron <- rowMeans(df[, c("ExcitatoryNeuron","InhibitoryNeuron")])
df <- df %>% select(-c("ExcitatoryNeuron","InhibitoryNeuron"))
# change to same cell type names
head(df)
colnames(df) 
# we might consider IPC as IPSC/StemCell or as NPC

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scCO <- zero.one

# quick visualization
heatmap(scCO, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Cerebral Organoids")
```

Now combine the organoids scRNA and then brain transcriptomic data

```{r}
# merge the brain orgainds

org1 <- scMO
org2 <- scCO
df_org <- merge(org1, org2, by = "row.names")
# add the row names to the merged data frame
rownames(df_org) <- df_org$Row.names
# remove the newly created Row.names column
df_org <- df_org %>% select(-"Row.names")
colnames(df_org)

# take means of the overlapping cell types
df_org$neurons <- rowMeans(df_org[, c("Neurons","Neuron")], na.rm = TRUE)
df_org$astrocytes <- rowMeans(df_org[, c("Astrocyte","Astrocytes")], na.rm = TRUE)
df_org$radialglia <- rowMeans(df_org[, c("RadialGlia","RG")], na.rm = TRUE)
df_org <- df_org %>% select(c("IPC","Epithelial","neurons","DA-Neurons",
                   "DA-NPC","astrocytes","radialglia"))
## now take means of the organoids and the brain sequencing
## first merge
df_RNA <- merge(df_org, RNAbrain, by = "row.names")
rownames(df_RNA) <- df_RNA$Row.names
df_RNA <- df_RNA %>% select(-"Row.names")
colnames(df_RNA)
# take means
df_RNA$Neuron <- rowMeans(df_RNA[, c("neuron","neurons")], na.rm = TRUE)
df_RNA$Neuron.DA <- rowMeans(df_RNA[, c("DA-Neurons","neuronDA")], na.rm = TRUE)
df_RNA$NPC.DA <- rowMeans(df_RNA[, c("DA-NPC","NPC-DA")], na.rm = TRUE)
df_RNA$StemLike <- rowMeans(df_RNA[, c("IPC","Stem")], na.rm = TRUE)
df_RNA$Astrocyte <- rowMeans(df_RNA[, c("Astrocytes","astrocytes")], na.rm = TRUE)
df_RNA$RG <- rowMeans(df_RNA[, c("radialglia","RadialGlia")], na.rm = TRUE)
colnames(df_RNA)
# select the mean values and other cell types
df_RNA <- df_RNA %>% select("Astrocyte","Astrocytes.Fetal","Epithelial",
                            "Endothelial","NPC","NPC.DA","Neuron",
                            "Neuron.DA","OPCs","Oligo","RG",
                            "RadialGliaDiv","StemLike")

# z-score
zero.one <- t(apply(df_RNA, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
z_RNA <- zero.one

# quick visualization
heatmap(z_RNA, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Organoid and Brain RNA")


```

Combine the new panel FC and transcriptomics

```{r}
# the gene names need to be changed back to the protein/antibody names
# Protein names:"CD133" = PROM1, "CD140a" = PDGFRA, "CD15" = FUT4,   "CD184" = CXCR4,"CD24" = CD24,"CD29" = ITGB1,"CD44" = CD44, "CD56"=NCAM1,"TH"= TH, "SSEA4" = ST3GAL2,  "CD49f"=ITGA6  

gene.prot <- c("CD24", "CD44","CD184","CD15","CD49f","CD29","CD56",
               "CD140a","CD133","SSEA4","TH")
df <- as.data.frame(z_RNA)
df$Markers <- gene.prot
rownames(df) <- df$Markers
z_RNA <- df[,1:13]
df_panel2 <- merge(z_RNA, fcz, by = "row.names")
colnames(df_panel2)
rownames(df_panel2) <- df_panel2$Row.names
# take means where available 
df_panel2$Astrocyte <- rowMeans(df_panel2[, c("Astrocyte.x",
                                              "Astrocyte.y")],
                                na.rm = TRUE)
df_panel2$opc <- rowMeans(df_panel2[, c("OPCs","OPC")],
                                na.rm = TRUE)
df_panel2$Oligos <- rowMeans(df_panel2[, c("Oligo.x","Oligo.y")],
                                na.rm = TRUE)
df_panel2$NPC <- rowMeans(df_panel2[, c("NPC.x","NPC.y")],
                                na.rm = TRUE)
df_panel2$NPCDA <- rowMeans(df_panel2[, c("NPC.DA","DANPC")],
                                na.rm = TRUE)
df_panel2$Neuron <- rowMeans(df_panel2[, c("Neuron.x","Neuron.y")],
                                na.rm = TRUE)
df_panel2$NeuronDA <- rowMeans(df_panel2[, c("Neuron.DA","DANeuron")],
                                na.rm = TRUE)
df_panel2$StemCell <- rowMeans(df_panel2[, c("IPSC","StemLike")],
                                na.rm = TRUE)

df_panel2 <- df_panel2 %>% select("Astrocyte","Astrocytes.Fetal",
                                  "Epithelial","Endothelial",
                                  "opc","Oligos","NPC","NPCDA",
                                  "Neuron","NeuronDA","RG",
                                  "RadialGliaDiv", "StemCell")
# z-score
zero.one <- t(apply(df_panel2, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
z_panel2 <- zero.one

# quick visualization
heatmap(z_panel2, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Reference Matrix")

# if some cell type need to remove the data should be z-scored again



# save the panel
output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(z_panel2, paste(output_path,"ReferenceMatrixPanel2.csv",sep = ""))




```



Read in Reference matrix and plot.  Change column and marker order as needed.

```{r}
df <- read.csv("/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/ReferenceMatrixPanel2.csv")

# the reference matrix for the correlation functions needs to be cell types as rows and markers as columns
mat2 <- as.data.frame(t(df))
write.csv(mat2, paste(output_path,"ReferenceMatrixPanel2Timecourse.csv",sep = ""))


# we need to remove the marker column and add these are row names
rownames(df) <- df$X
df <- df %>% select(-"X")

# heatmap function take a matix
mat <- as.matrix(df)

heatmap(mat, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Reference Matrix")



```








