---
title: "R Notebook Create Reference matrix"
output: html_notebook
---

This workbook contains:
1. Instructions on how to make a reference matrix manually or using in house and public data.
  a) Manual input values
  b) Read values from a csv (made in excel)
  c) Create a dataframe from Seurat objects.
  d) Combine different data sets to make one reference matrix.
  e) Save the reference matrix to read into 
2. Creation of the reference matrix used throughout the manuscript.
  a) Data inputs used and their sources.
  b) Combination and normalization of the inputs.
  c) Final reference matrix.
3. Creation of a second reference matrix with an adjusted antibody panel used for the time course analysis. 
  a) Data inputs used and their sources.
  b) Combination and normalization of the inputs.
  c) Final reference matrix.


To create the reference matrix you must first obtain expression data from different sources.

Not all users will be able to easily process these public data sources. However it is feasible to use excel and save one tab as a csv file then read this into R (examples below). The csv file (from excel) should formatted with column names:  
Marker, celltype1, celltype2, celltype3 ... etc
In the Marker column is the names of the antibody.  
In the other columns (cell types) are expression values.  
Values can be entered into directly into the spreadsheet by estimations from the literature or copied from browsers.  These would be qualitative estimations but will still work as a reference matrix in the correlation matrix. 

A third option is to enter values into R.  
An example is shown here.

To make your own reference matrix by estimating from publications you can score expression from 0-5 (or whatever scale you like).  Enter your own antibodies under markers.  Enter expression values for each cell type as shown here. 
```{r}
# example of directly entering values into R for a reference matrix.  Relative values estimated from literature search. 
# these values are not used in the manuscript and are placed here as an example and are not actually specific values from a literature search. 

# first we create a vector for each set of values
Marker <- c("CD24","CD56","CD29","CD15","CD184","CD133","CD71","CD44","GLAST","AQP4","HepaCAM", "CD140a","O4")

# create a vector for each cell type with an expression value for each antibody.  
# you must follow the order of the antibodies in your marker list.
StemCell <- c(5,2,5,5,5,5,1,1,1,1,1,3,0)
NPC <- c(3,3,5,5,3,3,2,1,1,1,1,1,0)
Neuron <- c(5,5,3,3,1,1,4,1,1,1,1,1,0)
Astrocyte <- c(1,1,2,1,1,1,3,4,4,5,5,2,0)
OPC <- c(2,2,2,2,2,2,2,3,3,1,1,5,2)
Oligo <- c(1,1,1,1,2,2,1,2,2,1,1,3,5)
RG <- c(3,3,4,4,5,6,4,4,4,3,1,3,0)
Endothelial <- c(2,2,2,4,4,4,3,4,1,1,1,1,0)
Epithelial <- c(1,1,2,5,1,3,2,3,3,1,2,2,0)

ref_dataframe <- data.frame(Marker, Astrocyte, Endothelial, Epithelial, Neuron, 
                            NPC, Oligo, OPC, RG, StemCell)
dim(ref_dataframe)
head(ref_dataframe)


```
Now we z-score the data frame

```{r}
df.value <- ref_dataframe[1:13, 2:10]  # Select the columns you want to normalize
zero.one <- t(apply(df.value, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
colnames(zero.one) <- colnames(df.value)
rownames(zero.one) <- ref_dataframe$Marker
head(zero.one)

# replace with normalized dataframe
ref_dataframe <- zero.one

# we can now save the reference matrix 
output_path <- "where/to/save/file/"   # example not a real file path

output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(ref_dataframe, paste(output_path,"ManualReferenceMatrix.csv",sep = ""))


```

Load libraries for data wrangling and plotting

```{r}
# load libraries
library(tidyverse)
library(ggplot2)
library(Seurat)
library(pryr) # for memory check

```

Now we can plot the reference matrix

```{r}

# we can use the normalized matrix without the column of antibodies names (these are the row names) to plot a heatmap

heatmap(ref_dataframe, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

# remember these are rough/made up values not used in the manuscript


```


# Making the reference matrix with different input data 
For antibody panel: "CD24","CD56","CD29","CD15","CD184","CD133","CD71","CD44","GLAST","AQP4","HepaCAM", "CD140a","O4"

1. 2D Flow (experimentally acquired this manuscript)
2. BenBarres bulk RNA from human isolated cells
3. scRNAseq public data from human brain
4. scRNAseq from previously acquired (and previously published) scRNAseq from midbrain organoids (165 days in final differentiation)
5. scRNAseq from cerebral organoids (public data)


Flow cytometry data from 2D cell cultures

```{r}
# read in the flow cytometry data 
# this is a csv that was created with the preprocessing workflow

input_path <- "where/you/have/the/data/file/"
input_path <- "/Users/rhalenathomas/Documents/Data/FlowCytometry/PhenoID/Analysis/2Dcells_surface/preprocessing/Allsamples/"

# this is the complete data
fc <- read.csv(paste(input_path,"2Dcells_surfacealigned_transformed_flowset.csv", sep = ""))

# we now want to get the mean expression values for each antibody in each cell type
# check what columns we have
colnames(fc)

# here "batch" is the cell type
fcf <- fc[4:17]
head(fcf)

colnames(fcf)
dim(fcf)
unique(fcf$Batch)

# now we want to make a new dataframe with mean expression values
# Calculate the mean value per column and per cell type
meanfc <- aggregate(. ~ fcf$Batch, data = fcf[, -14], FUN = mean)

# our table is transposed, we need to switch it

# set the Batch values (cell types) as rownames
rownames(meanfc) <- meanfc$`fcf$Batch`
fcmeans <- data.frame(t(meanfc[, -1]), stringsAsFactors = FALSE)
head(fcmeans)

# now we see that there are two astrocyte samples, two NPC samples, two oligodendrocytes samples.

# We will use oligo.e (early) as OPC and oligo.m (mature) as oligodendrocytes (these are still premyelinating)

# For NPC and Astrocytes we will take an average



```

```{r}
# Take the mean of the repeat samples
df_flow <- fcmeans

# Calculate the average for NPCs and create the new column
df_flow$NPCs <- rowMeans(df_flow[, c("NPC1", "NPC2")])

# Remove the individual NPC columns
df_flow <- df_flow %>% select(-c("NPC1", "NPC2"))

df_flow$Astrocytes <- rowMeans(df_flow[, c("Astrocytes2", "Astrocyes1")])
df_flow <- df_flow %>% select(-c("Astrocytes2", "Astrocyes1"))



#rename columns
colnames(df_flow) <- c("IPSC","Neuron","OPC","Oligo","NPC","Astrocyte")

head(df_flow)

```
Now z-score the flow data

```{r}
df.value <- df_flow 
zero.one <- t(apply(df.value, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
fcz <- zero.one


# we can use the normalized matrix without the column of antibodies names (these are the row names) to plot a heatmap
heatmap(fcz, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

# later we will see that the cell types not included in the flow cytometry data will have NA (empty spaces) for the missing cell types. 


# save the flow cytometery in 2D cells normalized matrix
output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(fcz, paste(output_path,"FC_ReferenceMatrix.csv",sep = ""))


```

Use RNA values from sorted human brain tissue. 
These values are from an excel spreadsheet downloaded as the supplemental data from https://www.brainrnaseq.org/
The data is in the supplemental data from doi: 10.1523/JNEUROSCI.1860-14.2014
I adjusted the column headings in excel so there is only one column name row at the top. 

```{r}

input_path <- "/Users/rhalenathomas/Documents/Data/RNA/BenBarres/"
RNA_adult_brain <- read.csv(paste(input_path,"HumanBrainRNASortedCells.csv",sep = ""))

colnames(RNA_adult_brain)
# there are multiple samples for each cell type.
# we will take averages for each cell type
# we will not included brain tumors
# first we need to filter for the genes that match our proteins targeted by the antibody panel.

```
Now we need to reformat the data frame so we can normalize that data
```{r}

# first we need to select the genes that match our proteins 
# O4 is a glycoprotein with unknown gene transcript. NKX6-2 and NKX2-2 are involved in oligodendrocyte differentiation and correlate with O4 staining

# antibody/proteins "CD24","CD56","CD29","CD15","CD184","CD133","CD71","CD44","GLAST","AQP4","HepaCAM", "CD140a","O4"
marker_genes = marker_genes = c("CD24","NCAM1","ITGB1","FUT4",
                            "CXCR4","PROM1","TFRC","CD44",
                            "SLC1A3","AQP4","HEPACAM", "PDGFRA",
                            "NKX2-2")
  

df <- subset(RNA_adult_brain, X %in% marker_genes)

# now we want to add the gene names as the row names
rownames(df) <- df$X
colnames(df)

# remove the X column of gene names, the Tumor cells and the Whole cortex
df <- df[,6:38]
colnames(df)

# now take the average of the different cell types

df$Astrocytes.Fetal <- rowMeans(df[, c("AstrocytesFetal",
                                           "AstrocytesFetal.1","AstrocytesFetal.2",
                                           "AstrocytesFetal.3","AstrocytesFetal.4",
                                           "AstrocytesFetal.5","AstrocytesFetal.6",
                                           "AstrocytesFetal.7","AstrocytesFetal.8",
                                           "AstrocytesFetal.9")])
head(df)
# Remove the individual columns
df <- df %>% select(-c("AstrocytesFetal",
                                           "AstrocytesFetal.1","AstrocytesFetal.2",
                                           "AstrocytesFetal.3","AstrocytesFetal.4",
                                           "AstrocytesFetal.5","AstrocytesFetal.6",
                                           "AstrocytesFetal.7","AstrocytesFetal.8",
                                           "AstrocytesFetal.9"))
head(df)

df$Astrocytes.Mature <- rowMeans(df[, c("AstrocytesMature","AstrocytesMature.1",
                                        "AstrocytesMature.2","AstrocytesMature.3",
                                        "AstrocytesMature.4","AstrocytesMature.5",
                                        "AstrocytesMature.6","AstrocytesMature.7",
                                        "AstrocytesMature.8","AstrocytesMature.9",
                                        "AstrocytesMature.10",
                                        "AstrocytesMature.11")])

# remove individual columns
df <- df %>% select(-c("AstrocytesMature","AstrocytesMature.1",
                                        "AstrocytesMature.2","AstrocytesMature.3",
                                        "AstrocytesMature.4","AstrocytesMature.5",
                                        "AstrocytesMature.6","AstrocytesMature.7",
                                        "AstrocytesMature.8","AstrocytesMature.9",
                                        "AstrocytesMature.10",
                                        "AstrocytesMature.11"))
colnames(df)

#means of oligo
df$Oligodendrocyte <- rowMeans(df[, c("Oligo","Oligo.1","Oligo.2",
                                      "Oligo.3","Oligo.4")])

df <- df %>% select(-c("Oligo","Oligo.1","Oligo.2", 
                       "Oligo.3","Oligo.4"))

colnames(df)

# means of macrophage/microglia - these will be removed later as we do not expect these cells in midbrain organoids
df$marcrophage.microglia <- rowMeans(df[, c("MacrophageMicroglia",
                                            "MacrophageMicroglia.1",
                                            "MacrophageMicroglia.2")])

df <- df %>% select(-c("MacrophageMicroglia","MacrophageMicroglia.1",
                                            "MacrophageMicroglia.2"))

colnames(df)

# mean from endothelial cells
df$marcrophage.microglia <- rowMeans(df[, c("Endothelial","Endothelial.1")])

df <- df %>% select(-c("Endothelial","Endothelial.1"))

colnames(df)
class(df)
head(df)

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
fc_rnaseq <- zero.one

# quick visualization
heatmap(fc_rnaseq, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

# the normalization is across each row (gene) and is effected by the cell types included.
# we can see that mature astrocytes match expected expression (Glast/SLC1A3, HEPACAM, AQP4 and CD79/TFRC and low in CD44) This matches the flow data except for Glast.  The fetal astroctyes match expection of glia stem cells/progenitors, higher in CD44 and precursor markers CD133/PROM1, CD184/CXCR4, unexpectedly the fetal astrocytes are high in CD24, reported high in neurons and stem cells. 

```
```{r}
# z-score the total RNAseq brain data without microglia

df <- df %>% select(-c("marcrophage.microglia"))
colnames(df)

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
fc_rnaseq <- zero.one

# quick visualization
heatmap(fc_rnaseq, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")


```
Adult brain data from single cell sequencing
All data sets were processed into Seurat objects.

1. Adult midbrain: https://doi.org/10.1038/s41593-022-01061-1.  (Kamath/Macosko) 
2. Developing brain: https://cells.ucsc.edu/?ds=cortex-dev DOI: 10.1126/science.aap8809 (Nowakowski/Kriegstein)


```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"

adult_midbrain <- readRDS(paste(input_path,"KamathTotal_downsample.RDS",sep = ""))

colnames(adult_midbrain@meta.data)
unique(adult_midbrain$Cell_Subtype)
unique(adult_midbrain$Cell_Type)

# we will use the cell type (main cell types)

```

```{r}
# get the mean expression values for each gene matching the protein targets in our antibody panel
# use seurat function mydatafetch

# get expression for each cell type
Idents(adult_midbrain) <- "Cell_Type"
unique(adult_midbrain$Cell_Type)
# we need to remove the NaN cells
seu <- subset(adult_midbrain, idents = c("astro", "endo",  "mg",
                                                   "olig", "nonda", "da", "opc"))
unique(seu$Cell_Type)
# scale the data
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))

expression <- FetchData(seu, vars = c("ident",marker_genes))


# get means for each cell type
ex <- as.tbl(expression)
meanEx <- ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean"))
head(meanEx)
# remove the gene name column and add these as row names

df <- as.data.frame(meanEx)
rownames(df) <- df$ident
df <- df[,2:14]
df <- as.data.frame(t(df))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")

# now we will normalize the data z-score the same as the other data
# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scAmidbrain <- zero.one

# quick visualization
heatmap(scAmidbrain, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

# as we won't separate DA neurons from other neurons I will take the mean here
# I will remove microglia


```
Clean up the dataframe

```{r}
# as we won't separate DA neurons from other neurons I will take the mean here
# I will remove microglia

df$Neuron <- rowMeans(df[, c("da","nonda")])
df <- df %>% select(-c("mg","da","nonda"))
# change to same cell type names
colnames(df) <- c("Astrocyte","Endothelial","Oligodendrocyte","OPC","Neuron")

# now we will normalize the data z-score the same as the other data
# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scAmidbrain <- zero.one

# quick visualization
heatmap(scAmidbrain, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

# there is no expression of CD24 in this data

```

Developing cortex
```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"

dev_cortex <- readRDS(paste(input_path,"Nowakowski_dev_cortext.RDS",sep = ""))

colnames(dev_cortex@meta.data)
unique(dev_cortex$Celltypes)
# I'll take only the cell types of interest
# IPC = intermediate progenitors
Idents(dev_cortex) <- "Celltypes"
seu <- subset(dev_cortex, idents = c("RG","EN","IN","IPC","OPC",
                                        "Astrocyte","Endothelial"))
# normalize and scale
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))


marker_genes = marker_genes = c("CD24","NCAM1","ITGB1","FUT4",
                            "CXCR4","PROM1","TFRC","CD44",
                            "SLC1A3","AQP4","HEPACAM", "PDGFRA",
                            "NKX2-2","NKX2.2")
# the genes are named differently in this data set

# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))

# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names

rownames(meanEx) <- meanEx$ident
head(meanEx)
meanEx <- meanEx[,2:14]
df <- as.data.frame(t(meanEx))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")

head(df)

# EN and IN are excitation and inhibitory neurons
# I'll take the average

df$Neuron <- rowMeans(df[, c("IN","EN")])
df <- df %>% select(-c("IN","EN"))
# change to same cell type names
head(df)
colnames(df) 
# we might consider IPC as IPSC/StemCell or as NPC

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scDevCor <- zero.one

# quick visualization
heatmap(scDevCor, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")




```

Midbrain and striatum two ages.
Bhaduri https://doi.org/10.1038/s41586-021-03910-8
https://cells.ucsc.edu/?ds=dev-brain-regions

```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"

midbrain_str <- readRDS(paste(input_path,"Bhaduri_midbrain_striatum.RDS",sep = ""))
colnames(midbrain_str@meta.data)
unique(midbrain_str$cell_type)
unique(midbrain_str$age)
unique(midbrain_str$cell_class) # cell class same as cell type
unique(midbrain_str$cell_cluster)
# I'll take only the cell types of interest
# IPC = intermediate progenitors
Idents(midbrain_str) <- "cell_type"
seu <- subset(midbrain_str, idents = c("dividing" ,"neuron","radial glia",
                                     "endothelial","oligodendrocyte",
                                        "astrocyte","interneuron"))
      
# normalize and scale
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))


marker_genes = marker_genes = c("CD24","NCAM1","ITGB1","FUT4",
                            "CXCR4","PROM1","TFRC","CD44",
                            "SLC1A3","AQP4","HEPACAM", "PDGFRA",
                            "NKX2-2","NKX2.2")
# the genes are named differently in this data set

# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))

# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names

rownames(meanEx) <- meanEx$ident
head(meanEx)
meanEx <- meanEx[,2:14]
df <- as.data.frame(t(meanEx))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")

head(df)
colnames(df)

# Neuron and Interneuron are excitation and inhibitory neurons
# I'll take the average

df$Neuron <- rowMeans(df[, c("neuron","interneuron")])
df <- df %>% select(-c("neuron","interneuron"))
# change to same cell type names
head(df)
colnames(df) 
# we might consider IPC as IPSC/StemCell or as NPC

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scMidStr <- zero.one

# quick visualization
heatmap(scMidStr, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")



```

Developing Forebrain
https://cells.ucsc.edu/?ds=human-forebraindev+sc-rna-seq

```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"

forebrain <- readRDS(paste(input_path,"Karolinski_DevForebrain_downsample_Level1.RDS",sep = ""))
colnames(forebrain@meta.data)
unique(forebrain$Age)
unique(forebrain$Level1)
unique(forebrain$Celltypes)
# I'll take only the cell types of interest
# IPC = intermediate progenitors
Idents(forebrain) <- "Celltypes"

seu <- subset(forebrain, idents = c("RG" ,"IN","EN",
                                     "Endothelial","NPC","Neurons","OPC"))
      
# normalize and scale
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))


marker_genes = marker_genes = c("CD24","NCAM1","ITGB1","FUT4",
                            "CXCR4","PROM1","TFRC","CD44",
                            "SLC1A3","AQP4","HEPACAM", "PDGFRA",
                            "NKX2-2","NKX2.2")
# the genes are named differently in this data set

# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))

# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names

rownames(meanEx) <- meanEx$ident
head(meanEx)
meanEx <- meanEx[,2:14]
df <- as.data.frame(t(meanEx))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")

head(df)
colnames(df)

# Neuron and Interneuron are excitation and inhibitory neurons
# I'll take the average

df$Neuron <- rowMeans(df[, c("Neurons","IN","EN")])
df <- df %>% select(-c("Neurons","IN","EN"))
# change to same cell type names
head(df)
colnames(df) 
# we might consider IPC as IPSC/StemCell or as NPC

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scForebrain <- zero.one

# quick visualization
heatmap(scForebrain, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")
```

Fetal midbrain https://doi.org/10.1016/j.cell.2016.09.027
La Manno

```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"
# this is a singlecellexperiment object
fetalmid <- readRDS(paste(input_path,"manno_human.rds",sep = ""))
# convert to seurat
seu <- as.Seurat(fetalmid, counts = "counts", data = "logcounts")
colnames(seu@meta.data)
unique(seu$cell_type1)
unique(seu$Source)
Idents(seu) <- "cell_type1"
levels(seu)
# there too many cell types these need to be grouped at a higher resolution
# [1] "eNb1"     "eNb2"     "eNb3"     "eNb4"     "eProg1a"  "eProg1b" 
# [7] "eProg2a"  "eProg2b"  "eRgla"    "eRglb"    "eRglc"    "eRgld"   
#[13] "eRgle"    "eRglf"    "eSCa"     "eSCb"     "eSCc"     "hDA0"    
#[19] "hDA1"     "hDA2"     "hEndo"    "hGaba"    "hMgl"     "hNbGaba" 
#[25] "hNbM"     "hNbML1"   "hNbML5"   "hNProg"   "hOMTN"    "hOPC"    
#31] "hPeric"   "hProgBP"  "hProgFPL" "hProgFPM" "hProgM"   "hRgl1"   
#[37] "hRgl2a"   "hRgl2b"   "hRgl2c"   "hRgl3"    "hRN"      "hSert"   
#[43] "iDAa"     "iDAb"     "iDAc"     "iMN1"     "iMN2"     "iNb1"    
#[49] "iNb2"     "iProg1"   "iProg2"   "iProg3"   "iRgl1"    "iRgl2"   
#[55] "iRN"      "Unk"    

# cell type main groups
celltypes <- c("NB","NB","NB","NB","Prog","Prog",
               "Prog","Prog","RG","RG","RG","RG",
               "RG","RG","ESC","ESC","ESC","DAN",
               "DAN","DAN","Endothelial","IN","MGL","IN",
               "NB","NB","NB","Prog","OMTN","OPC",
               "Pericyte","Prog","Prog","Prog","Prog","RG",
               "RG","RG","RG","RG","RN","Neurons",
               "DAN","DAN","DAN","MN","MN","NB",
               "NB","Prog","Prog","Prog","RG","RG",
               "RG","Unk")

names(celltypes) <- levels(seu)
seu <- RenameIdents(seu, celltypes)
seu <- AddMetaData(seu, metadata = Idents(seu), col.name = "Celltypes")

table(seu$Celltypes,seu$cell_type1)
unique(seu$Celltypes)
# select only the cell types I want to use
Idents(seu) <- "Celltypes"
seu <- subset(seu, idents = c("RG","Prog","DAN","Endothelial","Neurons","OPC","IN"))

seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))


marker_genes = marker_genes = c("CD24","NCAM1","ITGB1","FUT4",
                            "CXCR4","PROM1","TFRC","CD44",
                            "SLC1A3","AQP4","HEPACAM", "PDGFRA",
                            "NKX2-2","NKX2.2")
# the genes are named differently in this data set

# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))

# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names

rownames(meanEx) <- meanEx$ident
head(meanEx)
meanEx <- meanEx[,2:14]
df <- as.data.frame(t(meanEx))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")

head(df)
colnames(df)

# Neurons, DAN and IN are all neurons 
# I'll take the average

df$Neuron <- rowMeans(df[, c("Neurons","IN","DAN")])
df <- df %>% select(-c("Neurons","IN","DAN"))
# change to same cell type names
head(df)
colnames(df) 
# we might consider IPC as IPSC/StemCell or as NPC

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scManno <- zero.one

# quick visualization
heatmap(scManno, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

```



Now we will merge the brain scRNAseq data sets and the total RNA data set

```{r}
# we need to make sure the marker names and the cell type names match

sc1 <- scAmidbrain
sc2 <- scDevCor
sc3 <- scForebrain
sc4 <- scMidStr
sc5 <- scManno
total <- fc_rnaseq

rownames(sc1)
rownames(sc2)
rownames(sc3)
rownames(sc4)
rownames(sc5)
rownames(total)

# the total RNA has a different order to the rows
# reorder the rows

# Get the row order of sc1 in total
row_order <- match(rownames(sc1), rownames(total))

# Reorder rows of total based on sc1's order
t1 <- total[row_order, ]

rownames(t1)
rownames(total)
rownames(sc1)

```
Merge and take the average of the two data sets and then take the mean value when there is overlap

```{r}
# I want to merge by gene/protein 

# Join dataframes using merge
df_brain <- merge(sc1, sc2, by = "row.names")
# add the row names to the merged data frame
rownames(df_brain) <- df_brain$Row.names
# remove the newly created Row.names column
df_brain <- df_brain %>% select(-"Row.names")

df_brain <- merge(df_brain, sc3, by = "row.names")
rownames(df_brain) <- df_brain$Row.names
df_brain <- df_brain %>% select(-"Row.names")

df_brain <- merge(df_brain, sc4, by = "row.names")
rownames(df_brain) <- df_brain$Row.names
df_brain <- df_brain[,2:23]

df_brain <- merge(df_brain, sc5, by = "row.names")
rownames(df_brain) <- df_brain$Row.names
dim(df_brain)
df_brain <- df_brain[,2:28]

df_brain <- merge(df_brain, t1, by = "row.names")
rownames(df_brain) <- df_brain$Row.names
dim(df_brain)
df_brain <- df_brain[,2:32]

head(df_brain)
dim(df_brain)
colnames(df_brain)

# take the means
# copy the variable in case of an error to not have to run the above code again
scBrain <- df_brain

scBrain$neuron <- rowMeans(scBrain[, c("Neuron.x", "Neuron.y","Neuron.x.1",
                                       "Neuron.y.1", "Neurons","Neuron")], na.rm = TRUE)
scBrain$Astrocyte <- rowMeans(scBrain[, c("Astrocyte.x", "Astrocyte.y",
                                          "astrocyte","Astrocytes.Mature")], na.rm = TRUE)
scBrain$OPC <- rowMeans(scBrain[, c("OPC.x", "OPC.y","OPC.y.1","OPC.x.1")], na.rm = TRUE)
scBrain$Endothelial <- rowMeans(scBrain[, c("Endothelial.x", "Endothelial.y",
                                            "Endothelial.x.1","Endothelial.y.1",
                                            "endothelial")], na.rm = TRUE)
scBrain$Oligo <- rowMeans(scBrain[, c("Oligodendrocyte.x","oligodendrocyte",
                                     "Oligodendrocyte.y"), na.rm = TRUE])

scBrain$RadialGlia <- rowMeans(scBrain[, c("RG.x","RG.y","radial glia","RG")], na.rm = TRUE)
scBrain$Stem <- rowMeans(scBrain[, c("IPC","dividing","Prog")], na.rm = TRUE)

scBrain <- scBrain %>% select(c("Astrocyte","Endothelial","OPC",
                                "Oligo","neuron","RadialGlia","Stem"))
head(scBrain)

# z-score
zero.one <- t(apply(scBrain, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
RNAbrain <- zero.one

# quick visualization
heatmap(RNAbrain, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

# save the average RNA brain experession matrix
output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(RNAbrain, paste(output_path,"BrainRNAReferenceMatrix.csv",sep = ""))


```

scRNAseq from organoids
1. Midbrain organoids 165 days AST23 and isogenic control https://doi.org/10.1093/braincomms/fcab223

2.Cerebral organoids


AST23 and isogenic control 165 days in final differentiation media

```{r}

# read in the data from the 165 days midbrain organoids
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/AST23_BrainComm/"
MO <- readRDS(paste(input_path,"SNCAandiso_labels140522.rds",sep = ""))
colnames(MO@meta.data)
unique(MO$celltypes)
unique(MO$cluster.ids)
# I'll take only the cell types of interest

Idents(MO) <- "celltypes"

seu <- subset(MO, idents = c("RG" ,"Astrocytes",
                                     "Epithelial","NPC","Neurons","DA-Neurons"))
      
# normalize and scale
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))


marker_genes = marker_genes = c("CD24","NCAM1","ITGB1","FUT4",
                            "CXCR4","PROM1","TFRC","CD44",
                            "SLC1A3","AQP4","HEPACAM", "PDGFRA",
                            "NKX2-2")

# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))

# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names

rownames(meanEx) <- meanEx$ident
head(meanEx)
meanEx <- meanEx[,2:14]
df <- as.data.frame(t(meanEx))
rownames(df) <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")

head(df)
colnames(df)

# Neuron and DA neurons
# I'll take the average

df$Neuron <- rowMeans(df[, c("Neurons","DA-Neurons")])
df <- df %>% select(-c("Neurons","DA-Neurons"))
# change to same cell type names
head(df)
colnames(df) 
# we might consider IPC as IPSC/StemCell or as NPC

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scMO <- zero.one

# quick visualization
heatmap(scMO, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

output_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"
write.csv(scMO, paste(output_path,"scMOReferenceMatrix.csv",sep = ""))


```

Cerebral organoids

Tanaka et al. 2020. Cell Rep. https://cells.ucsc.edu/?ds=organoidatlas&meta=Cluster


```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/"
so <- readRDS(paste(input_path,"UCSCorg30000cellsSeurat.RDS"))

colnames(so@meta.data)
unique(so$Type)
unique(so$Subtype)
Idents(so) <- "Type"
# select the cell types 
#[1] "ExcitatoryNeuron" "RadialGlia"       "Astrocyte"       
#[4] "IPC"              "InhibitoryNeuron" "Outlier"         
#[7] "Unknown"
seu <- subset(so, idents = c("ExcitatoryNeuron","Astrocyte","RadialGlia","IPC",
                             "InhibitoryNeuron"))
      
# normalize and scale
seu <- NormalizeData(seu, features = rownames(seu))
seu <- ScaleData(seu, features = rownames(seu))


marker_genes = marker_genes = c("CD24","NCAM1","ITGB1","FUT4",
                            "CXCR4","PROM1","TFRC","CD44",
                            "SLC1A3","AQP4","HEPACAM", "PDGFRA",
                            "NKX2-2")

# check for genes
DoHeatmap(seu, features = marker_genes)

# get the expression data
expression <- FetchData(seu, vars = c("ident",marker_genes))

# get means for each cell type
ex <- as.data.frame(expression)
meanEx <- as.data.frame(ex %>% group_by(ident) %>% summarise_all(.funs = c(mean="mean")))
head(meanEx)
# remove the gene name column and add these as row names

rownames(meanEx) <- meanEx$ident
head(meanEx)
dim(meanEx)
meanEx <- meanEx[,2:8]
df <- as.data.frame(t(meanEx))
# some genes are missing
rownames(df)

rownames(df) <- c("CD24","NCAM1","ITGB1","CXCR4","PROM1","TRFC",
                  "SLC1A3")
# present: "CD24_mean"   "NCAM1_mean"  "ITGB1_mean"  "CXCR4_mean" 
# "PROM1_mean"  "TFRC_mean"   "SLC1A3_mean"
#absent NKX2-2, PDGFRA, HEPACAM, AQP4, CD44, FUT4
row.names <-  c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")

# we need to add in rows of NaN values for the missing genes
# Create a new dataframe with NaN rows for missing genes
# Create a new dataframe with NaN rows and desired row names and column names
new_df <- data.frame(matrix(NA, nrow = 6, ncol = ncol(df)))
rownames(new_df) <- c("NKX2-2", "PDGFRA", "HEPACAM",
                      "AQP4", "CD44", "FUT4")
colnames(new_df) <- colnames(df)

# Merge the new dataframe with the existing dataframe
# Combine the existing dataframe and new dataframe
combined_df <- rbind(df, new_df)

# reorder
# Reorder the rows based on the desired row names
reordered_df <- combined_df[row.names, ]
head(reordered_df)
# check the cell type names
colnames(reordered_df)

# [1] combine "ExcitatoryNeuron"  "InhibitoryNeuron"
# I'll take the average
df <- reordered_df

df$Neuron <- rowMeans(df[, c("ExcitatoryNeuron","InhibitoryNeuron")])
df <- df %>% select(-c("ExcitatoryNeuron","InhibitoryNeuron"))
# change to same cell type names
head(df)
colnames(df) 
# we might consider IPC as IPSC/StemCell or as NPC

# z-score
zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
scCO <- zero.one

# quick visualization
heatmap(scCO, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")


```


Take the mean of the two organoid datasets

```{r}

scMO <- read.csv("/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/scMOReferenceMatrix.csv")
rownames(scMO) <- scMO$X
scMO <- scMO %>% select(-"X")
sc1 <- scMO
rownames(sc1)
sc2 <- scCO
rownames(sc2)
rownames(sc2) <- c("CD24","NCAM1","ITGB1","FUT4",  
                   "CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4",
                   "HEPACAM", "PDGFRA","NKX2-2")
# Join dataframes using merge
df_Org <- merge(sc1, sc2, by = "row.names")
# add the row names to the merged data frame
rownames(df_Org) <- df_Org$Row.names
head(df_Org)
colnames(df_Org)
# remove the newly created Row.names column
df_Org <- df_Org %>% select(-"Row.names")

# take the means
# copy the variable in case of an error to not have to run the above code again
dfOrganoid <- df_Org
#df_Org <- dfOrganoid
colnames(df_Org)

df_Org$Neuron <- rowMeans(df_Org[, c("Neuron.x", "Neuron.y")], na.rm = TRUE)
df_Org$Astro <- rowMeans(df_Org[, c("Astrocytes", "Astrocyte")], na.rm = TRUE)

df_Org$Radial_Glia <- rowMeans(df_Org[, c("RG", "RadialGlia")], na.rm = TRUE)

df_Org <- df_Org %>% select(c("Astro","Neuron","Radial_Glia","NPC",
                              "IPC","Epithelial"))
zero.one <- t(apply(df_Org, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
organoid <- zero.one

# quick visualization
heatmap(organoid, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")

# save the organoid dataframe
write.csv(organoid,"/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/OrganoidReferenceMatrix.csv")


```


Finally I will combine the different inputs.
1. FC protein
2. Brain RNAseq
3. Organoid RNAseq

I will first take the mean of the transcriptomics then take the mean of the result with the FC levels to weight the protein higher in the correlation matrix.

```{r}
# I will read in the saved csv files and will have 3 dataframes with a column for the rownames

input_path <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/"

fc <- read.csv(paste(input_path,"FC_ReferenceMatrix.csv",sep = ""))
brain <- read.csv(paste(input_path,"BrainRNAReferenceMatrix.csv",sep = ""))
organoid <- read.csv(paste(input_path,"OrganoidReferenceMatrix.csv",sep = ""))


```

Merge the transcriptomics and take means of the same cell types

```{r}

rna_merge <- merge(brain,organoid, by = "X")
colnames(rna_merge)

# take means
rna_merge$Neurons <- rowMeans(rna_merge[, c("Neuron", "neuron")], na.rm = TRUE)
rna_merge$Astrocytes <- rowMeans(rna_merge[, c("Astrocyte", "Astro")], na.rm = TRUE)
rna_merge$StemCell <- rowMeans(rna_merge[, c("IPC", "Stem")], na.rm = TRUE)
rna_merge$RG <- rowMeans(rna_merge[, c("RadialGlia", "Radial_Glia")], na.rm = TRUE)
colnames(rna_merge)

df <- rna_merge %>% select(c("Endothelial","Epithelial", "Neurons",
                             "Astrocytes","StemCell","RG","OPC","NPC",
                             "Oligo"))

rownames(df) <- rna_merge$X

zero.one <- t(apply(df, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
trans_df <- zero.one

# quick visualization
heatmap(trans_df, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Heatmap of Normalized Matrix")



```


Now take the mean of the transcription and the FC protein levels

```{r}

# We need to change the gene names into the protein names
protein.names <- fc$X

rownames(fc) <- fc$X

gene.names <- rownames(trans_df)
# see the two lists
gene.names
protein.names
# rename the rownames of the transcription df with protein names
# the order needs to be the same as the gene names, which is not the same as the proteins because they are sorted alphabetically
gene.protein <- c("AQP4","CD24","CD44","CD184","CD15",   
                  "HepaCAM","CD29", "CD56",  "O4", "CD140a",
                  "CD133","GLAST", "CD71")

rownames(trans_df) <- gene.protein

# merge the dataframes by rownames
df_merge <- merge(fc,trans_df, by = 0)
# add in the rownames
rownames(df_merge) <- df_merge$Row.names

df_merge <- df_merge %>% select(-c("X","Row.names"))
# see the column names to know which to take the average
colnames(df_merge)

df <- df_merge
# take means
df$neurons <- rowMeans(df[, c("Neuron", "Neurons")], na.rm = TRUE)
df$astrocytes <- rowMeans(df[, c("Astrocyte", "Astrocytes")], na.rm = TRUE)
df$stemlike <- rowMeans(df[, c("IPSC", "StemCell")], na.rm = TRUE)
df$NPC <- rowMeans(df[, c("NPC.x", "NPC.y")], na.rm = TRUE)
df$OPC <- rowMeans(df[, c("OPC.x", "OPC.y")], na.rm = TRUE)
df$NPC <- rowMeans(df[, c("NPC.x", "NPC.y")], na.rm = TRUE)
df$oligodendrocyte <- rowMeans(df[, c("Oligo.x", "Oligo.y")], na.rm = TRUE)

# check the new column names
colnames(df)

df_final <- df %>% select(c("astrocytes","Endothelial","Epithelial",
                            "neurons","NPC","OPC","oligodendrocyte",
                            "RG","stemlike"))


zero.one <- t(apply(df_final, MARGIN = 1, FUN = function(X) (X - min(X, na.rm = TRUE))/diff(range(X, na.rm = TRUE))))

# Assign the column names from the reference data frame to the normalized matrix
ref_mat <- zero.one

# quick visualization
heatmap(ref_mat, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Refernce Matrix")





```

