---
title: "Figure 6"
output: github_document
---

# Figure 6
scRNAseq analysis of FAC sorted populations

# In this workbook.

Read in the labelled object 
Run the correlation comparison between the total scRNA per FACS pop and the AB levels
Calculate the proportion of the cell types from the original labels
UMAP of the cell POPS merged
UMAP of the subtypes labelled


```{r}
# load required libraries
library(Seurat)
library(dplyr)
library('reshape2') 
library("ggplot2")


```


Read in the labelled object single cell sequencing

```{r}

output_path <- "path/to/save/to/"
# saved in the figures folder 
seu.sc <- readRDS(paste(output_path,"Labelled4Pops.RDS",sep=""))


```


Read in the Flow Cytometry data

```{r}

seu.fc <- readRDS("PostSortedFlowExpressionseu.RDS")
unique(seu.fc$GatedPop)

```

Marker list

```{r}

marker.ab <- c("CD24","CD56","CD29","CD15","CD184","CD133","CD71","CD44","GLAST","AQP4","HepaCAM", "CD140a","O4")

marker.genes <-c("CD24","NCAM1","ITGB1","FUT4","CXCR4","PROM1","TFRC","CD44","SLC1A3","AQP4","HEPACAM", "PDGFRA","NKX6-2")

```


Make the expression matrixes

```{r}
Idents(seu.sc) <- 'orig.ident'
DefaultAssay(seu.sc) <- 'RNA'

# to include all genes in scale data add the features argument
seu.sc <- ScaleData(object = seu.sc, features = rownames(seu.sc))


scRNAseq.mean <- as.data.frame(AverageExpression(seu.sc,features = marker.genes, assays = 'RNA',
                                   group.by = 'orig.ident', slot= 'scale.data'))


# change gene names to protein/AB names
rownames(scRNAseq.mean) <- marker.ab

# get the mean expression from FACS 
DefaultAssay(seu.fc) <- 'RNA'
Idents(seu.fc) <- 'GatedPop'

#seu.fc <- ScaleData(seu.fc)

FC.mean <- as.data.frame(AverageExpression(seu.fc, assays = 'RNA',features = marker.ab,
                                   group.by = 'GatedPop', slot= 'scale.data'))

```

Calculate correlation between two matrixes

```{r}

# rename and reorder dataframes to match
colnames(FC.mean)
colnames(FC.mean) <- c("Astrocytes","RadialGlia","Neurons2","Neurons1")
colnames(FC.mean)

colnames(scRNAseq.mean)
colnames(scRNAseq.mean) <- c("Neurons2","Neurons1","Astrocytes","RadialGlia")
colnames(scRNAseq.mean)
rna.mean <- scRNAseq.mean %>% dplyr::select("Astrocytes","RadialGlia","Neurons1","Neurons2")
FC.mean <- FC.mean %>% dplyr::select("Astrocytes","RadialGlia","Neurons1","Neurons2")

colnames(rna.mean)
colnames(FC.mean)
df.cor <- cor(FC.mean, rna.mean)
# first one is the rows and second is the columns, FC measures on the y axis

rownames(FC.mean)
rownames(rna.mean)

# try calculating z scores
# this works on rows
data <- FC.mean
add.rownames <- rownames(data)
FC.mean.z <- as.data.frame(sapply(data, function(data) (data-mean(data))/sd(data)))

rownames(FC.mean.z) <-add.rownames
FC.mean.z


data <- rna.mean
add.rownames <- rownames(data)
rna.mean.z <- as.data.frame(sapply(data, function(data) (data-mean(data))/sd(data)))

rownames(rna.mean.z) <-add.rownames
rna.mean.z


df.cor <- cor(FC.mean.z, rna.mean.z)
df.cor

```


Plot expression RNA x FC for each antibody 

```{r}

# Combine the data into a single dataframe
combined_data <- data.frame(
  Marker = rep(rownames(FC.mean), each = ncol(FC.mean)),
  Cell_Type = rep(colnames(FC.mean), times = nrow(FC.mean)),
  FC_Value = as.vector(t(FC.mean)),
  RNA_Value = as.vector(t(rna.mean))
)

# Create the dot plot
gg <- ggplot(combined_data, aes(x = FC_Value, y = RNA_Value, color = Marker)) +
  geom_point() +
  facet_wrap(~ Cell_Type, scales = "free") +
  labs(x = "FC Expression Values", y = "RNA Expression Values", color = "Marker") +
  theme_minimal()

# Calculate correlation coefficient (R) for each facet
correlation_data <- combined_data %>%
  group_by(Cell_Type) %>%
  summarize(correlation = cor(FC_Value, RNA_Value))

# Add linear regression lines
gg <- gg +
  geom_smooth(method = "lm", se = FALSE, color = "black", linetype = "dashed") # Add linear regression line
  
# Annotate plot with correlation coefficient
gg <- gg +
  geom_text(data = correlation_data, aes(label = paste("R =", round(correlation, 2))), 
            x = -Inf, y = Inf, hjust = 0, vjust = 1, size = 4, color = "black", 
            inherit.aes = FALSE, show.legend = FALSE)

gg


# Function to z-score normalize a matrix
z_score_normalize <- function(mat) {
  apply(mat, 2, function(x) (x - mean(x)) / sd(x))
}

# Function to min-max scale a matrix
min_max_scale <- function(mat) {
  apply(mat, 2, function(x) (x - min(x)) / (max(x) - min(x)))
}

# Z-score normalize the FC and RNA matrices
FC_normalized <- min_max_scale(z_score_normalize(as.matrix(FC.mean)))
RNA_normalized <- min_max_scale(z_score_normalize(as.matrix(rna.mean)))

# Combine the data into a single dataframe
combined_data <- data.frame(
  Marker = rep(rownames(FC_normalized), each = ncol(FC_normalized)),
  Cell_Type = rep(colnames(FC_normalized), times = nrow(FC_normalized)),
  FC_Value = as.vector(t(FC_normalized)),
  RNA_Value = as.vector(t(RNA_normalized))
)

# Create the dot plot
gg <- ggplot(combined_data, aes(x = FC_Value, y = RNA_Value, color = Marker)) +
  geom_point() +
  facet_wrap(~ Cell_Type, scales = "free") +
  labs(x = "FC Expression Values", y = "RNA Expression Values", color = "Marker") +
  theme_minimal()

# Calculate correlation coefficient (R) for each facet
correlation_data <- combined_data %>%
  group_by(Cell_Type) %>%
  summarize(correlation = cor(FC_Value, RNA_Value))

# Add linear regression lines
gg <- gg +
  geom_smooth(method = "lm", se = FALSE, color = "black", linetype = "dashed") # Add linear regression line
  
# Annotate plot with correlation coefficient
gg <- gg +
  geom_text(data = correlation_data, aes(label = paste("R =", round(correlation, 2))), 
            x = -Inf, y = Inf, hjust = 0, vjust = 1, size = 4, color = "black", 
            inherit.aes = FALSE, show.legend = FALSE)

gg

```
Plot FC vs RNA for each marker by cell type

```{r}


# Reshape the FC dataframe
melted_FC <- FC.mean.z %>%
  rownames_to_column(var = "Marker") %>%
  pivot_longer(cols = -Marker, names_to = "Cell_type", values_to = "FC_expression")

# Reshape the RNA dataframe
melted_RNA <- rna.mean.z %>%
  rownames_to_column(var = "Marker") %>%
  pivot_longer(cols = -Marker, names_to = "Cell_type", values_to = "RNA_expression")

# Merge the melted dataframes
combined_data <- merge(melted_FC, melted_RNA, by = c("Marker", "Cell_type"))


correlation_data2 <- combined_data %>%
  group_by(Marker) %>%
  summarise(correlation = cor(FC_expression, RNA_expression, use = "pairwise.complete.obs"))

correlation_data2


plot2 <- ggplot(combined_data, aes(x = FC_expression, y = RNA_expression, color = Cell_type)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, aes(group = 1), linetype = "dashed", color = "black") + # Add linear regression line
  facet_wrap(~ Marker, scales = "free") + # Facet by Marker
  labs(x = "FC Expression", y = "RNA Expression", color = "Cell Types") +
  theme_minimal() +
  geom_text(data = correlation_data2, aes(x = Inf, y = -Inf, label = paste("R =", round(correlation, 2))),
            hjust = 1, vjust = 0, size = 3, inherit.aes = FALSE) + # Annotate with R values
  theme(legend.position = "bottom") # Place legend at the bottom

plot2

```

# Figure 6C
Plot the heatmap of the correlation between Flow (protein) and scRNAseq (RNA)

```{r}
# need to melt the matrix
library(reshape2)
library(ggplot2)

longData<- melt(df.cor)

head(longData)
colnames(longData) <- c("FC","scRNA","R2")

ggplot(longData, aes(x=scRNA,y=FC, fill =R2)) + geom_tile() +
  scale_fill_gradient2(low = "#075AFF",
                       mid = "#FFFFCC",
                       high = "#FF0000") +
   guides(fill = guide_colourbar(label = TRUE,
                                ticks = FALSE)) + theme_bw() +
  coord_fixed()  +scale_x_discrete(expand=c(0,0))+
  scale_y_discrete(expand=c(0,0))+ theme(text = element_text(size=16, colour = "black"),
        axis.text.x = element_text(size = 18, colour = "black", angle = 90), axis.text.y = element_text(size= 18, colour = "black"))  + xlab('scRNAseq') + ylab('Flow Cytometry') 


# save plot
pdf(paste(output_path,"FC_scRNA_correlation.pdf"), width = 8, height = 5)
ggplot(longData, aes(x=scRNA,y=FC, fill =R2)) + geom_tile() +
  scale_fill_gradient2(low = "#075AFF",
                       mid = "#FFFFCC",
                       high = "#FF0000", midpoint = 0, limit = c(-0.5,0.5)) +
   guides(fill = guide_colourbar(label = TRUE,
                                ticks = FALSE)) + theme_bw() +
  coord_fixed()  +scale_x_discrete(expand=c(0,0)) +
  scale_y_discrete(expand=c(0,0)) + 
  theme(text = element_text(size=16, colour = "black"),
        axis.text.x = element_text(size = 16, colour = "black", angle = 90), axis.text.y = element_text(size= 16, colour = "black"))  + xlab('scRNAseq') + ylab('Flow Cytometry') 
dev.off()

```

# Figure 6D
UMAP of the different FACS sorted populations in the merged data object.

```{r}

sample.order <- c("Astrocytes","RadialGlia","Neurons1","Neurons2")
sample.order <- rev(sample.order)

# colour order to match cell type order
clust.colours <- c("royalblue", "indianred2","springgreen4","palegreen2")
 
Idents(seu.sc) <- 'orig.ident'
     
DimPlot(seu.sc, order = sample.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, pt.size = 0.1, label = FALSE)

### Figure 6D

pdf(paste(output_path,"UMAPscRNAseqMerge4.pdf"),width = 7, height = 4)
DimPlot(seu.sc, order = sample.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, pt.size = 0.1, label = FALSE, label.size = 6) +
  theme(legend.text = element_text(size=16), axis.title.y = element_text(size=16), 
        axis.title.x = element_text(size=16), axis.text.y = element_text(size =16),
        axis.text.x = element_text(size =16))
dev.off()


## Figure S19 

cell.order <- c("Astrocytes","Endothelial","Glia","NPC","Neurons","Other","Radial Glia")
cell.order <- rev(cell.order)

# colour order to match cell type order
clust.colours <- c("chocolate1","deepskyblue","steelblue4","red2","mediumpurple3","burlywood3", 
                   "pink2")

 
Idents(seu.sc) <- 'Cell_Types'
      
# designated the order of the splits factor
seu.sc$orig.ident <- factor(x = seu.sc$orig.ident, levels = c("Astrocytes","RadialGlia","Neurons1","Neurons2"))


DimPlot(seu.sc, order = cell.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, pt.size = 0.1, label = FALSE, split.by = 'orig.ident', ncol = 2)



pdf(paste(output_path,"UMAP_merge_splitbyorigident.pdf"),width = 12, height = 7.5)
DimPlot(seu.sc, order = cell.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, pt.size = 0.1, label = FALSE, split.by = 'orig.ident', label.size = 6, ncol = 2) +
  theme(legend.text = element_text(size=16), axis.title.y = element_text(size=16), 
        axis.title.x = element_text(size=16), axis.text.y = element_text(size =16),
        axis.text.x = element_text(size =16))
dev.off()


# also plot one not split to see the whole thing


pdf(paste(output_path,"UMAP_merge_CellTypes.pdf"),width = 6.7, height = 4.1)
DimPlot(seu.sc, order = cell.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, pt.size = 0.1, label = FALSE, label.size = 6) +
  theme(legend.text = element_text(size=16), axis.title.y = element_text(size=16), 
        axis.title.x = element_text(size=16), axis.text.y = element_text(size =16),
        axis.text.x = element_text(size =16))
dev.off()





```

# Figure 6E
Proportions of main cell types in each sorted population

```{r}

sample.lables <- as.data.frame(table(seu.sc$orig.ident, seu.sc$Cell_Type2))

sample.lables$Freq <- as.double(sample.lables$Freq)


pop.order <- c("Astrocytes","RadialGlia","Neurons1","Neurons2")

sample.order <- c("Astrocytes","Mix","DANeurons","Neurons",
                  "NPC","RadialGlia")

clust.colours <- c("#FF9400",         # oranges Astrocytes  
                   "steelblue",   # mix 
                   "#528A47",    # greens DA neurons   
                   "#8E36D2", # neurons 2  CP, GRIA2
                   "red2", # NPC
                   "#FCB1C6"
                   )


df <- sample.lables %>% mutate(Var2 = factor(Var2, levels = sample.order))
df <- df %>% mutate(Var1 = factor(Var1, levels = pop.order))


ggplot(df, aes(x = Var1,y=Freq ,fill = Var2)) + geom_bar(position= "fill", stat = "identity")  +
  RotatedAxis() +
  scale_y_continuous(labels = scales::percent_format()) + theme_classic() + theme(text = element_text(size=15),
        axis.text.x = element_text(angle=90, hjust=1))  + xlab('Sorted Population') + ylab('Proportion of Cell Subtype')  + scale_fill_manual(values = clust.colours)



pdf(paste(output_path,"Fig6FbarPopsMaintypes.pdf"), width = 4, height = 5.5)
ggplot(df, aes(x = Var1,y=Freq ,fill = Var2)) + geom_bar(position= "fill", stat = "identity")  +
  RotatedAxis() +
  scale_y_continuous() + theme_classic() + theme(text = element_text(size=15),
        axis.text.x = element_text(angle=90, hjust=1))  + xlab('Sorted Population') + ylab('Proportion of Cell Types')  + scale_fill_manual(values = clust.colours)
dev.off()

```


# Figure 6F
UMAP coloured by subtypes of cells


```{r}
# choose the order for cell types to appear in the legend
# by default this will go by cluster size
sample.order <- c("Astrocytes-immature","Astrocytes-reactive", "Astrocytes-resting",
                  "DANeurons-SN","DANeurons-VM","DANeurons-VTA",
                  "Mix",
                  "Neurons-Endocrine","Neurons-Excitatory","Neurons-GABAergic",
                  "Neurons-Glutamatergic","Neurons-Mature","Neurons_stem_potential",
                  "Neurons-5HT","NPC",
                  "RadialGlia-VZ", "RadialGlia-SVZ","RadialGlia-outer",
                  "RadialGlia-NSC", "RadiaGlia-proliferating")
sample.order <- rev(sample.order)


clust.colours <- c("#F1C166","#FF9400","#E95901",          # oranges Astrocytes
                   "#02E8BA","#528A47","#0AE90C",     # greens DA neurons
                   "steelblue",   # mix 
                   "#E3C3FC", "#B166F6","#917EA5",
                   "#6B03F6", "#9770CB","#C9B5EF",   # neurons 2  CP, GRIA2
                   "#8E36D2", "red2", # NPC
                   "#C092A1","#F5C3FC","#F74D84"
                   ,"#FCB1C6","#FA7A92"
                   )

#DimPlot(seu.sc, pt.size = 0.1, order = sample.order, label = TRUE, group.by = 'Cell_Subtype_Markers')
Idents(seu.sc) <- 'Subtypes' 

DimPlot(seu.sc, order = sample.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, pt.size = 0.25, label = TRUE, repel = TRUE)
DimPlot(seu.sc, order = sample.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, pt.size = 0.25, label = FALSE)



pdf(paste(output_path,"UMAP_CellSubtypes.pdf"),width = 10, height = 5)
DimPlot(seu.sc, order = sample.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, pt.size = 0.25, label = FALSE)
dev.off()

# Figures S25
# Make the split dim plots with the same colours
Idents(seu.sc) <- 'Subtypes' 

DimPlot(seu.sc, order = sample.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, split.by = "orig.ident",pt.size = 0.25, label = FALSE, ncol = 2)


pdf(paste(output_path,"UMAP_CellSubtypesSplit.pdf"),width = 10, height = 5)
DimPlot(seu.sc, order = sample.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, split.by = "orig.ident",pt.size = 0.25, label = FALSE, ncol = 2)
dev.off()



```

# Figure 6G
Proportions of DA subtypes in neurons1 and neurons2

```{r}
# get count of cell subtypes for each sorted population
# get all neuron subtypes
sample.lables <- as.data.frame(table(seu.sc$orig.ident, seu.sc$Subtypes))

sample.lables$Freq <- as.double(sample.lables$Freq)

sample.table <- sample.lables %>% filter(Var1 %in% c("Neurons1","Neurons2"))


sample.table.da <- sample.table %>% filter(Var2 %in% c("DANeurons-SN","DANeurons-VM","DANeurons-VTA"))


pop.order <- c("Neurons1","Neurons2")

# for DA neurons in the two neurons pops
sample.order <- c("DANeurons-SN","DANeurons-VM","DANeurons-VTA")
df <- sample.table.da %>% mutate(Var2 = factor(Var2, levels = sample.order))
df <- df %>% mutate(Var1 = factor(Var1, levels = pop.order))
clust.colours<- c("#02E8BA","#528A47","#0AE90C")


ggplot(df, aes(x = Var1,y=Freq ,fill = Var2)) + geom_bar(position= "fill", stat = "identity")  +
  RotatedAxis() +
  scale_y_continuous(labels = scales::percent_format()) + theme_classic() + theme(text = element_text(size=15),
        axis.text.x = element_text(angle=90, hjust=1))  + xlab('Sorted Population') + ylab('Proportion of Cell Subtype')  + scale_fill_manual(values = clust.colours)






```



Figure S19
counts of subtypes for each main cell type, grouped by the four sorted populations.

```{r}
# subset the DA neurons
Idents(seu.sc) <- "Cell_Type2"
seu.da <- subset(seu.sc, idents = "DANeurons")


sample.lables <- as.data.frame(table(seu.da$orig.ident, seu.da$Subtypes))
sample.lables$Freq <- as.double(sample.lables$Freq)

sample.table <- sample.lables %>% filter(Var2 %in% c("DANeurons-VTA","DANeurons-VM","DANeurons-SN"))

pop.order <- c("Astrocytes","RadialGlia","Neurons1","Neurons2")

sample.order <- c("DANeurons-VTA","DANeurons-VM","DANeurons-SN")



clust.colours <- c("#02E8BA","#528A47","#0AE90C"
                   )


df <- sample.table %>% mutate(Var2 = factor(Var2, levels = sample.order))
df <- df %>% mutate(Var1 = factor(Var1, levels = pop.order))


# 
ggplot(df, aes(x = Var1, y= Freq ,fill = factor(Var2))) + geom_bar(stat = "identity", position = "stack")  + 
  RotatedAxis() +
 theme_classic() + theme(text = element_text(size=15),
        axis.text.x = element_text(angle=90, hjust=1))  + xlab('Sorted Population') + ylab('Number of Cells')  + scale_fill_manual(values = clust.colours)



pdf(paste(output_path,"Fig6G_CountDAsub.pdf"), width = 5, height = 5)
ggplot(df, aes(x = Var1, y= Freq ,fill = factor(Var2))) + geom_bar(stat = "identity", position = "stack")  + 
  RotatedAxis() +
 theme_classic() + theme(text = element_text(size=16),
        axis.text.x = element_text(angle=90, hjust=1))  + xlab('Sorted Population') + ylab('Number of Cells')  + scale_fill_manual(values = clust.colours)
dev.off()


```

Counts for astrocytes

```{r}

# make a data frame with the subtype counts by FACS population
sample.lables <- as.data.frame(table(seu.sc$orig.ident, seu.sc$Subtypes))

sample.lables$Freq <- as.double(sample.lables$Freq)

# filter for the astrocytes
sample.table.a <- sample.lables %>% filter(Var2 %in% c("Astrocytes-immature","Astrocytes-reactive", "Astrocytes-resting"))

pop.order <- c("Astrocytes","RadialGlia","Neurons1","Neurons2")

sample.order <- c("Astrocytes-immature","Astrocytes-reactive", "Astrocytes-resting")
clust.colours <- c("#F1C166","#FF9400","#E95901")         # oranges Astrocytes


df <- sample.table.a %>% mutate(Var2 = factor(Var2, levels = sample.order))
df <- df %>% mutate(Var1 = factor(Var1, levels = pop.order))


# 
ggplot(df, aes(x = Var1, y= Freq ,fill = factor(Var2))) + geom_bar(stat = "identity", position = "stack")  + 
  RotatedAxis() +
 theme_classic() + theme(text = element_text(size=15),
        axis.text.x = element_text(angle=90, hjust=1))  + xlab('Sorted Population') + ylab('Number of Cells')  + scale_fill_manual(values = clust.colours)



pdf(paste(output_path,"Fig6SubAstrocounts.pdf"), width = 5, height = 5)
ggplot(df, aes(x = Var1, y= Freq ,fill = factor(Var2))) + geom_bar(stat = "identity", position = "stack")  + 
  RotatedAxis() +
 theme_classic() + theme(text = element_text(size=16),
        axis.text.x = element_text(angle=90, hjust=1))  + xlab('Sorted Population') + ylab('Number of Cells')  + scale_fill_manual(values = clust.colours)
dev.off()


```

Counts for Neurons

```{r}
sample.lables <- as.data.frame(table(seu.sc$orig.ident, seu.sc$Subtypes))

sample.lables$Freq <- as.double(sample.lables$Freq)

sample.table.n <- sample.lables %>% filter(Var2 %in% c("Neurons-Endocrine","Neurons-Excitatory",
                  "Neurons-GABAergic",
                  "Neurons-Glutamatergic","Neurons-Mature",
                  "Neurons_stem_potential",
                  "Neurons-5HT","NPC"))

pop.order <- c("Astrocytes","RadialGlia","Neurons1","Neurons2")

sample.order <- c("Neurons-Endocrine","Neurons-Excitatory",
                  "Neurons-GABAergic",
                  "Neurons-Glutamatergic","Neurons-Mature",
                  "Neurons_stem_potential",
                  "Neurons-5HT","NPC")

clust.colours <- c("#E3C3FC", "#B166F6","#917EA5",
                   "#6B03F6", "#9770CB","#C9B5EF","#8E36D2", "red2")




df <- sample.table.n %>% mutate(Var2 = factor(Var2, levels = sample.order))
df <- df %>% mutate(Var1 = factor(Var1, levels = pop.order))


# 
ggplot(df, aes(x = Var1, y= Freq ,fill = factor(Var2))) + geom_bar(stat = "identity", position = "stack")  + 
  RotatedAxis() +
  scale_y_continuous(expand = c(0,0)) +
 theme_classic() + theme(text = element_text(size=15),
        axis.text.x = element_text(angle=90, hjust=1))  + xlab('Sorted Population') + ylab('Number of Cells')  + scale_fill_manual(values = clust.colours)



pdf(paste(output_path,"Fig6Sub_NeuronsCount.pdf"), width = 5.5, height = 5)
ggplot(df, aes(x = Var1, y= Freq ,fill = factor(Var2))) + geom_bar(stat = "identity", position = "stack")  + 
  RotatedAxis() +
  scale_y_continuous(expand = c(0,0)) +
 theme_classic() + theme(text = element_text(size=16),
        axis.text.x = element_text(angle=90, hjust=1))  + xlab('Sorted Population') + ylab('Number of Cells')  + scale_fill_manual(values = clust.colours)
dev.off()


```

Counts for Radial Glia

```{r}

sample.lables <- as.data.frame(table(seu.sc$orig.ident, seu.sc$Subtypes))

sample.lables$Freq <- as.double(sample.lables$Freq)

sample.table.r <- sample.lables %>% filter(Var2 %in% c("RadialGlia-VZ", "RadialGlia-SVZ","RadialGlia-outer",
                  "RadialGlia-NSC", "RadiaGlia-proliferating"))

pop.order <- c("Astrocytes","RadialGlia","Neurons1","Neurons2")

sample.order <- c("RadialGlia-VZ", "RadialGlia-SVZ","RadialGlia-outer",
                  "RadialGlia-NSC", "RadiaGlia-proliferating")



clust.colours <- c( "#C092A1","#F5C3FC","#F74D84"
                   ,"#FCB1C6","#FA7A92"
                   )




df <- sample.table.r %>% mutate(Var2 = factor(Var2, levels = sample.order))
df <- df %>% mutate(Var1 = factor(Var1, levels = pop.order))


# 
ggplot(df, aes(x = Var1, y= Freq ,fill = factor(Var2))) + geom_bar(stat = "identity", position = "stack")  + 
  RotatedAxis() +
 theme_classic() + theme(text = element_text(size=15),
        axis.text.x = element_text(angle=90, hjust=1))  + xlab('Sorted Population') + ylab('Number of Cells')  + scale_fill_manual(values = clust.colours)



pdf(paste(output_path,"Fig6Sub_RGcounts.pdf"), width = 5.5, height = 5)
ggplot(df, aes(x = Var1, y= Freq ,fill = factor(Var2))) + geom_bar(stat = "identity", position = "stack")  + 
  RotatedAxis() +
 theme_classic() + theme(text = element_text(size=16),
        axis.text.x = element_text(angle=90, hjust=1))  + xlab('Sorted Population') + ylab('Number of Cells')  + scale_fill_manual(values = clust.colours)
dev.off()


```



For Table S8

```{r}

# get all neuron subtypes
sample.lables <- as.data.frame(table(seu.sc$orig.ident, seu.sc$Subtypes))

s.table <- reshape(sample.lables, idvar = "Var2", timevar = "Var1", direction = "wide")

head(s.table)
write.csv(s.table, paste(output_path,"freq_subtypes_Feb7.csv",sep = ""))


# table of main cell types
sample.lables <- as.data.frame(table(seu.sc$orig.ident, seu.sc$Cell_Type2))

s.table <- reshape(sample.lables, idvar = "Var2", timevar = "Var1", direction = "wide")

head(s.table)
write.csv(s.table, paste(output_path,"freq_MaintypeFreq_Feb8.csv",sep = ""))




```

Subtype markers
DotPlots of markers in cell type subsets


```{r}

# neuron subtype
n.markers <- c("ASCL1","CP","GRIA2","MGP","SPARCL1","TPH1","TFPI2","CD24")

DotPlot(neurons, features = n.markers) + 
  theme(text = element_text(size=16, colour = "black"),axis.text.x = element_text(angle = 90))


# DA neuron subytype markers
da.markers <- c("RAB3B","TPBG","TTR","TH","SOX6","CALB1","SLC17A6")

DotPlot(neurons.da, features = da.markers) + 
  theme(text = element_text(size=16, colour = "black"),axis.text.x = element_text(angle = 90))

# new markers
da1 <- c("RAB3B","SCG2","CALB1","RASGRP","STMN2","ADCYAP1","PTPRO")
da2 <- c("TPBG","PTGDS","CD9","DLK1","SPARC","RBP1","HES1")
da3 <- c("TTR","NEUROD1","NDUFA4","SYT1","SLC1A2","OLFM2","FAM19A4")
da1 <- c("RAB3B","SCG2","CALB1","RASGRP","ADCYAP1","PTPRO")
da2 <- c("TPBG","PTGDS","CD9","DLK1","SPARC","RBP1")
da3 <- c("NEUROD1","NDUFA4","SYT1","SLC1A2","OLFM2","FAM19A4")
da.markers <- c(da1,da2,da3)
DotPlot(neurons.da, features = da.markers) + 
  theme(text = element_text(size=16, colour = "black"),axis.text.x = element_text(angle = 90))

pdf(paste(output_path,"Dotplot_DANeuSubMarkersJan25.pdf",sep = ""),width = 7, height = 4)
DotPlot(neurons.da, features = da.markers) + 
  theme(text = element_text(size=16, colour = "black"),axis.text.x = element_text(angle = 90))
dev.off()


rg.markers <- c("CYP1B1","NEAT1","PTN","RPL41","TOP2A","VCAN","SOX2","SLIT2","HES1","VIM")

pdf(paste(output_path,"Dotplot_RGMarkersOct17.pdf",sep = ""),width = 7, height = 4)
DotPlot(rg, features = rg.markers) + 
  theme(text = element_text(size=16, colour = "black"),axis.text.x = element_text(angle = 90))
dev.off()

# Apoe, Gfap, Aqp4 and Slc1a3
astro.markers <- c("COL3A1","FABP5","HPD","APOE","S100B","IGFBP2","DBI",
                   "PRSS56", "IGTP","LFIT3","LIGP1",
                   "COL1A2")

pdf(paste(output_path,"Dotplot_AstroMarkersOct17.pdf",sep = ""),width = 7, height = 4)
DotPlot(astro, features = astro.markers) + 
  theme(text = element_text(size=16, colour = "black"),axis.text.x = element_text(angle = 90))
dev.off()


```



Proportion of cell types tests
Compare FACS Neurons1 to Neurons2
Figure S20

Check the proportions of subtypes of DA neurons
Check proportions of all main cell types

```{r}

library("scProportionTest")

prop_test <- sc_utils(seu.sc)

prop_test <- permutation_test(
	prop_test, cluster_identity = "Cell_Type2",
	sample_1 = "Neurons1", sample_2 = "Neurons2",
	sample_identity = "orig.ident"
)

permutation_plot(prop_test)


prop_test <- sc_utils(seu.da)

prop_test <- permutation_test(
	prop_test, cluster_identity = "Subtypes",
	sample_1 = "Neurons1", sample_2 = "Neurons2",
	sample_identity = "orig.ident"
)

permutation_plot(prop_test)


```


