---
title: "R Notebook"
output: html_notebook
---

1. Get the proportions of main cell types from each FACS sorted sample
2. 


```{r}

library(Seurat)
library(dplyr)
library(Matrix)
library(ggplot2)


```


Read in the 4 sorted Populations separately processed and annotated

```{r}

pathway <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PhenoID/scRNAseqSorted/objs/"

Neurons1 <- readRDS(paste(pathway,"Neuron1LabledSeu30092022.RDS",sep = ""))
Neurons2 <- readRDS(paste(pathway,"Neurons2LabelsSeu30092022.RDS",sep = ""))
Astro <- readRDS(paste(pathway,"Glia1LabledSeu301102022.RDS",sep = ""))
RG <- readRDS(paste(pathway,"Glia2LabledSeu03102022.RDS",sep = ""))



```


Check each object for main cell type labels

```{r}

table(Neurons1$Cell_Types)
dim(Neurons1)  # total cells 1723

table(Neurons2$Cell_Types)
dim(Neurons2) # total cells 8884

table(Astro$Cell_Type)
dim(Astro) # 20000

table(RG$Cell_Types)
dim(RG)

```


Fix each separate populations 
- Add FACS sort name
- Down sample
- check proportions are maintained
- rename Cell_Type in Astro to be the same name for 'Cell_Types'


```{r}

# cell type variables weren't named the same and the project ID are not the population names
# I'll further down sample Glia1 (LaunchSample3)

# change orig.ident names
Idents(Neurons2) <- 'orig.ident'
Neurons2$orig.ident <- 'Neurons2'
Idents(Astro) <- 'orig.ident'
Astro$orig.ident <- 'Astrocytes'
Idents(RG) <- 'orig.ident'
RG$orig.ident <- 'RadialGlia'


# rename metadata 
Idents(Astro) <- 'Cell_Type'
Astro$Cell_Types <- Idents(Astro)
table(Astro$Cell_Types)
Astro$Cell_Type <- NULL

# down sample
Idents(Astro) <- 'orig.ident'
Astro.sub <- subset(Astro, downsample = 9000)
table(Astro.sub$Cell_Types)
# good proportions are maintained
# save downsampled object






all.pops <- merge(Neurons1, y = c(Neurons2, Astro.sub, RG), project = "PhenoID")


```
Try to down sample to get closer proportions of starting populations

```{r}

# down sample
Idents(Astro) <- 'orig.ident'
Astro.sub <- subset(Astro, downsample = 3000)
table(Astro$Cell_Types)
table(Astro.sub$Cell_Types)

# down sample
Idents(Neurons2) <- 'orig.ident'
Neurons2.sub <- subset(Neurons2, downsample = 2000)
dim(Neurons2)
table(Neurons2$Cell_Types)
table(Neurons2.sub$Cell_Types)

# down sample
Idents(RG) <- 'orig.ident'
RG.sub <- subset(RG, downsample = 2000)
dim(RG)
table(RG$Cell_Types)
table(RG.sub$Cell_Types)

all.pops <- merge(Neurons1, y = c(Neurons2.sub, Astro.sub, RG.sub), project = "PhenoID")
table(all.pops$orig.ident)


```





Check merge

```{r}

table(all.pops$orig.ident)
table(all.pops$Cell_Types)

# I might want to down sample all the other populations accept Neurons1 
# they cluster fairly well
# save the merged object 
pathway <- "/Users/rhalenathomas/Documents/Data/scRNAseq/PhenoID/scRNAseqSorted/objs/"
saveRDS(all.pops, paste(pathway,"All4popssub07102022.RDS"))

#seu.q <- readRDS(paste(pathway,"All4pops07102022.RDS"))
#seu.q <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PhenoID/scRNAseqSorted/objs/All4popssub07102022.RDS")

```

Normalize and cluster

```{r}

# standard workflow 
# repeated same for all.pops and all.pops sub
seu <- NormalizeData(all.pops, normalization.method = "LogNormalize", scale.factor = 10000)
seu <- FindVariableFeatures(seu, selection.method = "vst", nfeatures = 2500)
seu <- ScaleData(seu)
seu <- RunPCA(seu)
ElbowPlot(seu, ndims = 30)
# best choice PC = 15

# didn't save the objects need to repeat
seu.q <- NormalizeData(seu.q, normalization.method = "LogNormalize", scale.factor = 10000)
seu.q <- FindVariableFeatures(seu.q, selection.method = "vst", nfeatures = 2500)
seu.q <- ScaleData(seu.q)
seu.q <- RunPCA(seu.q)
ElbowPlot(seu.q, ndims = 30)




```

Find clusters

```{r}

seu <- RunUMAP(seu, reduction = "pca", n.neighbors = 25, dims = 1:20)
DimPlot(seu, reduction = "umap")

seu.q <- FindNeighbors(seu, dims = 1:20, k.param = 25)
seu.q <- FindClusters(seu.q, resolution = c(0,0.2,0.6,1,1.2, 1.5, 1.8))

# clear out other clustering to be able to use clustree
library(clustree)

#columns.to.remove <- c("RNA_snn_res.0.05", "RNA_snn_res.0.1","RNA_snn_res.0.25","RNA_snn_res.0.5",
#                       "RNA_snn_res.0.4","RNA_snn_res.0.8")
#for(i in columns.to.remove) {
#  seu.q[[i]] <- NULL
#}


#DimPlot(seu.q, reduction = "umap", group.by = 'RNA_snn_res.0.05')
#DimPlot(seu.q, reduction = "umap", group.by = 'RNA_snn_res.0.1')
DimPlot(seu.q, reduction = "umap", group.by = 'RNA_snn_res.0.2', label = TRUE)
#DimPlot(seu.q, reduction = "umap", group.by = 'RNA_snn_res.0.4')
DimPlot(seu.q, reduction = "umap", group.by = 'RNA_snn_res.0.6',label = TRUE)
DimPlot(seu.q, reduction = "umap", group.by = 'RNA_snn_res.1', label = TRUE)
DimPlot(seu.q, reduction = "umap", group.by = 'RNA_snn_res.1.2', label = TRUE)
DimPlot(seu.q, reduction = "umap", group.by = 'RNA_snn_res.1.8', label = TRUE)
DimPlot(seu.q, group.by = 'orig.ident', label = TRUE)
#VlnPlot(seu.q, features = 'nFeature_RNA', group.by = 'orig.ident')
DimPlot(seu.q, group.by = 'Cell_Types', label = TRUE)

DimPlot(seu.q, group.by = 'Cell_Types', split.by = 'orig.ident')
DimPlot(seu.q, group.by = 'RNA_snn_res.1.8', split.by = 'orig.ident')

# repeat when object wasn't saved
seu.q <- RunUMAP(seu.q, reduction = "pca", n.neighbors = 25, dims = 1:20)
DimPlot(seu.q, reduction = "umap")

seu.q <- FindNeighbors(seu.q, dims = 1:20, k.param = 25)
seu.q <- FindClusters(seu.q, resolution = c(0,0.2,0.6,1,1.2, 1.8))

# clear out other clustering to be able to use clustree
library(clustree)

columns.to.remove <- c("RNA_snn_res.0.05", "RNA_snn_res.0.1","RNA_snn_res.0.25","RNA_snn_res.0.5",
                       "RNA_snn_res.0.4","RNA_snn_res.0.8")
for(i in columns.to.remove) {
  seu.q[[i]] <- NULL
}



DimPlot(seu.q, reduction = "umap", group.by = 'RNA_snn_res.1.8', label = TRUE)
DimPlot(seu.q, group.by = 'orig.ident', label = TRUE)
DimPlot(seu.q, group.by = 'Cell_Types', label = TRUE)
DimPlot(seu.q, group.by = 'Cell_Types', split.by = 'orig.ident')
DimPlot(seu.q, group.by = 'RNA_snn_res.1.8', split.by = 'orig.ident')


clustree(seu.q)



```

Number FACS populations in each cluster
Number of Cell Types in each cluster


```{r}

saveRDS(seu.q, paste("/Users/rhalenathomas/Documents/Data/scRNAseq/PhenoID/scRNAseqSorted/objs/Merge4PopsLabels.RDS"))

seu.q <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PhenoID/scRNAseqSorted/objs/Merge4PopsLabels.RDS")

DimPlot(seu.q, group.by = 'Level3')
DimPlot(seu.q, group.by = 'Levels')

```




```{r}

# FACS populations per cluster
table(seu.q$RNA_snn_res.1.8, seu.q$orig.ident)

# Cell Types per cluster
table(seu.q$RNA_snn_res.1.8, seu.q$Cell_Types)

```


Re annotate these clusters with subgroups

```{r}

Idents(seu.q) <- 'RNA_snn_res.1.8'
# from k.param 49
cluster.ids <- c("Astro1","Neurons1","RG1","Astro2",
                 "RG2","Neurons2","Neurons3","Astro3",
                 "Neurons3","Astro4","Astro5","Neurons4",
                 "RG3","RG4","Neurons5","Neurons6",
                 "Endothelial","Neurons7")
# from k.param 249
cluster.ids <- c("RG1","Neurons1","Astro1","Astro2","Neurons2",
                 "Astro-Mix","Neurons-RG","RG2",
                 "Neurons3",
                 "Endothelial")
# k.param 125
cluster.ids <- c("Astro1","Neurons1","Astro2","Neurons1",
                 "RG1-Neurons","Neurons2","Astro3-Neurons-RG","RG1",
                 "Neurons3","Neurons4","RG2",
                 "Endothelial","Mix")
# k.param 25
cluster.ids <- c("Astro1","Neurons1","RG-Neurons1","Astro2","Neurons2",
                 "Astro-RG","Astro3","Neurons2","RG2","Neurons3",
                 "Astro3","Neurons4","RG3","Neurons4","Neurons5",
                 "Mix","RG4","Endothelial","Neurons6","RG-Neurons2",
                 "Neurons7","Glia","RG5","Mix","Mix"
                 )


names(cluster.ids) <- levels(seu.q)
seu.q <- RenameIdents(seu.q, cluster.ids)
seu.q$Cluster.ids <- Idents(seu.q)

DimPlot(seu.q, reduction = "umap", label = TRUE, group.by = 'Cluster.ids', repel = TRUE)
DimPlot(seu.q, reduction = "umap", label = TRUE, group.by = 'Cluster.ids',split.by = 'orig.ident')

# some cluster are problematic 
# cluster 3 has 650 RG and 180 Neurons


Idents(seu.q) <- 'RNA_snn_res.1.8'

cluster.ids <- c("Astro","Neurons","RG","Astro",
                 "RG","Neurons","Neurons","Astro",
                 "Neurons","Astro","Astro","Neurons",
                 "RG","RG","Neurons","Neurons",
                 "Endothelial","Neurons")

cluster.ids <- c("RG","Neurons","Astro","Astro","Neurons",
                 "Astro-Mix","Neurons-RG","RG",
                 "Neurons",
                 "Endothelial")

# k.param 125
cluster.ids <- c("Astro","Neurons","Astro","Neurons",
                 "RG1-Neurons","Neurons","Astro3-Neurons-RG","RG",
                 "Neurons","Neurons","RG",
                 "Endothelial","Mix")
# K 25
cluster.ids <- c("Astro","Neurons","RG-Neur","Astro","Neurons",
                 "Astro-RG","Astro","Neurons","RG","Neurons",
                 "Astro","Neurons","RG","Neurons","Neurons",
                 "Mix","RG","Endothelial","Neurons","RG-Neur",
                 "Neurons","Glia","RG","Mix","Mix"
                 )

names(cluster.ids) <- levels(seu.q)
seu.q <- RenameIdents(seu.q, cluster.ids)
seu.q$Level1 <- Idents(seu.q)

DimPlot(seu.q, reduction = "umap", label = TRUE, group.by = 'Level1', repel = TRUE)
DimPlot(seu.q, reduction = "umap", label = TRUE, group.by = 'Level1',split.by = 'orig.ident')



```






Make a table of the proportion of cell types and a plot

```{r}

# cell type labels done separately 

df <- as.data.frame(table(seu.q$orig.ident, seu.q$Cell_Types))


cell.order <- c("Astrocytes","Endothelial","Glia","Neurons","NPC","Other","Radial Glia")
cell.order <- rev(cell.order)


ggplot(df, aes(x = Var1, y = Freq, fill = Var2)) + 
  geom_col(position = "fill") + 
  
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, size = 12), axis.text.y = element_text(size = 12)) +
  labs(y = "Proportion of Cells", x = "") + theme_classic() +
  theme(axis.title.y = element_text(size = 15), legend.text = element_text(size = 12), legend.title = element_text(size = 14)) 


# save the proportion chart with colours matching the UMAPs 



# colour order to match cell type order
clust.colours <- c("chocolate1","deepskyblue","steelblue4","mediumpurple3","red2","burlywood3", 
                   "pink2")



pdf(paste(output_path,"BarChartProportionCellTypesin4pops.pdf"), height = 5, width = 7)
ggplot(df, aes(x = Var1, y = Freq, fill = Var2)) + 
  geom_col(position = "fill") + theme_classic() +
  scale_fill_manual(values = clust.colours) +
  #scale_x_discrete(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0,0)) +
  theme(axis.text.x = element_text(angle = 90, size = 16), axis.text.y = element_text(size = 16)) +
  labs(y = "Proportion of Cells", x = "") +
  theme(axis.title.y = element_text(size = 16), legend.text = element_text(size = 16), legend.title = element_text(size = 16)) 
dev.off()

# change the colours

# try with the new label based on the merge object clusters

df.2 <- as.data.frame(table(seu.q$orig.ident, seu.q$Cluster.ids))

ggplot(df.2, aes(x = Var1, y = Freq, fill = Var2)) + 
  geom_col(position = "fill") + 
  #scale_fill_manual(values = c("#0073C2FF", "#EFC000FF")) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, size = 12), axis.text.y = element_text(size = 12)) +
  labs(y = "Proportion of Cells", x = "") + theme_classic() +
  theme(axis.title.y = element_text(size = 15), legend.text = element_text(size = 12), legend.title = element_text(size = 14)) 

# main subgroups renamed
df.3 <- as.data.frame(table(seu.q$orig.ident, seu.q$Level1))
ggplot(df.3, aes(x = Var1, y = Freq, fill = Var2)) + 
  geom_col(position = "fill") + 
  #scale_fill_manual(values = c("#0073C2FF", "#EFC000FF")) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, size = 12), axis.text.y = element_text(size = 12)) +
  labs(y = "Proportion of Cells", x = "") + theme_classic() +
  theme(axis.title.y = element_text(size = 15), legend.text = element_text(size = 12), legend.title = element_text(size = 14)) 








```


Make a proprotion table


```{r}

library(stringr)
library(reshape)

# for original cell types 


df.r <- reshape(df, idvar = "Var2", timevar = "Var1", direction = "wide")
dat1 <- df.r
dat1[] <- lapply(dat1[], function(x){
  # Check if the column is numeric
  if (is.numeric(x)){
    return(x/sum(x)*100)
  } else{
    return(x)
  }
})
dat1
write.csv(dat1, paste(output_path,"proportionofCelltypesin4pops.csv"))

### for new main cell types 

df.r <- reshape(df.3, idvar = "Var2", timevar = "Var1", direction = "wide")

dat3 <- df.r
dat3[] <- lapply(dat3[], function(x){
  # Check if the column is numeric
  if (is.numeric(x)){
    return(x/sum(x)*100)
  } else{
    return(x)
  }
})
dat3


### with the subtypes of cells
df.r <- reshape(df.2, idvar = "Var2", timevar = "Var1", direction = "wide")

dat2 <- df.r
dat2[] <- lapply(dat2[], function(x){
  # Check if the column is numeric
  if (is.numeric(x)){
    return(x/sum(x)*100)
  } else{
    return(x)
  }
})
dat2





```


Process Kamath 

```{r}


seu <- NormalizeData(seu.r, normalization.method = "LogNormalize", scale.factor = 10000)
seu <- FindVariableFeatures(seu, selection.method = "vst", nfeatures = 2000)
seu <- ScaleData(seu)
seu <- RunPCA(seu)

saveRDS(seu,"/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/KamathTotal_downsample.RDS")



```








Re annotate the new clustered object

Get predictions:
Use Kamath all cells down sampled (doesn't contain Radial Glia)
Developing Forebrain
Organoids AIW002 120 days

```{r}

# Kamath midbrain Postmortem sn

seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/KamathTotal_downsample.RDS")
#Idents(seu.r) <- "Cell_Type"
#seu.r <- subset(seu.r, idents = c("astro","da","nonda","endo"))

Idents(seu.r) <- "Cell_Subtype"

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = seu.q, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$Cell_Subtype, k.weight = 50)
seu.q <- AddMetaData(seu.q, predictions$predicted.id, col.name = "mb.pred")
seu.q <- AddMetaData(seu.q, predictions$prediction.score.max, col.name = "prediction.score.max")

table(seu.q$mb.pred)


seu.q$mb.pred.thresh <- ifelse(seu.q$prediction.score.max > 0.7, seu.q$mb.pred, "none")

DimPlot(seu.q, group.by = 'mb.pred.thresh')
table(seu.q$mb.pred.thresh)


#***** subtype predictions don't work 
#but when I had the separate subytes they did work on neurons and astro separately. It could be there just aren't enough reference cells - I only have 1000 each main group.  

# I'm going to try with the main cell types groups instead

Idents(seu.r) <- "Cell_Type"
table(seu.r$Cell_Type)

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = seu.q, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$Cell_Type)
seu.q <- AddMetaData(seu.q, predictions$predicted.id, col.name = "mb.pred")
seu.q <- AddMetaData(seu.q, predictions$prediction.score.max, col.name = "prediction.score.max")

table(seu.q$mb.pred)


seu.q$mb.pred.thresh <- ifelse(seu.q$prediction.score.max > 0.7, seu.q$mb.pred, "none")


DimPlot(seu.q, group.by = 'mb.pred.thresh', label = TRUE)
table(seu.q$mb.pred.thresh)

# create the prediction top 2 cell types per cluster dataframe

# still useless

# best to use the subtype predictions from the separate objects after subsetting


```


Try the Bhaduri whole brain dataset

```{r}

# midbrain and striatum

seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/Bhaduri_midbrain_striatum.RDS")

Idents(seu.r) <- "cell_cluster"

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = seu.q, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$cell_cluster)

seu.q <- AddMetaData(seu.q, predictions$predicted.id, col.name = "mbs.pred")
seu.q <- AddMetaData(seu.q, predictions$prediction.score.max, col.name = "prediction.score.max")

table(seu.q$mbs.pred)
seu.q$mbs.pred.thresh <- ifelse(seu.q$prediction.score.max > 0.7, seu.q$mbs.pred, "none")


DimPlot(seu.q, group.by = 'mbs.pred.thresh', label = TRUE)
table(seu.q$mb.pred.thresh)


# read in the reference dataset
# whole brain Bhaduri down sampled
seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/Bhaduri_downsample.RDS")

Idents(seu.r) <- "cell_cluster"

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = seu.q, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$cell_cluster)
seu.q <- AddMetaData(seu.q, predictions$predicted.id, col.name = "brain.pred")
seu.q <- AddMetaData(seu.q, predictions$prediction.score.max, col.name = "prediction.score.max")

table(seu.q$brain.pred)
seu.q$brain.pred.thresh <- ifelse(seu.q$prediction.score.max > 0.7, seu.q$brain.pred, "none")


DimPlot(seu.q, group.by = 'brain.pred.thresh', label = TRUE)
DimPlot(seu.q, group.by = "brain.pred")
table(seu.q$brain.pred.thresh)

```


Label transfer with the developing forbrain

```{r}

seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/Karolinski_DevForebrain_downsample_Level1.RDS")
colnames(seu.r@meta.data)
DefaultAssay(seu.r) <- 'RNA'

# clusters has subgroups
# levels are main cell groups 
Idents(seu.r) <- "Level1"

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = seu.q, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$Level1)
seu.q <- AddMetaData(seu.q, predictions$predicted.id, col.name = "dfb.pred")
seu.q <- AddMetaData(seu.q, predictions$prediction.score.max, col.name = "prediction.score.max")

table(seu.q$dfb.pred)
seu.q$dfb.pred.thresh <- ifelse(seu.q$prediction.score.max > 0.85, seu.q$dfb.pred, "none")

table(seu.q$dfb.pred.thresh)
DimPlot(seu.q, group.by = 'dfb.pred.thresh', label = FALSE)
DimPlot(seu.q, group.by = "dfb.pred")
DimPlot(seu.q, group.by = 'orig.ident', split.by = 'dfb.pred.thresh')
DimPlot(seu.q, split.by = 'orig.ident', group.by = 'dfb.pred.thresh')


```



AIW002 120 days predictions

```{r}

seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/AIWtrio120days/MOintegratedClusterK123res0.8.names_nov16_2021")

### maybe the knockouts are causing a problem
Idents(seu.r) <- 'orig.ident'
seu.r <- subset(seu.r, idents = )

Idents(seu.r) <- "res08names"
DefaultAssay(seu.r) <- "RNA"

anchors <- FindTransferAnchors(reference = seu.r ,query = seu.q, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$res08names)

seu.q <- AddMetaData(seu.q, predictions$predicted.id, col.name = "hmo.pred")
seu.q <- AddMetaData(seu.q, predictions$prediction.score.max, col.name = "prediction.score.max")

table(seu.q$hmo.pred)
seu.q$hmo.pred.thresh <- ifelse(seu.q$prediction.score.max > 0.85, seu.q$hmo.pred, "none")

table(seu.q$hmo.pred.thresh)
DimPlot(seu.q, group.by = 'hmo.pred.thresh', label = FALSE)
DimPlot(seu.q, group.by = "hmo.pred")
DimPlot(seu.q, group.by = 'orig.ident', split.by = 'hmo.pred.thresh')
DimPlot(seu.q, split.by = 'orig.ident', group.by = 'hmo.pred.thresh')

```

Make a summary table of predictions

```{r}


# AIW002 120 days predictions - take the thresholded options
t.lables <- as.data.frame(table(seu.q$RNA_snn_res.1.8, seu.q$hmo.pred.thresh))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 
top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)




```


Try the 60 days organoids AIW

```{r}

seu.r <-readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/AIWtrio60days/AWI002ParkinKOPinkKO60days_labels_14052022.rds")

Idents(seu.r) <- "cluster.ids"
DefaultAssay(seu.r) <- "RNA"

anchors <- FindTransferAnchors(reference = seu.r ,query = seu.q, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$cluster.ids)

seu.q <- AddMetaData(seu.q, predictions$predicted.id, col.name = "hmo60.pred")
seu.q <- AddMetaData(seu.q, predictions$prediction.score.max, col.name = "prediction.score.max")

table(seu.q$hmo60.pred)
seu.q$hmo60.pred.thresh <- ifelse(seu.q$prediction.score.max > 0.85, seu.q$hmo60.pred, "none")

table(seu.q$hmo60.pred.thresh)
DimPlot(seu.q, group.by = 'hmo60.pred.thresh', label = FALSE)
DimPlot(seu.q, group.by = "hmo60.pred")
DimPlot(seu.q, group.by = 'orig.ident', split.by = 'hmo60.pred.thresh')
DimPlot(seu.q, split.by = 'orig.ident', group.by = 'hmo60.pred.thresh')



```


Make predictions from 60 days

```{r}
# AIW002 120 days predictions - take the thresholded options
t.lables <- as.data.frame(table(seu.q$RNA_snn_res.1.8, seu.q$hmo60.pred.thresh))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 
top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)


```


Look at the expression of gene lists

```{r}

Idents(seu.q) <- 'RNA_snn_res.1.8'

feature_list = c("MKI67","SOX2","POU5F1","DLX2","PAX6","SOX9","HES1","NES","RBFOX3","MAP2","NCAM1","CD24","GRIA2","GRIN2B","GABBR1","GAD1","GAD2","GABRA1","GABRB2","TH","ALDH1A1","LMX1B","NR4A2","CORIN","CALB1","KCNJ6","CXCR4","ITGA6","SLC1A3","CD44","AQP4","S100B", "PDGFRA","OLIG2","MBP","CLDN11","VIM","VCAM1")

#DoHeatmap(seu.q, features = feature_list, size=3, angle =90, group.bar.height = 0.02)
DotPlot(seu.q, features = feature_list) +RotatedAxis()

PD_poulin = c("TH","SLC6A3","SLC18A2","SOX6","NDNF","SNCG","ALDH1A1","CALB1","TACR2","SLC17A6","SLC32A1","OTX2","GRP","LPL","CCK","VIP")

#DoHeatmap(seu.q, features = PD_poulin, size=3, angle =90, group.bar.height = 0.02, group.by = 'RNA_snn_res.0.6')
DotPlot(seu.q, features = PD_poulin)+RotatedAxis()

ealryNeur = c("DCX","NEUROD1","TBR1")
proliferation = c("PCNA","MKI67")
neuralstem = c("SOX2","NES","PAX6","MASH1")

feature_list <- c("DCX","NEUROD1","TBR1","PCNA","MKI67","SOX2","NES","PAX6","MASH1")
#DoHeatmap(seu.q, features = feature_list, size=3, angle =90, group.bar.height = 0.02, group.by = 'RNA_snn_res.0.6')
DotPlot(seu.q, features = feature_list)+RotatedAxis()
# no proliferation marker expression  PCNA or MKI67
# cluster 4 DA neurons - shows early neuron marker and low PAX 4
# cluster 3 has higher SOX2 - neuroblast marker / NPC marker

mat_neuron = c("RBFOX3","SYP","DLG45","VAMP1","VAMP2","TUBB3","SYT1","BSN","HOMER1","SLC17A6") 
# NeuN is FOX3 - RBFOX3
# PSD95 also SP-90 or DLG4
# VGLUT2 is SLC17A6
#DoHeatmap(seu.q, features = mat_neuron, size=3, angle =90, group.bar.height = 0.02, group.by = 'RNA_snn_res.0.6')
# cluster 4 also show mature neuron markers
DotPlot(seu.q, features = mat_neuron)+RotatedAxis()
# excitatory neuron markers
ex = c("GRIA2","GRIA1","GRIA4","GRIN1","GRIN2B","GRIN2A","GRIN3A","GRIN3","GRIP1","CAMK2A")
#DoHeatmap(seu.q, features = ex, size=3, angle =90, group.bar.height = 0.02, group.by = 'RNA_snn_res.0.6')
DotPlot(seu.q, features = ex)+RotatedAxis()
# inhibitory neuron markers
inh = c("GAD1","GAD2", "GAT1","PVALB","GABR2","GABR1","GBRR1","GABRB2","GABRB1","GABRB3","GABRA6","GABRA1","GABRA4","TRAK2")
#DoHeatmap(seu.q, features = inh, size=3, angle =90, group.bar.height = 0.02, group.by = 'RNA_snn_res.0.6')
DotPlot(seu.q, features = inh)+RotatedAxis()
# cluster 4 is more excitatory than inhbitory but neither marker set has much expression 

### glia markers
microglia = c("PTPRC","AIF1","ADGRE1")  # ADGRE1 is a microglia marker F4/80, CD45 is PTPRC, gene name IBA1 is AIF1
astolgNPCpromicro = c("GFAP","S100B","SLC1A2","MBP","SOX10","SPP1","DCX","NEUROD1","TBR1","PCNA","MKI67","PTPRC","AIF1","ADGRE1")
# note GLT1 is EAAT2 which is SLC1A2 glutatmate transporter
# epithelial
epi = c("HES1","HES5","SOX2","SOX10","NES","CDH1","NOTCH1") # e-cadherin is CDH1

#DoHeatmap(seu.q, features = astolgNPCpromicro, size=3, angle =90, group.bar.height = 0.02, group.by = 'RNA_snn_res.0.6')
DotPlot(seu.q, features = astolgNPCpromicro)+RotatedAxis()
# cluster 4 is more excitatory than inhbitory but neither marker set has much expression 
#DoHeatmap(seu.q, features = epi, size=3, angle =90, group.bar.height = 0.02, group.by = 'RNA_snn_res.0.6')
DotPlot(seu.q, features = epi)+RotatedAxis()

# also add Radial glia marker overlap with Glia and Neurons

features <- c("PTPRC","AIF1","ADGRE1", "VIM", "TNC","PTPRZ1","FAM107A","HOPX","LIFR",
              "ITGB5","IL6ST")
#DoHeatmap(seu.q, features = features, size=3, angle =90, group.bar.height = 0.02, group.by = 'RNA_snn_res.0.6')
DotPlot(seu.q, features = features)+RotatedAxis()
```



Rename cell types 

```{r}

Idents(seu.q) <- 'RNA_snn_res.1.8'

cluster.ids <- c("Astro-RG-Pre","Neurons1-im","RG1-NPC","Astro1","Neurons2-im",
                 "Astro-NPC","Astro","Neurons3","RG2","Neurons4",
                 "Astro3","Neurons5","RG3-NPC","Neurons6-DA","Neurons7-NPC",
                 "Mix-RG4","RG5","Endothelial-NPC","Neurons8-inh-max","Neurons9-DA",
                 "Neurons10-DA","RG6","RG7-epi","NPC-Astro","RG-OPC"
                 )


names(cluster.ids) <- levels(seu.q)
seu.q <- RenameIdents(seu.q, cluster.ids)
seu.q$Level2 <- Idents(seu.q)

DimPlot(seu.q, group.by = 'Level2', label = TRUE)
DimPlot(seu.q, group.by = 'Level2', label = TRUE, split.by = 'orig.ident')
DimPlot(seu.q, group.by =  'orig.ident')

# name the major groups together and see how it matches up with space

cluster.ids <- c("Astro1","Neurons1","RG1","Astro2","Neurons2",
                 "Astro3","Astro4","Neurons3","RG2","Neurons4",
                 "Astro5","Neurons5","RG3-NPC","DANeurons1","Neurons6",
                 "Mix","RG4","Endothelial-NPC","Neurons8-inh-max","DANeurons2",
                 "DANeurons3","RG5","RG6-epi","RG7","RG-OPC"
                 )


names(cluster.ids) <- levels(seu.q)
seu.q <- RenameIdents(seu.q, cluster.ids)
seu.q$Level3 <- Idents(seu.q)

DimPlot(seu.q, group.by = 'Level3', label = TRUE, repel = TRUE)
DimPlot(seu.q, group.by = 'Level3', label = TRUE, split.by = 'orig.ident')
DimPlot(seu.q, group.by =  'orig.ident')

Idents(seu.q) <- 'RNA_snn_res.1.8'

cluster.ids <- c("Astro","Neurons","RG","Astro","Neurons",
                 "Astro","Astro","Neurons","RG","Neurons",
                 "Astro","Neurons","RG","DANeurons","Neurons",
                 "Mix","RG","Endothelial","Neurons","DANeurons",
                 "DANeurons","RG","RG","RG","RG"
                 )


names(cluster.ids) <- levels(seu.q)
seu.q <- RenameIdents(seu.q, cluster.ids)
seu.q$Levels <- Idents(seu.q)

DimPlot(seu.q, group.by = 'Levels', label = TRUE, repel = TRUE)
DimPlot(seu.q, group.by = 'Levels', label = TRUE, split.by = 'orig.ident')
DimPlot(seu.q, group.by =  'orig.ident')




```


See how the labels fit with sorted cell types

```{r}

t.lables <- as.data.frame(table(seu.q$orig.ident, seu.q$Level3))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)

#Not a bad breakdown mix
t.lables <- as.data.frame(table(seu.q$orig.ident, seu.q$Levels))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)


```


Make the cell type proportion plots - make the input colours should be different than cell type colours
Cell type 
Use colours from other figures if possible:
Astrocytes = orange
Orange
Radial Glia = Pink
Neurons = Purple
Endothelial bright blue


I'll make a UMAP with the input colours  - match to 2D - 

```{r}

# Figure 6 C
# UMAP with the 4 orig.idents

sample.order <- c("Astrocytes","RadialGlia","Neurons1","Neurons2")
sample.order <- rev(sample.order)

# colour order to match cell type order
clust.colours <- c("royalblue", "indianred2","springgreen4","palegreen2")
 
Idents(seu.q) <- 'orig.ident'
      
DimPlot(seu.q, order = sample.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, pt.size = 0.1, label = FALSE)

output_path <- "/Users/rhalenathomas/Documents/Projects_Papers/PhenoID/ForFigures/scRNA/"

### Figure 6C

pdf(paste(output_path,"UMAPscRNAseqMerge4.11102022.pdf"),width = 7, height = 4)
DimPlot(seu.q, order = sample.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, pt.size = 0.1, label = FALSE, label.size = 6) +
  theme(legend.text = element_text(size=16), axis.title.y = element_text(size=16), 
        axis.title.x = element_text(size=16), axis.text.y = element_text(size =16),
        axis.text.x = element_text(size =16))
dev.off()


## Figure S23 panel 

cell.order <- c("Astrocytes","Endothelial","Glia","NPC","Neurons","Other","Radial Glia")
cell.order <- rev(cell.order)

# colour order to match cell type order
clust.colours <- c("chocolate1","deepskyblue","steelblue4","red2","mediumpurple3","burlywood3", 
                   "pink2")

 
Idents(seu.q) <- 'Cell_Types'
      
# designated the order of the splits factor
seu.q$orig.ident <- factor(x = seu.q$orig.ident, levels = c("Astrocytes","RadialGlia","Neurons1","Neurons2"))


DimPlot(seu.q, order = cell.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, pt.size = 0.1, label = FALSE, split.by = 'orig.ident', ncol = 2)




pdf(paste(output_path,"UMAP_merge_splitbyorigident.11102022.pdf"),width = 12, height = 7.5)
DimPlot(seu.q, order = cell.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, pt.size = 0.1, label = FALSE, split.by = 'orig.ident', label.size = 6, ncol = 2) +
  theme(legend.text = element_text(size=16), axis.title.y = element_text(size=16), 
        axis.title.x = element_text(size=16), axis.text.y = element_text(size =16),
        axis.text.x = element_text(size =16))
dev.off()


# also plot one not split to see the whole thing


pdf(paste(output_path,"UMAP_merge_CellTypes.11102022.pdf"),width = 6.7, height = 4.1)
DimPlot(seu.q, order = cell.order, cols = clust.colours, shuffle = TRUE, raster=FALSE, pt.size = 0.1, label = FALSE, label.size = 6) +
  theme(legend.text = element_text(size=16), axis.title.y = element_text(size=16), 
        axis.title.x = element_text(size=16), axis.text.y = element_text(size =16),
        axis.text.x = element_text(size =16))
dev.off()




```


Find Cluster markers with all clusters:

```{r}

Idents(seu.q) <- 'Level3'

ClusterMarkers <- FindAllMarkers(seu.q)

top3 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=3, wt = avg_log2FC)
DoHeatmap(seu.q, features = top3$gene, size=3, angle =90, group.bar.height = 0.02)

write.csv(output_path,"ClusterMarkersLevel3all.csv")


```

Same cell type markers are overlapping


```{r}

# error in dot plot because of repeat gene
marker.top3 <- unique(top3$gene)
DotPlot(seu.q, features = marker.top3) + RotatedAxis()

# RG-NPC is similar to mix and DAneurons2

```

Not all astros overlap with astro markers and not all neurons overlap neurons
The RG overlap with all kinds of cells. 


Subset out populations for some label transfers from the public data.

Subset out Neurons 

```{r}

# subset neurons
# Level2
cluster.ids <- c("Astro-RG-Pre","Neurons1-im","RG1-NPC","Astro1","Neurons2-im",
                 "Astro-NPC","Astro","Neurons3","RG2","Neurons4",
                 "Astro3","Neurons5","RG3-NPC","Neurons6-DA","Neurons7-NPC",
                 "Mix-RG4","RG5","Endothelial-NPC","Neurons8-inh-max","Neurons9-DA",
                 "Neurons10-DA","RG6","RG7-epi","NPC-Astro","RG-OPC"
                 )
# Level 3
cluster.ids <- c("Astro1","Neurons1","RG1","Astro2","Neurons2",
                 "Astro3","Astro4","Neurons3","RG2","Neurons4",
                 "Astro5","Neurons5","RG3-NPC","DANeurons1","Neurons6",
                 "Mix","RG4","Endothelial-NPC","Neurons8-inh-max","DANeurons2",
                 "DANeurons3","RG5","RG6-epi","RG7","RG-OPC"
                 )

Idents(seu.q) <- 'Level3'
neuron.id <- c("Neurons1","Neurons2","Neurons3","Neurons4",
                 "Neurons5","DANeurons1","Neurons6",
                 "Mix","Neurons8-inh-max","DANeurons2",
                 "DANeurons3")
# without mix
neuron.id <- c("Neurons1","Neurons2","Neurons3","Neurons4",
                 "Neurons5","DANeurons1","Neurons6",
                 "Neurons8-inh-max","DANeurons2",
                 "DANeurons3")

neurons.sub <- subset(seu.q, idents = neuron.id)

table(neurons.sub$Level3)

DimPlot(neurons.sub, label = TRUE)



```


Find Cluster markers to distinguish Neuron subgroups

```{r}

table(neurons.sub$Level3)
Idents(neurons.sub) <- 'Level3'

ClusterMarkers <- FindAllMarkers(neurons.sub, only.pos = TRUE)

top3 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=3, wt = avg_log2FC)
DoHeatmap(neurons.sub, features = top3$gene, size=3, angle =90, group.bar.height = 0.02)

marker.top3 <- unique(top3$gene)
DotPlot(neurons.sub, features = marker.top3) + RotatedAxis()

write.csv(ClusterMarkers, paste(output_path,"ClusterMarkersLevel3NeuronSubnoMix.csv", sep = ""))


```
Neurons 1 has overlapping markers with DAneurons2 
Neurons 2, Neurons 4, Neurons 6 are very similar in the top markers - these are all in the Neurons1 FACS pop
And the top markers are mito genes.


Look at the DA marker genes and see if there are differences there
Also some neuronal markers

```{r}

PD_poulin = c("TH","SLC6A3","SLC18A2","SOX6","NDNF","SNCG","ALDH1A1","CALB1","TACR2","SLC17A6","SLC32A1","OTX2","GRP","LPL","CCK","VIP")

DoHeatmap(neurons.sub, features = PD_poulin)
DotPlot(neurons.sub, features = PD_poulin) + RotatedAxis()


DotPlot(neurons.sub, features = "TH")
DA.sub <- c("CALB1","SOX6","RBP4","DDT")
DotPlot(neurons.sub, features = DA.sub) + RotatedAxis()


```

DA neurons 1 has the most different DA markers (TH, CALB1, SLC17A6, OTX2) - highest CALB1
DA neurons 2 has most cells and highest level of TH expression, most cells expressing SOX2 at a low levels
DA neurons 3 SLC17A6, OTX2, OTX2 highest 


Get markers for just DA neurons

```{r}

Idents(neurons.sub) <- 'Level3'
DA.sub <- subset(neurons.sub, idents = c("DANeurons1","DANeurons2","DANeurons3"))

ClusterMarkers.DA <- FindAllMarkers(DA.sub, only.pos = TRUE)

top3 <- ClusterMarkers.DA %>% group_by(cluster) %>% top_n(n=3, wt = avg_log2FC)
DoHeatmap(DA.sub, features = top3$gene, size=3, angle =90, group.bar.height = 0.02)

marker.top3 <- unique(top3$gene)
DotPlot(neurons.sub, features = marker.top3) + RotatedAxis()
DotPlot(DA.sub, features = marker.top3) + RotatedAxis()

write.csv(ClusterMarkers, paste(output_path,"ClusterMarkersDANeurons.csv", sep = ""))


```
Good markers DA1 - RAB3B
DA2 - TPBG or HES1  - TPBG is proposed as a DA marker  (HES1 midbrain patterning)
DA3 - TPH1 - sertononergic neurons or TTR oxidative stress protective


Compare the other neurons and find markers

```{r}

Idents(neurons.sub) <- 'Level3'
neur.sub <- subset(neurons.sub, idents = c("DANeurons1","DANeurons2","DANeurons3"), invert = TRUE)
table(neur.sub$Level3)
ClusterMarkers.neur <- FindAllMarkers(neur.sub, only.pos = TRUE)

top3 <- ClusterMarkers.neur %>% group_by(cluster) %>% top_n(n=3, wt = avg_log2FC)
DoHeatmap(neur.sub, features = top3$gene, size=3, angle =90, group.bar.height = 0.02)

top3 <- ClusterMarkers.neur %>% group_by(cluster) %>% top_n(n=3, wt = avg_log2FC)
DoHeatmap(neur.sub, features = top3$gene, size=3, angle =90, group.bar.height = 0.02)

marker.top3 <- unique(top3$gene)
#DotPlot(neurons.sub, features = marker.top3) + RotatedAxis()
DotPlot(neur.sub, features = marker.top3) + RotatedAxis()

write.csv(ClusterMarkers.neur, paste(output_path,"ClusterMarkersotherNeurons.csv", sep = ""))


```


Have a look at some markers in feature plots

```{r}

DimPlot(neur.sub)
FeaturePlot(neur.sub, features = c("RUNX1T1", "ASCL1","GAD1","GRIN2A","GRIA2","VAMP2"))


mat_neuron = c("RBFOX3","SYP","DLG45","VAMP1","VAMP2","TUBB3","SYT1","BSN","HOMER1","SLC17A6") 
# NeuN is FOX3 - RBFOX3
# PSD95 also SP-90 or DLG4
# VGLUT2 is SLC17A6
#DoHeatmap(seu.q, features = mat_neuron, size=3, angle =90, group.bar.height = 0.02, group.by = 'RNA_snn_res.0.6')
# cluster 4 also show mature neuron markers
DotPlot(neur.sub, features = mat_neuron)+RotatedAxis()
# excitatory neuron markers
ex = c("GRIA2","GRIA1","GRIA4","GRIN1","GRIN2B","GRIN2A","GRIN3A","GRIN3","GRIP1","CAMK2A")
#DoHeatmap(seu.q, features = ex, size=3, angle =90, group.bar.height = 0.02, group.by = 'RNA_snn_res.0.6')
DotPlot(neur.sub, features = ex)+RotatedAxis()
# inhibitory neuron markers
inh = c("GAD1","GAD2", "GAT1","PVALB","GABR2","GABR1","GBRR1","GABRB2","GABRB1","GABRB3","GABRA6","GABRA1","GABRA4","TRAK2")

DotPlot(neur.sub, features = inh) + RotatedAxis()






```


I'll separate out Neurons 2,4,6

```{r}

Idents(neurons.sub) <- 'Level3'
neurons1.sub <- subset(neurons.sub, idents = c("Neurons2","Neurons4","Neurons6"))

ClusterMarkers.n1 <- FindAllMarkers(neurons1.sub, only.pos = TRUE)

top3 <- ClusterMarkers.n1 %>% group_by(cluster) %>% top_n(n=3, wt = avg_log2FC)
DoHeatmap(neurons1.sub, features = top3$gene, size=3, angle =90, group.bar.height = 0.02)

marker.top3 <- unique(top3$gene)
DotPlot(neurons1.sub, features = marker.top3) + RotatedAxis()
DotPlot(DA.sub, features = marker.top3) + RotatedAxis()

write.csv(ClusterMarkers.n1, paste(output_path,"ClusterMarkersNeurons246.csv", sep = ""))

```

```{r}

Idents(neur.sub) <- 'Level3'
neurons2.sub <- subset(neur.sub, idents = c("Neurons2","Neurons4","Neurons6"), invert = TRUE)

ClusterMarkers.n2 <- FindAllMarkers(neurons2.sub, only.pos = TRUE)

top3 <- ClusterMarkers.n2 %>% group_by(cluster) %>% top_n(n=3, wt = avg_log2FC)
DoHeatmap(neurons2.sub, features = top3$gene, size=3, angle =90, group.bar.height = 0.02)

marker.top3 <- unique(top3$gene)
DotPlot(neurons2.sub, features = marker.top3) + RotatedAxis()


write.csv(ClusterMarkers.n2, paste(output_path,"ClusterMarkersNeurons1358.csv", sep = ""))


```

Look at the GO terms when separating out the neuron subtypes

```{r}
library(enrichR)

setEnrichrSite("Enrichr") # Human genes
# list of all the databases

# libaries with cell types
#dbs <- listEnrichrDbs()
#dbs
db <- c('GO_Cellular_Component_2015','GO_Biological_Process_2015',
        'CellMarker_Augmented_2021','Azimuth_Cell_Types_2021')



group = "Neurons3"

#for(group in group.list){
print(group)
Up.list <- ClusterMarkers.n2 %>% filter(cluster == group & avg_log2FC > 0)
genes <- Up.list$gene

Er <- enrichr(genes, databases = db)
print(plotEnrich(Er[[1]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[3]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[4]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))

t.GOcell <- Er[[1]] %>% select(Term, Genes, Combined.Score)
print(t.GOcell)

t.GObio <- Er[[2]] %>% select(Term, Genes, Combined.Score)
print(t.GObio)

t.CellMarker <- Er[[3]] %>% select(Term, Genes, Combined.Score)
print(t.CellMarker)


t.Azi <- Er[[4]] %>% select(Term, Genes, Combined.Score)
print(t.Azi)
#}



```



Compare the neurons1 groups to the neurons2 groups

```{r}

Idents(neurons.sub) <- 'Level3'
ClusterMarkers <- FindMarkers(neurons.sub, ident.1 = c("Neurons1","Neurons3","Neurons5","Neurons8-inh-max"), ident.2 = c("Neurons2","Neurons4","Neurons6"))

up.n1 <- ClusterMarkers %>% top_n(n=5, wt = avg_log2FC)
up.n2 <- ClusterMarkers %>% top_n(n=-5, wt = avg_log2FC)


DoHeatmap(neurons.sub, features = c(rownames(up.n1), rownames(up.n2)), size=3, angle =90, group.bar.height = 0.02)

DotPlot(neurons.sub, features = c(rownames(up.n1), rownames(up.n2))) + RotatedAxis()


write.csv(ClusterMarkers, paste(output_path,"ClusterMarkersNeurons1vs2.csv", sep = ""))

DotPlot(neur.sub, features = c("CD24", "NCAM1")) + RotatedAxis()

### neurons 1 and 2 must be mixed up in the sorting
# Needs to be neurons CD24 and neurons CD56 
# where neurons2 is the CD24 population


```

Check the GO terms and celltype libraries for the two groups of neurons

```{r}
library(enrichR)

setEnrichrSite("Enrichr") # Human genes
# list of all the databases

# libaries with cell types
#dbs <- listEnrichrDbs()
#dbs
db <- c('GO_Cellular_Component_2015','GO_Biological_Process_2015',
        'CellMarker_Augmented_2021','Azimuth_Cell_Types_2021')


# up regulated in Neurons 2 (CD24 high)

Up.list <- ClusterMarkers %>% filter(avg_log2FC > 0)
genes <- rownames(Up.list)


# Neurons 1 (CD24 low)
down.list <- ClusterMarkers %>% filter(avg_log2FC < 0)
genes <- rownames(down.list)

Er <- enrichr(genes, databases = db)
print(plotEnrich(Er[[1]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[3]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[4]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))

t.GOcell <- Er[[1]] %>% select(Term, Genes, Combined.Score)
print(t.GOcell)
t.GObio <- Er[[2]] %>% select(Term, Genes, Combined.Score)
print(t.GObio)
t.CellMarker <- Er[[3]] %>% select(Term, Genes, Combined.Score)
print(t.CellMarker)
t.Azi <- Er[[4]] %>% select(Term, Genes, Combined.Score)
print(t.Azi)


```

Compare directly Neurons 1 (low CD24) with Neurons 2 (high CD24)
For DEG and cell type predictions

```{r}

Idents(neurons.sub) <- 'orig.ident'
neurons.1.2 <- subset(neurons.sub, idents = c("Neurons1", "Neurons2")) 
table(neurons.1.2$Level3,neurons.1.2$orig.ident)

# rename Idents to make 3 groups and then do DGE and predictions on those groups
# note the prediction are across all the cell types

# reprocess 
seu <- neurons.1.2
seu <- NormalizeData(seu, normalization.method = "LogNormalize", scale.factor = 10000)
seu <- FindVariableFeatures(seu, selection.method = "vst", nfeatures = 2000)
seu <- ScaleData(seu)
seu.q <- RunPCA(seu)

DimPlot(seu.q)

```
```{r}

Idents(seu.q) <- 'Level3'

cluster.ids <- c("Neurons-CD24","Neurons","Neurons-CD24","Neurons",
                 "Neurons-CD24","DAN",
                 "Neurons","Neurons-CD24","DAN","DAN"
                 )

names(cluster.ids) <- levels(seu.q)
seu.q <- RenameIdents(seu.q, cluster.ids)
seu.q$NeuronGroups <- Idents(seu.q)
DimPlot(seu.q, group.by = 'Level3', split.by = 'orig.ident')
DimPlot(seu.q, group.by = 'NeuronGroups', split.by = 'orig.ident')

```

See the predicted markers

```{r}

DimPlot(seu.q, group.by = 'MBOAST23.pred', split.by = 'orig.ident')
DimPlot(seu.q, group.by = 'MBOAIW.pred', split.by = 'orig.ident')
DimPlot(seu.q, group.by = 'AIW60.pred', split.by = 'orig.ident')
DimPlot(seu.q, group.by = 'Bha.mid.stri.pred', split.by = 'orig.ident')
DimPlot(seu.q, group.by = 'da.pred', split.by = 'orig.ident')
DimPlot(seu.q, group.by = 'K.pred', split.by = 'orig.ident')


```


Double check and rename the two neuron populations

```{r}

FeaturePlot(neurons.sub, features = c("CD24","NCAM1", "MAP2", "TUBB3"), split.by = "orig.ident")

```



Look at cell marker libraries and GO terms in all neurons contrasts


```{r}

library(enrichR)

setEnrichrSite("Enrichr") # Human genes
# list of all the databases

# libaries with cell types
dbs <- listEnrichrDbs()
dbs
db <- c('GO_Cellular_Component_2015','GO_Biological_Process_2015',
        'CellMarker_Augmented_2021','Azimuth_Cell_Types_2021')

# enrichr(genes, databases = NULL)

group.list <- as.character(unique(ClusterMarkers$cluster))

# try to run a loop 
# the loop runs but it's hard to tell the output 
# need to run one at a time
[1] "Neurons1"         "Neurons2"         "Neurons3"         "Neurons4"         "Neurons5"        
 [6] "DANeurons1"       "Neurons6"         "Mix"              "Neurons8-inh-max" "DANeurons2"      
[11] "DANeurons3" 


group.list = "DANeurons3"

for(group in group.list){
print(group)
Up.list <- ClusterMarkers %>% filter(cluster == group & avg_log2FC > 0)
genes <- Up.list$gene

Er <- enrichr(genes, databases = db)
print(plotEnrich(Er[[1]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[3]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[4]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))

t.GOcell <- Er[[1]] %>% select(Term, Genes, Combined.Score)
print(t.GOcell)

t.GObio <- Er[[2]] %>% select(Term, Genes, Combined.Score)
print(t.GObio)

t.CellMarker <- Er[[3]] %>% select(Term, Genes, Combined.Score)
print(t.CellMarker)


t.Azi <- Er[[4]] %>% select(Term, Genes, Combined.Score)
print(t.Azi)
}



```



Label transfer Kamath DA neurons in Neurons Subset (has both DA neurons and Others as predicted by AIW002 label transfers)

```{r}


seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/Macosko_Data/DAsubgroups_processed.Rds")

Idents(seu.r) <- "Cell_Subtype"
Idents(neurons.sub) <- "Level3"

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = neurons.sub, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$Cell_Subtype, k.weight = 30)
neurons.sub <- AddMetaData(neurons.sub, predictions$predicted.id, col.name = "da.pred")
neurons.sub <- AddMetaData(neurons.sub, predictions$prediction.score.max, col.name = "prediction.score.max")

table(neurons.sub$da.pred)

neurons.sub$da.pred.thresh <- ifelse(neurons.sub$prediction.score.max > 0.55, neurons.sub$da.pred, "none")

DimPlot(neurons.sub, group.by = 'da.pred.thresh')
table(neurons.sub$da.pred.thresh)






```


See the predictions from Kamath DA subtypes

```{r}

#DA subtypes

# all predicitions
t.lables <- as.data.frame(table(neurons.sub$Level3, neurons.sub$da.pred))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)

# with threshold
t.lables <- as.data.frame(table(neurons.sub$Level3, neurons.sub$da.pred.thresh))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top.thresh <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top.thresh) <- NULL
df.top.thresh$I <- row.names(df.top.thresh)


```

Label from Bhadari adult brain -all brain down sampled

```{r}

seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/Karolinski_DevForebrain_downsample_Level1.RDS")

Idents(seu.r) <- "Level1"
Idents(neurons.sub) <- "Level3"

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = neurons.sub, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$Level1, k.weight = 50)
neurons.sub <- AddMetaData(neurons.sub, predictions$predicted.id, col.name = "Bha.pred")
neurons.sub <- AddMetaData(neurons.sub, predictions$prediction.score.max, col.name = "prediction.score.max")

table(neurons.sub$Bha.pred)

neurons.sub$Bha.pred.thresh <- ifelse(neurons.sub$prediction.score.max > 0.80, neurons.sub$Bha.pred, "none")

DimPlot(neurons.sub, group.by = 'Bha.pred.thresh')
table(neurons.sub$Bha.pred.thresh)



```

```{r}

#Whole adult brain 

# all predicitions
t.lables <- as.data.frame(table(neurons.sub$Level3, neurons.sub$Bha.pred))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)

# with threshold
t.lables <- as.data.frame(table(neurons.sub$Level3, neurons.sub$Bha.pred.thresh))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top.thresh <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top.thresh) <- NULL
df.top.thresh$I <- row.names(df.top.thresh)
```

Try with only the midbrain and striatum Bhaduri

```{r}


seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/Bhaduri_midbrain_striatum.RDS")
Idents(seu.r) <- "cell_cluster"
Idents(neurons.sub) <- "Level3"

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = neurons.sub, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$cell_cluster, k.weight = 50)
neurons.sub <- AddMetaData(neurons.sub, predictions$predicted.id, col.name = "Bha.m.pred")
neurons.sub <- AddMetaData(neurons.sub, predictions$prediction.score.max, col.name = "prediction.score.max")

table(neurons.sub$Bha.m.pred)

neurons.sub$Bha.m.pred.thresh <- ifelse(neurons.sub$prediction.score.max > 0.70, neurons.sub$Bha.m.pred, "none")

DimPlot(neurons.sub, group.by = 'Bha.m.pred.thresh')
table(neurons.sub$Bha.m.pred.thresh)


# all predicitions
t.lables <- as.data.frame(table(neurons.sub$Level3, neurons.sub$Bha.m.pred))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)

# with threshold
t.lables <- as.data.frame(table(neurons.sub$Level3, neurons.sub$Bha.m.pred.thresh))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top.thresh <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top.thresh) <- NULL
df.top.thresh$I <- row.names(df.top.thresh)


```

Use the developing forbrain dataset Karolinski

```{r}

seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/Karolinski_DevForebrain_downsample_Level1.RDS")
Idents(seu.r) <- "Clusters"
Idents(neurons.sub) <- "Level3"

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = neurons.sub, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$Clusters, k.weight = 50)
neurons.sub <- AddMetaData(neurons.sub, predictions$predicted.id, col.name = "K.pred")
neurons.sub <- AddMetaData(neurons.sub, predictions$prediction.score.max, col.name = "prediction.score.max")

table(neurons.sub$K.pred)

neurons.sub$K.pred.thresh <- ifelse(neurons.sub$prediction.score.max > 0.70, neurons.sub$K.pred, "none")

DimPlot(neurons.sub, group.by = 'K.pred.thresh')
table(neurons.sub$K.pred.thresh)


# all predicitions
t.lables <- as.data.frame(table(neurons.sub$Level3, neurons.sub$K.pred))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)

# with threshold
t.lables <- as.data.frame(table(neurons.sub$Level3, neurons.sub$K.pred.thresh))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top.thresh <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top.thresh) <- NULL
df.top.thresh$I <- row.names(df.top.thresh)

```

Nowakowski developing cortex

```{r}
seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/Nowakowski_dev_cortext.RDS")
Idents(seu.r) <- "WGCNAcluster"
Idents(neurons.sub) <- "Level3"

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = neurons.sub, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$WGCNAcluster, k.weight = 50)
neurons.sub <- AddMetaData(neurons.sub, predictions$predicted.id, col.name = "N.pred")
neurons.sub <- AddMetaData(neurons.sub, predictions$prediction.score.max, col.name = "prediction.score.max")

table(neurons.sub$N.pred)

neurons.sub$N.pred.thresh <- ifelse(neurons.sub$prediction.score.max > 0.70, neurons.sub$N.pred, "none")

DimPlot(neurons.sub, group.by = 'N.pred.thresh')
table(neurons.sub$N.pred.thresh)


# all predicitions
t.lables <- as.data.frame(table(neurons.sub$Level3, neurons.sub$N.pred))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)

# with threshold
t.lables <- as.data.frame(table(neurons.sub$Level3, neurons.sub$N.pred.thresh))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top.thresh <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top.thresh) <- NULL
df.top.thresh$I <- row.names(df.top.thresh)

```

Check the DA neurons and None-neurons in the Kamath


```{r}

saveRDS(neurons.sub, paste(output_path,"neurons.sub.RDS",sep = ""))
saveRDS(astro.sub2, paste(output_path,"astro.sub.RDS",sep = ""))

neurons.sub <- readRDS(paste(output_path,"neurons.sub.RDS",sep = ""))

seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/Macosko_Data/Kamath_DA_NonDAsub.RDS")
Idents(seu.r) <- "Cell_Subtype"
Idents(neurons.sub) <- "Level3"

seu.r <- NormalizeData(seu.r, normalization.method = "LogNormalize", scale.factor = 10000)
seu.r <- FindVariableFeatures(seu.r, selection.method = "vst", nfeatures = 2000)
seu.r <- ScaleData(seu.r)
seu.r <- RunPCA(seu.r)

saveRDS(seu.r, "/Users/rhalenathomas/Documents/Data/scRNAseq/Macosko_Data/Kamath_DA_NonDAsub.RDS")

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = neurons.sub, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$Cell_Subtype, k.weight = 30)
neurons.sub <- AddMetaData(neurons.sub, predictions$predicted.id, col.name = "neur.pred")
neurons.sub <- AddMetaData(neurons.sub, predictions$prediction.score.max, col.name = "prediction.score.max")

table(neurons.sub$neur.pred)

neurons.sub$neur.pred.thresh <- ifelse(neurons.sub$prediction.score.max > 0.45, neurons.sub$neur.pred, "none")

DimPlot(neurons.sub, group.by = 'neur.pred.thresh')
table(neurons.sub$neur.pred.thresh)


# all predictions
t.lables <- as.data.frame(table(neurons.sub$Level3, neurons.sub$neur.pred))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)

# with threshold
t.lables <- as.data.frame(table(neurons.sub$Level3, neurons.sub$neur.pred.thresh))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top.thresh <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top.thresh) <- NULL
df.top.thresh$I <- row.names(df.top.thresh)

```





Label again after more transfer labels 






Subset out the Astrocytes for annotation

```{r}

Idents(seu.q) <- 'Level3'
# Level 3
cluster.ids <- c("Astro1","Neurons1","RG1","Astro2","Neurons2",
                 "Astro3","Astro4","Neurons3","RG2","Neurons4",
                 "Astro5","Neurons5","RG3-NPC","DANeurons1","Neurons6",
                 "Mix","RG4","Endothelial-NPC","Neurons8-inh-max","DANeurons2",
                 "DANeurons3","RG5","RG6-epi","RG7","RG-OPC"
                 )

astro.list <- c("Astro1","Astro2","Astro3","Astro4",
                 "Astro5","Mix"
                 )

astro.sub <- subset(seu.q, idents = astro.list)

table(astro.sub$Level3)
DimPlot(astro.sub, label = TRUE, repel = TRUE)


```

Identify markers of astrocyte subgroups

```{r}

Idents(seu.q) <- 'Level3'

ClusterMarkers <- FindAllMarkers(astro.sub)

top3 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=3, wt = avg_log2FC)
DoHeatmap(astro.sub, features = top3$gene, size=3, angle =90, group.bar.height = 0.02)

marker.top3 <- unique(top3$gene)
DotPlot(astro.sub, features = marker.top3) + RotatedAxis()

write.csv(ClusterMarkers, paste(output_path,"ClusterMarkersLevel3AstroSub.csv"))



```


Astro 1 and 2, Astro 3 and 4 are similar in makers.  These might be merged.


Find the markers between astro 3 and 4 only

```{r}

Idents(seu.q) <- 'Level3'

ClusterMarkers34 <- FindMarkers(astro.sub, ident.1 = "Astro3", ident.2 = "Astro4", only.pos = TRUE)
ClusterMarkers34 <- FindMarkers(astro.sub, ident.1 = "Astro4", ident.2 = "Astro3", only.pos = TRUE)

top.up <- ClusterMarkers34 %>% top_n(n=5, wt = avg_log2FC)
top.down <- ClusterMarkers34 %>% top_n(n=-5, wt = avg_log2FC)
DoHeatmap(astro.sub, features = c(rownames(top.up),rownames(top.down)), size=3, angle =90, group.bar.height = 0.02)

# these are all up in astro4
up.down <- c("SPINT2", "RGCC", "NNMT", "TRA2B", "BDP1",
             "DDR2", "COPB2", "SELENOS", "ALDH3A2", "DDX42", "CP", "CPVL", "ARL4C")
# the expression was all low, higher in mixed
# no good markers

# up in astro3 but not significant
up.down <- c("ATP1A2", "PTN", "APOD", "DLK1", "PABPN1", "TTYH1", "CDO1")
DotPlot(astro.sub, features = up.down) + RotatedAxis() 


# PTN is up in Astro3

```

I will merge Astro3 and Astro4
Maybe 1 and 2 also need to be merged


```{r}

Idents(astro.sub) <- 'Level3'

ClusterMarkers12 <- FindMarkers(astro.sub, ident.1 = "Astro1", ident.2 = "Astro2", only.pos = TRUE)
ClusterMarkers21 <- FindMarkers(astro.sub, ident.1 = "Astro2", ident.2 = "Astro1", only.pos = TRUE)

# astro2 up
up.down <- c("ESM1", "RSPO2", "HTRA1", "KCNQ10T1", "SOX2", "CCSER2", "CALD1", "PJA2")
DotPlot(astro.sub, features = up.down) + RotatedAxis() 

# good markers astro2: ESM1, SOX2, PJA2

# astro1 up not significant
up.down <- c("NDUFAB1", "RPS29", "ATP5PO", "RNF5", "HDDC2", "POLR1D", "FAM229B", "EEF1D")
DotPlot(astro.sub, features = up.down) + RotatedAxis() 
# not good marker for astro 1


```

I'll merge astro1 and astro2,  also astro3 and astro4

```{r}

Idents(seu.q) <- 'RNA_snn_res.1.8'

#Idents(seu.q) <- 'Level3'
# Level 3
# merge astro1 + astro2 = astro1, astro3+ astro4 = astro2, now astro5 = astro3
cluster.ids <- c("Astro1","Neurons1","RG1","Astro1","Neurons2",
                 "Astro2","Astro2","Neurons3","RG2","Neurons4",
                 "Astro3","Neurons5","RG3-NPC","DANeurons1","Neurons6",
                 "Mix","RG4","Endothelial-NPC","Neurons8-inh-max","DANeurons2",
                 "DANeurons3","RG5","RG6-epi","RG7","RG-OPC"
                 )


names(cluster.ids) <- levels(seu.q)
seu.q <- RenameIdents(seu.q, cluster.ids)
seu.q$Level4 <- Idents(seu.q)




astro.list <- c("Astro1","Astro2","Astro3","Mix"
                 )

astro.sub <- subset(seu.q, idents = astro.list)

table(astro.sub$Level4)
DimPlot(astro.sub, group.by = "Level4")



```

```{r}

# cluster markers with new Astro groupings
Idents(seu.q) <- 'Level4'

ClusterMarkers <- FindAllMarkers(astro.sub, only.pos = TRUE)

top3 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=5, wt = avg_log2FC)
DoHeatmap(astro.sub, features = top3$gene, size=3, angle =90, group.bar.height = 0.02)

marker.top3 <- unique(top3$gene)
DotPlot(astro.sub, features = marker.top3) + RotatedAxis()

write.csv(ClusterMarkers, paste(output_path,"ClusterMarkersLevel4AstroSub.csv"))

### how will this look without mix?


astro.list <- c("Astro1","Astro2","Astro3"
                 )

astro.sub2 <- subset(seu.q, idents = astro.list)

table(astro.sub2$Level4)
DimPlot(astro.sub2, group.by = "Level4")

Idents(seu.q) <- 'Level4'

ClusterMarkers <- FindAllMarkers(astro.sub2, only.pos = TRUE)

top3 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=5, wt = avg_log2FC)
DoHeatmap(astro.sub2, features = top3$gene, size=3, angle =90, group.bar.height = 0.02)

marker.top3 <- unique(top3$gene)
DotPlot(astro.sub2, features = marker.top3) + RotatedAxis()

write.csv(ClusterMarkers, paste(output_path,"ClusterMarkersLevel4AstroSubnoMix.csv"))


```

There are clearer markers without the mix population

Check the GO terms in the astrocyte subtypes

```{r}
library(enrichR)

setEnrichrSite("Enrichr") # Human genes
# list of all the databases

# libaries with cell types
dbs <- listEnrichrDbs()
dbs
db <- c('GO_Cellular_Component_2015','GO_Biological_Process_2015',
        'CellMarker_Augmented_2021','Azimuth_Cell_Types_2021')

# enrichr(genes, databases = NULL)

group.list <- as.character(unique(ClusterMarkers$cluster))


group <- "Astro1"

#for(group in group.list){
print(group)
Up.list <- ClusterMarkers %>% filter(cluster == group & avg_log2FC > 0)
genes <- Up.list$gene

Er <- enrichr(genes, databases = db)
print(plotEnrich(Er[[1]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[3]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[4]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))

t.GOcell <- Er[[1]] %>% select(Term, Genes, Combined.Score)
print(t.GOcell)

t.GObio <- Er[[2]] %>% select(Term, Genes, Combined.Score)
print(t.GObio)

t.CellMarker <- Er[[3]] %>% select(Term, Genes, Combined.Score)
print(t.CellMarker)


t.Azi <- Er[[4]] %>% select(Term, Genes, Combined.Score)
print(t.Azi)
#}



```


Predict the Astrocyte subgroups

```{r}
seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/Nowakowski_dev_cortext.RDS")
Idents(seu.r) <- "WGCNAcluster"
Idents(astro.sub2) <- "Level4"

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = astro.sub2, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$WGCNAcluster, k.weight = 50)
astro.sub2 <- AddMetaData(astro.sub2, predictions$predicted.id, col.name = "N.pred")
astro.sub2 <- AddMetaData(astro.sub2, predictions$prediction.score.max, col.name = "prediction.score.max")

table(astro.sub2$N.pred)

astro.sub2$N.pred.thresh <- ifelse(astro.sub2$prediction.score.max > 0.50, astro.sub2$N.pred, "none")

DimPlot(astro.sub2, group.by = 'N.pred.thresh')
table(astro.sub2$N.pred.thresh)


# all predicitions
t.lables <- as.data.frame(table(astro.sub2$Level4, astro.sub2$N.pred))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)

# with threshold
t.lables <- as.data.frame(table(astro.sub2$Level4, astro.sub2$N.pred.thresh))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top.thresh <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top.thresh) <- NULL
df.top.thresh$I <- row.names(df.top.thresh)

```
Predictions developing forebrain 

```{r}
seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/Karolinski_DevForebrain_downsample_Level1.RDS")

Idents(seu.r) <- "Clusters"
Idents(astro.sub2) <- "Level4"


astro.sub2 <- ScaleData(astro.sub2)

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = astro.sub2, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$Clusters, k.weight = 30)
astro.sub2 <- AddMetaData(astro.sub2, predictions$predicted.id, col.name = "k.pred")
astro.sub2 <- AddMetaData(astro.sub2, predictions$prediction.score.max, col.name = "prediction.score.max")

table(astro.sub2$k.pred)

astro.sub2$k.pred.thresh <- ifelse(astro.sub2$prediction.score.max > 0.70, astro.sub2$k.pred, "none")

DimPlot(astro.sub2, group.by = 'N.pred.thresh')
table(astro.sub2$k.pred.thresh)


# all predicitions
t.lables <- as.data.frame(table(astro.sub2$Level4, astro.sub2$k.pred))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)

# with threshold
t.lables <- as.data.frame(table(astro.sub2$Level4, astro.sub2$k.pred.thresh))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top.thresh <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top.thresh) <- NULL
df.top.thresh$I <- row.names(df.top.thresh)

```

Kamath astrocyte subgroups 

```{r}

seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/Macosko_Data/PD_astro.Rds")
Idents(seu.r) <- "Cell_Subtype"
Idents(astro.sub2) <- "Level4"

seu.r <- NormalizeData(seu.r, normalization.method = "LogNormalize", scale.factor = 10000)
seu.r <- FindVariableFeatures(seu.r, selection.method = "vst", nfeatures = 2000)
seu.r <- ScaleData(seu.r)
seu.r <- RunPCA(seu.r)

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = astro.sub2, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$Cell_Subtype, k.weight = 30)
astro.sub2 <- AddMetaData(astro.sub2, predictions$predicted.id, col.name = "Mas.pred")
astro.sub2 <- AddMetaData(astro.sub2, predictions$prediction.score.max, col.name = "prediction.score.max")

table(astro.sub2$Mas.pred)

astro.sub2$Mas.pred.thresh <- ifelse(astro.sub2$prediction.score.max > 0.90, astro.sub2$Mas.pred, "none")

DimPlot(astro.sub2, group.by = 'Mas.pred.thresh')
table(astro.sub2$Mas.pred.thresh)


# all predicitions
t.lables <- as.data.frame(table(astro.sub2$Level4, astro.sub2$Mas.pred))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)

# with threshold
t.lables <- as.data.frame(table(astro.sub2$Level4, astro.sub2$Mas.pred.thresh))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top.thresh <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top.thresh) <- NULL
df.top.thresh$I <- row.names(df.top.thresh)

```

Adult brain midbrain and striatum 

```{r}

seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/Bhaduri_midbrain_striatum.RDS")
Idents(seu.r) <- "cell_cluster"
Idents(astro.sub2) <- "Level4"

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = astro.sub2, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$cell_cluster, k.weight = 30)
# note that if there are not enough anchors you need to reduce the k.weight from default 50
astro.sub2 <- AddMetaData(astro.sub2, predictions$predicted.id, col.name = "Bha.pred")
astro.sub2 <- AddMetaData(astro.sub2, predictions$prediction.score.max, col.name = "prediction.score.max")

table(astro.sub2$Bha.pred)

astro.sub2$Bha.pred.thresh <- ifelse(astro.sub2$prediction.score.max > 0.90, astro.sub2$Bha.pred, "none")

DimPlot(astro.sub2, group.by = 'Bha.pred.thresh')
table(astro.sub2$Bha.pred.thresh)


# all predicitions
t.lables <- as.data.frame(table(astro.sub2$Level4, astro.sub2$Bha.pred))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)

# with threshold
t.lables <- as.data.frame(table(astro.sub2$Level4, astro.sub2$Bha.pred.thresh))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top.thresh <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top.thresh) <- NULL
df.top.thresh$I <- row.names(df.top.thresh)

```
Will be best to choose some subtype markers
Astro1 - most predicted as astrocytes, has GO terms differentation 
Astro2 - GO terms involving extracellular matrix
Astro3 - Go terms involving cell adhsion and proteins 




Subset Radial Glia to get subtypes

```{r}

DimPlot(seu.q, group.by = 'Level3')
DimPlot(seu.q, group.by = 'Levels')

Idents(seu.q) <- 'Levels'
RG.sub <- subset(seu.q, idents = "RG")
DimPlot(RG.sub, group.by = 'Level3', split.by = 'orig.ident')



```


First calculate DGE and determine if some subgroups should be merged

```{r}

Idents(RG.sub) <- 'Level3'
ClusterMarkers <- FindAllMarkers(RG.sub, only.pos = TRUE)

top3 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=3, wt = avg_log2FC)
DoHeatmap(RG.sub, features = top3$gene, size=3, angle =90, group.bar.height = 0.02)

marker.top3 <- unique(top3$gene)
DotPlot(RG.sub, features = marker.top3) + RotatedAxis()

output_path <- "/Users/rhalenathomas/Documents/Projects_Papers/PhenoID/ForFigures/scRNA/"
write.csv(ClusterMarkers, paste(output_path,"ClusterMarkersRG.csv"))

```
Many significant markers but not a large LFC for any
Indeed the marker genes are highly overlapping between many groups
RG7 and RG-OPC are distinctive from the others

RG1 markers are expressed in all groups
RG1,RG2,RG3,RG4 are all very overlapping
RG5 RG6 are overlapping

RG5 and RG6 aren't even in the Radial Glia sorted population

Subset just the original RG population and look for markers there 

```{r}

Idents(RG.sub) <- 'orig.ident'

RG.FACS <- subset(RG.sub, idents = "RadialGlia")

DimPlot(RG.FACS, group.by = 'Level3')


```


Get markers for these groups

```{r}

Idents(RG.FACS) <- 'Level3'
ClusterMarkers <- FindAllMarkers(RG.FACS, only.pos = TRUE)

top3 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=3, wt = avg_log2FC)
DoHeatmap(RG.FACS, features = top3$gene, size=3, angle =90, group.bar.height = 0.02)

marker.top3 <- unique(top3$gene)
DotPlot(RG.FACS, features = marker.top3) + RotatedAxis()

top.10 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=10, wt = avg_log2FC)


```

Maybe RG1 and RG2 should be merged 

```{r}

# RG1 vs RG2

clustermarkers.1.2 <- FindMarkers(RG.FACS, ident.1 = "RG1", ident.2 = "RG2")

clustermarkers.2.4 <- FindMarkers(RG.FACS, ident.1 = "RG2", ident.2 = "RG4")


```

No positive markers for RG1 vs RG2 however there are lots of up regulated markers in RG2


Subset out RG1-4

```{r}

Idents(RG.sub) <- 'Level3'
RG.sub2 <- subset(RG.sub, idents = c("RG1","RG2","RG3-NPC","RG4"))


Idents(RG.sub2) <- 'Level3'
ClusterMarkers <- FindAllMarkers(RG.sub2, only.pos = TRUE)

top3 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=3, wt = avg_log2FC)
DoHeatmap(RG.sub2, features = top3$gene, size=3, angle =90, group.bar.height = 0.02)

marker.top3 <- unique(top3$gene)
DotPlot(RG.sub2, features = marker.top3) + RotatedAxis()

top.10 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=10, wt = avg_log2FC)

```
RG2 and 4 are also overlapping in marker expression - but in 2 vs 4 there are positive markers in both directions 

I will merge RG2 and RG4

Test if RG5 and RG6 should be merged into one group

```{r}

clustermarkers.5.6 <- FindMarkers(RG.sub, ident.1 = "RG5", ident.2 = "RG6-epi")

```

There are very different and shouldn't be merged


```{r}

# merge RG1 and RG2
DimPlot(RG.sub)
# now the numbers will all shift although maybe RG6 should be named epithelial cells and RG-OPC as OPCs

cluster.ids <- c("RG1","RG1","RG2","RG3","RG4","epi","RG5","opc"
                 )


names(cluster.ids) <- levels(RG.sub)
RG.sub <- RenameIdents(RG.sub, cluster.ids)
RG.sub$Level4 <- Idents(RG.sub)
DimPlot(RG.sub, group.by = 'Level4')


```

Check the cluster markers again

```{r}


Idents(RG.sub) <- 'Level4'

ClusterMarkers <- FindAllMarkers(RG.sub, only.pos = TRUE)

top3 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=3, wt = avg_log2FC)
DoHeatmap(RG.sub, features = top3$gene, size=3, angle =90, group.bar.height = 0.02)

marker.top3 <- unique(top3$gene)
DotPlot(RG.sub, features = marker.top3) + RotatedAxis()

top.10 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=10, wt = avg_log2FC)

# these markers are terrible
# need to calculate separately 
#RG4 vs Epi

```
The markers are all actually lower even though it says a positive LFC for RG1 (merged with 2)

```{r}
Idents(RG.sub) <- "Level3"

RG1.markers.1 <- FindMarkers(RG.sub, ident.1 = "RG1", ident.2 = c("RG2","RG-OPC"))
RG1.markers.2 <- FindMarkers(RG.sub, ident.1 = "RG1", ident.2 = c("RG2","RG3-NPC","RG4" ))
RG1.markers.3 <- FindMarkers(RG.sub, ident.1 = "RG1", ident.2 = c("RG2","RG3-NPC","RG4","RG5","RG6-epi","RG-OPC"))

table(RG.sub$Level3)

```




I'll have a look at the predictions and marker list

```{r}


# predictions
t.lables <- as.data.frame(table(RG.sub$Level3, RG.sub$Bha.mid.stri.pred))
t.lables <- as.data.frame(table(RG.sub$Level3, RG.sub$AIW60.pred))
t.lables <- as.data.frame(table(RG.sub$Level3, RG.sub$AIW120.pred))
t.lables <- as.data.frame(table(RG.sub$Level3, RG.sub$MBOAIW.pred)) # same as AIW120 but predictions are slighlty different, mostly the same
t.lables <- as.data.frame(table(RG.sub$Level3, RG.sub$MBOAST23.pred))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)



```

There are not really RG in mature brains

Do some more predictions with the developing Brain

```{r}

seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/Karolinski_DevForebrain_downsample_Level1.RDS")

Idents(seu.r) <- "Clusters"
Idents(RG.sub) <- "Level3"


seu.q <- ScaleData(RG.sub)

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = seu.q, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$Clusters, k.weight = 30)
seu.q <- AddMetaData(seu.q, predictions$predicted.id, col.name = "k.pred")
seu.q <- AddMetaData(seu.q, predictions$prediction.score.max, col.name = "prediction.score.max")

table(seu.q$k.pred)


#  predictions
t.lables <- as.data.frame(table(seu.q$Level3, seu.q$k.pred))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 

top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)




```



Now the developing cortex data

```{r}


seu.r <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/PublicData/Nowakowski_dev_cortext.RDS")
Idents(seu.r) <- "WGCNAcluster"
Idents(RG.sub) <- "Level3"

seu.q <- ScaleData(RG.sub)

# find the reference anchors
anchors <- FindTransferAnchors(reference = seu.r, query = seu.q, dims = 1:25)
print("getting predictions")
predictions <- TransferData(anchorset = anchors, refdata = seu.r$WGCNAcluster, k.weight = 30)
seu.q <- AddMetaData(seu.q, predictions$predicted.id, col.name = "devcor.pred")
seu.q <- AddMetaData(seu.q, predictions$prediction.score.max, col.name = "prediction.score.max")

table(seu.q$devcor.pred)


#  predictions
t.lables <- as.data.frame(table(seu.q$Level3, seu.q$devcor.pred))
t.lables$Freq <- as.double(t.lables$Freq)
ggplot(t.lables, aes(y = Freq, x = Var1, fill = Var2)) + geom_bar(position = "stack", stat= "identity") + RotatedAxis() 




top.pred.celltype <- as.data.frame(t.lables  %>% group_by(Var1)  %>% top_n(2, Freq))
df.top <- top.pred.celltype[order(top.pred.celltype$Var1,-top.pred.celltype$Freq),]
row.names(df.top) <- NULL
df.top$I <- row.names(df.top)
head(df.top)

```


Have a look at the potential markers for RG1 and RG2

```{r}
markers <- c("TMSB4X", "TFPI2","RBP1", "LRRC75A", "RPS12", "RPL34", "RPS13")

DotPlot(RG.sub, features= markers, group.by = 'Level3') +RotatedAxis()
DotPlot(RG.sub, features= markers, group.by = 'Level4') +RotatedAxis()

```

All the RG1 markers are high in RG-OPC, but that small cluster is predicted differently from RG1


```{r}

# there are no good positive markers for RG1
# calculate again with negative markers included
Clustermarkers.all <- FindAllMarkers(RG.sub, logfc.threshold = 0.5, test.use = "MAST", min.cells.feature = 15)

genes.down.RG1 <- Clustermarkers.all %>% group_by(cluster) %>% filter(avg_log2FC < 0 & p_val_adj < 0.05 & cluster == "RG1")
genes <- genes.down.RG1$gene


Clustermarkers <- FindAllMarkers(RG.sub, test.use = "MAST", only.pos = TRUE)

```


Check out the down regulated pathways

```{r}


library(enrichR)

setEnrichrSite("Enrichr") # Human genes
# list of all the databases

# libaries with cell types
dbs <- listEnrichrDbs()
dbs
db <- c('GO_Cellular_Component_2015','GO_Biological_Process_2015',
        'Human_Gene_Atlas', 'KEGG_2013')

# enrichr(genes, databases = NULL)
#Up.list <- ClusterMarkers %>% filter(cluster == group & avg_log2FC > 0)
#genes <- Up.list$gene

#down in RG1


Er <- enrichr(genes, databases = db)
print(plotEnrich(Er[[1]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[3]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))
print(plotEnrich(Er[[4]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value"))

t.GOcell <- Er[[1]] %>% select(Term, Genes, Combined.Score)
print(t.GOcell)

t.GObio <- Er[[2]] %>% select(Term, Genes, Combined.Score)
print(t.GObio)

t.CellMarker <- Er[[3]] %>% select(Term, Genes, Combined.Score)
print(t.CellMarker)


t.Azi <- Er[[4]] %>% select(Term, Genes, Combined.Score)
print(t.Azi)

# not useful


```

Check markers from 
Pollen ... Kriegstein 

```{r}

# RG markers were up in one group of RG and down another 
# SLC1A3, PAX6, SOX2, PDGFD, GLI3 up in own group, the other has usp STMN2 and NEUROD6
# rg.con <- c("VIM","HES")
rg <- c("SLC1A3", "PAX6", "SOX2", "PDGFD", "GLI3", "STMN2", "NEUROD6", "VIM", "HES1")
DotPlot(RG.sub, features= rg, group.by = 'Level3') +RotatedAxis()

# RG4 matches the first pop SLCI1A3, PAX6 SOX2 also one new marker PTPRz1 very high


# progenitor markers canonical and novel
# EOMES = TBR2
pro <- c("TBR2", "ELAVL4", "NEUROG1", "NEUROD1", "NEUROD4", "PPP1R17", "PENK")
DotPlot(RG.sub, features= pro, group.by = 'Level3') +RotatedAxis()

# RP7 has some progen markers NEUROD4
# RG-OPC expresses PENK progenitor markers

# CRYAB is a novel G1 marker

# ventral RG markers
rgv <- c("CRYAB", "PDGFD", "TAGLN2", "FBXO32", "PALLD")
# out svz 
osvz <- c("HOPX", "PTPRZ1", "TNC", "FAM107A", "MOXD1")
DotPlot(RG.sub, features= rgv, group.by = 'Level3') +RotatedAxis()
DotPlot(RG.sub, features= osvz, group.by = 'Level3') +RotatedAxis()

# RG2 has high levels of CRYAB in some cells



```
RG4,5,6 classic RG markers and indicated as outer SVZ RG
RG7 has intermediate progenitor markers
RG2 - deviding cells use CRYAB as a marker
RG1 is likely the ventral RG - mark as SOX2- 
RG-OPC has his VIM and also Ventral - TAGLN2

Not sure which marker is good for RG3

```{r}

# check some markers
genes.rg3 <- Clustermarkers %>%filter(cluster == "RG3-NPC" & p_val_adj < 0.5 & avg_log2FC > 1)


rg3 <- c("NEAT1", "POLR2J3.q", "ZMAT1", "DDX17", "PNISR", "LUC7L3", 
         "DST", "MACF1", "ARGUL1", "NKTR", "WSB1", "SFPQ")

DotPlot(RG.sub, features= rg3, group.by = 'Level3') +RotatedAxis()

# All these markers have high expression in RG3, only in about 20-40% cells


```

Still need markers for 5 and 6
Check the list and also distinguish from each
other

- I've already found cluster markers 5 vs 6 above

```{r}

markers.5 <- clustermarkers.5.6 %>% filter(p_val_adj < 0.05 & avg_log2FC > 0.5)

rg.r <- rownames(markers.5)

DotPlot(RG.sub, features = rg.r) + RotatedAxis()

# no good the 5 markers are high in 4 or 2 or RG-OPC

# from all markers the top3 are mitochondrial genes in 5 and 6
# have a look at some other ones

# check some markers
genes.rg5 <- Clustermarkers %>%filter(cluster == "RG5" & p_val_adj < 0.5 & avg_log2FC > 0.5)

# 5 overlaps markers with 4 a lot
rg5 <- c("EMS1", "TRH", "AKR1C1", "PEBP1", "CLU", "PTPRZ1", "MEGF10", "RAB3B", "SELENOW", "BWX1", "NELL2")

DotPlot(RG.sub, features = rg5) + RotatedAxis()

# for RG6-epi 
genes.rg6 <- Clustermarkers %>%filter(cluster == "RG6-epi" & p_val_adj < 0.5 & avg_log2FC > 0.5)
rg6 <- c("EMX2","ECEL1", "CP", "GPHN", "KRT8", "SLC35F2", "SEZ6L2", "ANO10", "KCNJ13",
         "MRPS6", "DMKN", "BRD7", "KRT18", "SLC4A10", "SOX2","NES")
# neuroepithelial cell marker are supposed to be SOX2 and NESTIN

DotPlot(RG.sub, features = rg6) + RotatedAxis()

# for 6 almost all the markers are exclusive / very high in 6 these are not high at all
# The early prediction of epithelial cells doens't seem correct
# GPHN is gyphrin which is a scalphold protein at synapses - weird to be expressed here
# CP is an metalloprotein that binds iron, possibly in transport
# EMX2 is a transcription factor for cortical development
# ECEL1 is a enodopetidase membrane protein

FeaturePlot(seu.q, features = c("CP", "ECEL1", "EMX2", "VIM", "SOX2","HES1","PAX6"))
DimPlot(seu.q, label = TRUE)

# see each of the markers 

rg.sub.markers <- c("SOX2","CRYAB","DDX17","RTPRZ1","CLU","CP","MALAT1","NEUROD4","PENK")
DimPlot(seu.q, label = TRUE)
FeaturePlot(seu.q, features = rg.sub.markers)

FeaturePlot(RG.sub, features = "SOX2", split.by = 'Level3')
# RG-OPC is SOX2 negative
# RG epi and RG7 very low and few 
# RG1 mostly negative but a few very positive cells

FeaturePlot(RG.sub, features = "HES1", split.by = 'Level3')
# a little all over almost all cells high in RG4
FeaturePlot(RG.sub, features = "CRYAB", split.by = 'Level3')



```


Time to add on the final subgroup annotations - double check what makes sense
It might be easier to split the Neurons and Radial glia into subgroups and then compare between those subgroups.
RG it's harder different groups overlap

It seems like RG1,2,4 are more similar

RG4 and RG6 are also very similar
RG4 is the most 'true' RG

RG4 and 2 overlap a lot, also RG6 overlaps with 4 but not with 6

All have VIM
RG1 is negative low other markers HES1, SOX2, PAX6
RG-OPC is negative in PAX6
RG5 and 6 have high upregulation of mitochondrial energy production transcripts


MALAT1 high in 5,6,7 and none in the other groups 
RG7 more and RG-OPC have some intermediate progenitor markers

5 and 6 are not having tones of differentially expressed genes from eachother, but those genes are overlapping with the other groups

 I might merge 2 and 4
 previously tested 2vs4 and nothing is up in 2 compared to 4 
 


```{r}

# neuroD1 very high expression in RG7 but maybe not in all the cells
# test many markers - I've type in and removed for speed
# "SPARCL1"
# DLK1 distinguishes 5 vs 6 but it expressed in most other than 6
# save thing for PTN and ESM1
# up in 6 : MGP actually low in 6 but consistant, not in 5
# FHIT no expressed in all 6 and low everywhere
# ID1 expressed in most of 6 

FeaturePlot(RG.sub, features = c("ID1"), split.by = 'Level3')




```


```{r}





```








