---
title: "Figure 5"
output: github_document
---

# This notebook contains the code for the generation of plots in figure 5
1. CelltypeR annotation of new AIW002 samples from two new batches
2. Comparison between 4 batches of AIW002 hMOs.
3. Running hypergate
4. Analysis of cell populations gated in flowjo based on the hypergate outputs


```{r}
require(Seurat)
require(tidyverse)
require(CelltypeR)

```

# Annotate new AIW002 samples

Read in the Live single cells gated in FlowJo

```{r}
# input path is the pathway to the folder with the data

input_path <- "/GITHUB/CelltypeR/FlowCytometry_Data/AIW_repeat_for_sorting"
flowset <- fsc_to_fs(input_path, downsample = "none")


library(flowWorkspace)  # this library has the function "sampleNames"
sampleNames(flowset)

```

Rename with shorter sample names

```{r}
# rename samples 
sampleNames(flowset) <- c("M10_A","M10_B","M10_C","M10_D",
                           "J10_A", "J10_B", "J10_C")
sampleNames(flowset)

```

Transform and plot to see which features have multiple peaks to use for alignment

```{r}
flowset_biexp <- harmonize(flowset, processing = 'biexp')

plotdensity_flowset(flowset_biexp)
```


Align peaks and create dataframe
```{r}
# few samples have two peaks, the signel peaks are still aligned.
flowset_retro <- harmonize(flowset, processing = 'retro', 
                           two_peaks = c(1),
                       one_peak = c(2:20), threshold = 0.01)

df <- flowset_to_csv(flowset_retro)
write.csv(df, "/GITHUB/CelltypeR/ExampleOuts/AIW002batchesCandD.csv")

```


Now create the Seurat object and cluster

```{r}
AB <- c("CD24","CD56","CD29","CD15","CD184","CD133","CD71","CD44","GLAST","AQP4","HepaCAM", "CD140a","O4")

seu <- make_seu(df, AB_vector = AB)

seu <- get_clusters(seu, method = "louvain",
                         df_input = df,
                         k = 60,
                         resolution = c(0.8,1,1.2),
                         plots = FALSE,
                         save_plots = FALSE)

DimPlot(seu, raster = FALSE, group.by = "Sample", reduction = "umap")
DimPlot(seu, raster = FALSE, label = TRUE, group.by = "RNA_snn_res.1.2")



```



Now annotate the cell types using:
1.Visualization of heatmaps 
2. correlation predictions
3. RandomForest
4. Seurat transfer


Visualize expression

```{r}
AB <- c("CD24","CD56","CD29","CD15","CD184","CD133","CD71","CD44","GLAST","AQP4","HepaCAM", "CD140a","O4")
Idents(seu) <- "RNA_snn_res.1.2"
for (i in AB) {
  print(FeaturePlot(seu, features = i, min.cutoff = 'q1', max.cutoff = 'q97', label = TRUE, raster = FALSE, slot = "scale.data"))
}

```

Heat map by cluster

```{r}
library(data.table)
length(unique(seu$seurat_clusters))
plotmean(plot_type = 'heatmap',seu = seu, group = 'seurat_clusters',
         markers = AB, 
               var_names = c(0:28), slot = 'scale.data', xlab = "Cluster",
               ylab = "Markers")
```
Correlation predictions and summary bar charts

```{r}
# correlation predictions
reference_data <- read.csv("/GITHUB/CelltypeR/ExampleOuts/ReferenceMatrices/FinalReferenceMatrix.csv")
# the reference matrix need to be in the format cell types as rows and markers as columns
# there is a column X with the cell type names
df1 <- reference_data
rownames(df1) <- df1$X # add row names (these are the markers)
df1 <- df1 %>% select(-"X") # remove the column with the marker names
colnames(df1) <- c("Astrocytes","Endothelial","Epithelial","Neurons",
                   "NPC","OPC","Oligo","RadialGlia","StemCell")
df.ref <- as.data.frame(t(df1))
df.ref$X <- rownames(df.ref)

input_df <- read.csv("/GITHUB/CelltypeR/ExampleOuts/AIW002batchesCandD.csv")

cor <- find_correlation(test = input_df, 
                        reference = df.ref, 
                        min_corr = 0.35, 
                        min_diff = 0.01)


```

```{r}
plot_corr(cor, threshold = 0.35, min_cells = 200)
```



See the top correlation predictions
```{r}
seu <- AddMetaData(object=seu, metadata=cor$cell.label, col.name = 'cor.labels')
cor.ann <- get_annotation(seu, seu.cluster = seu$RNA_snn_res.1.2, 
                          seu.label = seu$cor.labels, top_n = 3, 
                          filter_out = c("Unknown","unknown","Mixed", 
                                         "unassigned","Unassigned"), 
                          Label = "CAM")

# top 5 per groups
t.lables <- as.data.frame(table(seu$RNA_snn_res.1.2, seu$cor.labels))
  t.lables$Freq <- as.double(t.lables$Freq)
  colnames(t.lables) <- c("Cluster", "Label","Freq")
  top.labs <- t.lables  %>% group_by(Cluster) %>% top_n(5, Freq)
  sort.tops <- top.labs %>% as.data.frame() %>% arrange(desc(Freq))  %>% arrange(Cluster)
  print(sort.tops)
```

Visualize the cluster predictions

```{r}
# filter cells out that are double labelled and very few 
thresh = 200
freq.cor <- data.frame(table(seu$cor.labels))

keep <- freq.cor %>% filter(Freq > thresh)

cellident <- as.character(keep$Var1)

Idents(seu) <- "cor.labels"
seu.sub <- subset(seu, ident = cellident, invert = FALSE)

plot_lab_clust(seu.sub, seu.sub$seurat_clusters, seu.sub$cor.labels, filter_out = "Unassigned")

```

UMAP with cell types predicted > 200

```{r, fig.width=4}
DimPlot(seu.sub, group.by = "cor.labels", raster = FALSE, label = TRUE)
```


Predictions with RFM

```{r}
# Random Forest predictions
# read in the saved model
rf <- readRDS("SavedTrainedRFM.RDS")


AB <- c("CD24","CD56","CD29","CD15","CD184","CD133","CD71","CD44","GLAST","AQP4","HepaCAM", "CD140a","O4")


rfm.pred <- RFM_predict(seu, rf)
head(rfm.pred)
unique(rfm.pred$Prediction)
seu <- AddMetaData(object=seu, metadata= as.factor(rfm.pred$Prediction), col.name = 'rfm.pred2')
DimPlot(seu, group.by = 'rfm.pred2', raster = FALSE, label = TRUE, repel = TRUE)

plot_lab_clust(seu, seu$seurat_clusters, seu$rfm.pred2, filter_out = c( "Unassigned","Mixed"))

```


Top 5 predictions with RFM and the top prediction per cluster

```{r}

rfm.ann <- get_annotation(seu, seu.cluster = seu$RNA_snn_res.1.2, 
                          seu.label = seu$rfm.pred2, top_n = 3, 
                          filter_out = c("Unknown","unknown","Mixed", 
                                         "unassigned","Unassigned"), 
                          Label = "RFM")

# top 5 per groups
t.lables <- as.data.frame(table(seu$RNA_snn_res.1.2, seu$rfm.pred2))
  t.lables$Freq <- as.double(t.lables$Freq)
  colnames(t.lables) <- c("Cluster", "Label","Freq")
  top.labs <- t.lables  %>% group_by(Cluster) %>% top_n(5, Freq)
  sort.tops <- top.labs %>% as.data.frame() %>% arrange(desc(Freq))  %>% arrange(Cluster)
  print(sort.tops)


```

UMAP of the RFM top predictions

```{r, fig.width=4}
DimPlot(seu, group.by = "rfm.pred2", raster = FALSE, label = TRUE)
```


Predict cell types with Seurat label transfer

```{r}
# seurat transfer anchor predictions
# Full annotated dataset, this has no unknown cells
seu.r<- readRDS("All9MOannaote.RDS") 

# take the AIW hMOs and prepare the reference object
Idents(seu.r) <- "Sample"
levels(seu.r)
seu.r <- subset(seu.r, idents = c("AIW002_0306", "AIW002_0317A", "AIW002_0317B"))
seu.aiw <- seu.r

Idents(seu.r) <- "Celltypes"
seu.r <- subset(seu.r, downsample = 1000)

seu.r <- ScaleData(seu.r, features = AB)
seu.r <- RunPCA(seu.r, features = AB, verbose = FALSE, approx = FALSE)

### predict cell types
seu <- seurat_predict(seu, seu.r, ref_id = 'Celltypes', 
                        refdata = seu.r$Celltypes,
                           down.sample = "none", 
                        markers = AB, kfilter = 50)


seu.ann <- get_annotation(seu, seu.cluster = seu$RNA_snn_res.1.2, 
                          seu.label = seu$seu.pred, top_n = 1, 
                          filter_out = c("Unknown","unknown","Mixed", 
                                         "unassigned","Unassigned"), 
                          Label = "seu")

# top 5 per groups
t.lables <- as.data.frame(table(seu$RNA_snn_res.1.2, seu$seu.pred))
  t.lables$Freq <- as.double(t.lables$Freq)
  colnames(t.lables) <- c("Cluster", "Label","Freq")
  top.labs <- t.lables  %>% group_by(Cluster) %>% top_n(5, Freq)
  sort.tops <- top.labs %>% as.data.frame() %>% arrange(desc(Freq))  %>% arrange(Cluster)
  print(sort.tops)

```
Seurat predictions UMAP
```{r}
DimPlot(seu, label = TRUE, raster = FALSE, group.by = "seu.pred")
```
Make a table of the 3 predictions
```{r}
an.list <- list(cor.ann, rfm.ann, seu.ann)
df.an <- annotate_df(an.list)
df.an
```


Annotated clusters of two new AIW002-02 hMO batches

```{r, fig.width= 10}


cluster.ids <- c("Glia-lineage","Astrocytes 2","Astrocytes 1","Neural-lineage","Radial Glia 1",   # 0-4
                 "Epithelial","Neurons 1","Neurons 1","Radial Glia 1","Neurons 1",  #5-9
                 "OPC","Radial Glia 2","Astrocytes 1", "Neurons 2","Endothelial", #10-14
                 "Radial Glia 1", "Radial Glia 1a", "Glia-lineage","NPC","OPC-like",  #15-19
                 "Neurons 2", "Stem cell like","Neurons 2", "Astrocytes 2", "NPC", #20-24
                 "Neurons 2","Radial Glia 3","Astrocytes 2","Oligodendrocytes"
                 )

seu <- CelltypeR::annotate(seu, annotations = cluster.ids, to_label = "RNA_snn_res.1.2", 
                annotation_name = "Celltypes")

DimPlot(seu, label = TRUE, raster = FALSE, label.size = 6)

cell.order <- c("Astrocytes 1", "Astrocytes 2",
                     "Radial Glia 1","Radial Glia 1a","Radial Glia 2",
                     "Radial Glia 3","Glia-lineage",
                     "Epithelial","Endothelial",
                     "Neurons 1","Neurons 2","NPC","Neural-lineage",
                     "Oligodendrocytes","OPC","OPC-like",
                     "Stem cell like")

clust.colours <- c("chocolate2",# Astrocytes 1
                   "darkorange", # Astrocytes 2
                   "pink", # RG1
                   "deeppink",# RG1a
                   "plum1", #RG2
                   "lightpink3",# RG3
                   "mistyrose2", # Glia-lineage
                   "steelblue3",# epithelial
                   "deepskyblue", # endothelial
                   "mediumpurple1",# neurons 1
                   "purple",# Neurons2
                   "plum3", # NPC
                   "mediumslateblue", # Neural lineage
                   "seagreen3",#Oligo
                   "olivedrab4", # OPC
                   "darkseagreen3",#OPC like
                   "tomato3",# stem like
                   "burlywood3" #extra
                   )

DimPlot(seu, cols = clust.colours, shuffle = TRUE, 
        raster=FALSE, pt.size = 0.1, label = TRUE, 
        group.by = 'Celltypes',
        order = rev(cell.order))

```

save the plot: Figure 5A
```{r}

pdf("UMAP.all.cells2.pdf",width = 9, height = 4.5)
DimPlot(seu, cols = clust.colours, shuffle = TRUE, 
        raster=FALSE, pt.size = 0.001, label = FALSE, 
        group.by = 'Celltypes',
        order = rev(cell.order)) +
theme(legend.text = element_text(size=15), axis.title.y = element_text(size=15), 
         axis.title.x = element_text(size=15), axis.text.y = element_text(size =15),
        axis.text.x = element_text(size =15))
dev.off()


```

Add batch annotations

```{r}
Idents(seu) <- "Sample"
levels(seu)

batch <- c("C","C","D","D","C","C","C")
seu <- annotate(seu, annotations = batch, 
                  to_label= "Sample", 
                  annotation_name = "Batch")

table(seu$Sample, seu$Batch)

```


# Compare across 4 AIW002 batches
Combine the two seurate data objects containing AIW002 samples and make barcharts to compare proportions of cells across batches

```{r}
# change the orig.ident to keep track of samples after merging
seu.aiw$orig.ident <- "AandB"
seu$orig.ident <- "CandD"
# make the merged object
seu.all <- merge(seu.aiw, y = seu, project = "AIW002all.batches")

```

Make a barchart of the proportion of cell types - Figure 5B

```{r}

# cell order and clust.colours defined above)
seu.all$Celltypes <- factor(seu.all$Celltypes, levels = c("Astrocytes 1", "Astrocytes 2",
                     "Radial Glia 1","Radial Glia 1a","Radial Glia 2",
                     "Radial Glia 3","Glia-lineage",
                     "Epithelial","Endothelial",
                     "Neurons 1","Neurons 2","NPC","Neural-lineage",
                     "Oligodendrocytes","OPC","OPC-like",
                     "Stem cell like"))


proportionplots(seu.all, seu.var = seu.all$Batch, seu.lable = seu.all$Celltypes, groups = "AIW002 Batch", my_colours = clust.colours)

```

save the plot figure 5B
```{r}

pdf("PercentCelltypesBatch.pdf", width = 5, height = 4)
proportionplots(seu.all, seu.var = seu.all$Batch, seu.lable = seu.all$Celltypes, groups = "AIW002 Batch", my_colours = clust.colours)
dev.off()


```

Proportionality tests

```{r}

library(scProportionTest)
# create a proportion test object

Idents(seu.all) <- "Celltypes"
prop_test <- sc_utils(seu.all)

# set up the comparison
# need to compare separately
prop_test <- permutation_test(
	prop_test, cluster_identity = "Celltypes",
	sample_1 = "A", sample_2 = "B",
	sample_identity = "Batch"
)

# make the plot
permutation_plot(prop_test)

# save the plot 
# png("prp.test.batchAvsB.png", width = 600, height = 300)
# permutation_plot(prop_test) +  theme(axis.text.x = element_text(size = 12),
#         axis.text.y = element_text(size = 12),
#         axis.title = element_text(size = 14),    # Adjust axis title size
#   legend.text = element_text(size = 12),   # Adjust legend text size
#   legend.title = element_text(size = 14))
# dev.off()


# need to compare separately
prop_test <- permutation_test(
	prop_test, cluster_identity = "Celltypes",
	sample_1 = "A", sample_2 = "C",
	sample_identity = "Batch"
)

# make the plot
permutation_plot(prop_test)

# save the plot 
# png("prp.test.batchAvsC.png", width = 600, height = 300)
# permutation_plot(prop_test) +  theme(axis.text.x = element_text(size = 12),
#         axis.text.y = element_text(size = 12),
#         axis.title = element_text(size = 14),    # Adjust axis title size
#   legend.text = element_text(size = 12),   # Adjust legend text size
#   legend.title = element_text(size = 14))
# dev.off()

# need to compare separately
prop_test <- permutation_test(
	prop_test, cluster_identity = "Celltypes",
	sample_1 = "A", sample_2 = "D",
	sample_identity = "Batch"
)

# make the plot
permutation_plot(prop_test)

# save the plot 
# png("prp.test.batchAvsD.png", width = 600, height = 300)
# permutation_plot(prop_test) +  theme(axis.text.x = element_text(size = 12),
#         axis.text.y = element_text(size = 12),
#         axis.title = element_text(size = 14),    # Adjust axis title size
#   legend.text = element_text(size = 12),   # Adjust legend text size
#   legend.title = element_text(size = 14))
# dev.off()

# need to compare separately
prop_test <- permutation_test(
	prop_test, cluster_identity = "Celltypes",
	sample_1 = "B", sample_2 = "C",
	sample_identity = "Batch"
)

# make the plot
permutation_plot(prop_test)

# save the plot 
# png("prp.test.batchBvsC.png", width = 600, height = 300)
# permutation_plot(prop_test) +  theme(axis.text.x = element_text(size = 12),
#         axis.text.y = element_text(size = 12),
#         axis.title = element_text(size = 14),    # Adjust axis title size
#   legend.text = element_text(size = 12),   # Adjust legend text size
#   legend.title = element_text(size = 14))
# dev.off()

# need to compare separately
prop_test <- permutation_test(
	prop_test, cluster_identity = "Celltypes",
	sample_1 = "B", sample_2 = "D",
	sample_identity = "Batch"
)

# make the plot
permutation_plot(prop_test)

# save the plot 
# png("prp.test.batchBvsD.png", width = 600, height = 300)
# permutation_plot(prop_test) +  theme(axis.text.x = element_text(size = 12),
#         axis.text.y = element_text(size = 12),
#         axis.title = element_text(size = 14),    # Adjust axis title size
#   legend.text = element_text(size = 12),   # Adjust legend text size
#   legend.title = element_text(size = 14))
# dev.off()


# need to compare separately
prop_test <- permutation_test(
	prop_test, cluster_identity = "Celltypes",
	sample_1 = "C", sample_2 = "D",
	sample_identity = "Batch"
)

# make the plot
permutation_plot(prop_test)

# save the plot 
# png("prp.test.batchCvsD.png", width = 600, height = 300)
# permutation_plot(prop_test) +  theme(axis.text.x = element_text(size = 12),
#         axis.text.y = element_text(size = 12),
#         axis.title = element_text(size = 14),    # Adjust axis title size
#   legend.text = element_text(size = 12),   # Adjust legend text size
#   legend.title = element_text(size = 14))
# dev.off()


```
# Running hypergate to get the threshold to gate cells defined by CelltypeR

Read in the labelled object from all cells from the 3 genotypes 9 MO dataset

```{r}
seu <- readRDS("All9MOannaoteAug.RDS") 
```

See the cell type annotations
```{r}
Idents(seu) <- "Celltypes"
levels(seu)

```

Merge together the Astrocytes and Radial Glia subgroups
```{r}

new.ids <- c("Astrocytes", "Astrocytes", "RadialGlia", "RadialGlia", "RadialGlia", "RadialGlia", "Glia-lineage",
             "Epithelial", "Endothelial", "Neurons1", "Neurons2", "NPC", "Neural-lineage", "Oligodendrocytes",
             "OPC", "OPC-like", "Stemlike")

seu <- CelltypeR::annotate(seu, annotations = new.ids, 
                  to_label= "Celltypes", 
                  annotation_name = "GatingCells")

table(seu$GatingCells)

```
See the cell type annotation for gating

```{r}
DimPlot(seu, group.by = "GatingCells", raster = FALSE, label = TRUE)
```

Prepare data for running hypergate

```{r}

# set the identity to the labels to by the hypergate input
Idents(seu) <- 'GatingCells'

# down sample
# repeat again with new subsample
i = 800
set.seed(i)
seu.down <- subset(x= seu, downsample = i)
DimPlot(seu.down, label = TRUE)
table(seu.down$GatingCells)

# create a matrix to be the input for hypergate
input.xm = as.matrix(GetAssayData(seu.down, slot = 'data'))
xm.t <- t(input.xm)
cluster.labels <- as.vector(seu.down$GatingCells)


```

Run hypergate for each cell type
```{r}
library(hypergate)
cell.types <- c("Astrocytes",  "RadialGlia", "Glia-lineage",
             "Epithelial", "Endothelial", "Neurons1", "Neurons2", "NPC", "Neural-lineage", "Oligodendrocytes",
             "OPC", "OPC-like", "Stemlike")

# run gating and get the output
for (cell in cell.types) {
  hg_output <- hypergate(xp = xm.t, gate_vector = cluster.labels, level = cell)
  gating_predicted = subset_matrix_hg(hg_output,xm.t)
  conf.table <- table(ifelse(gating_predicted, "Gated-in", "Gated-out"), ifelse(cluster.labels ==  cell, cell, "Others"))
  print(cell)
  print(conf.table)
  print(hgate_rule(hg_output))
 
}

```


# Analysis of populations gated in FlowJO
Based on the hypergate results cells were gated in FlowJo uses the defined threshold for each relevent antibody. The populations were exported as fcs.  The gated populations from all samples from the 9 hMO dataset (with 3 genotypes) were merged.


```{r}
# read in the cell types gated in FlowJo
# make seurat object
input_path <- "/GITHUB/CelltypeR/FlowCytometry_Data/FlowJoGated"
flowset <- fsc_to_fs(input_path, downsample = 5000)

library(flowWorkspace)  # this library has the function "sampleNames"
sampleNames(flowset)


sampleNames(flowset) <- c("Astrocytes","Endothelial","Epithelial","Glial_lineage","Neural_lineage","Neurons1",
                           "Neurons2", "NPC", "Oligodendrocytes","RadialGlia")
sampleNames(flowset)

```

Check peaks
```{r}

flowset_biexp <- harmonize(flowset, processing = 'biexp')

plotdensity_flowset(flowset_biexp)

```

Align and create dataframe of area expression values FSC-A for each channel/Marker

```{r}
flowset_retro <- harmonize(flowset, processing = 'retro', 
                           two_peaks =  c(8:22),
                       one_peak = c(1:7), threshold = 0.01)

df <- flowset_to_csv(flowset_retro)

write.csv(df, "FlowJoGatedCellPops.csv")
```

Make the seurat object and cluster

```{r}
# marker names to take from the expression df
AB <- c("CD24","CD56","CD29","CD15","CD184","CD133","CD71","CD44","GLAST","AQP4","HepaCAM", "CD140a","O4")
# make the seurat object
seu <- make_seu(df, AB_vector = AB)

# find clusters
seu <- get_clusters(seu, method = "louvain",
                         df_input = df,
                         k = 20,
                         resolution = 1.8,
                         plots = FALSE,
                         save_plots = FALSE)

```

Visualize marker expression
```{r}
for (i in AB) {
  print(FeaturePlot(seu, features = i, min.cutoff = 'q1', max.cutoff = 'q97', label = TRUE, raster = FALSE))
}

```

Heatmap of marker expression by cluster

```{r}
library(data.table)
length(unique(seu$seurat_clusters))
plotmean(plot_type = 'heatmap',seu = seu, group = 'RNA_snn_res.1.8',
         markers = AB, 
               var_names = c(0:34), slot = 'scale.data', xlab = "Cluster",
               ylab = "Markers")

```
# Annotate clusters



Calculate the correlations

```{r}
reference_data <- read.csv("GITHUB/CelltypeR/ExampleOuts/ReferenceMatrices/FinalReferenceMatrix.csv")
# the reference matrix need to be in the format cell types as rows and markers as columns
# there is a column X with the cell type names
df1 <- reference_data
rownames(df1) <- df1$X # add row names (these are the markers)
df1 <- df1 %>% select(-"X") # remove the column with the marker names
colnames(df1) <- c("Astrocytes","Endothelial","Epithelial","Neurons",
                   "NPC","OPC","Oligo","RadialGlia","StemCell")
df.ref <- as.data.frame(t(df1))
df.ref$X <- rownames(df.ref)

df_test <- read.csv("FlowJoGatedCellPops.csv") # the X column needs to be the first column
# run correlation
cor <- find_correlation(test = df_test, 
                        reference = df.ref, 
                        min_corr = 0.25, 
                        min_diff = 0.01)

write.csv(cor, "GatedCellsCorPredictUpdated.csv")

plot_corr(cor, threshold = 0.25, min_cells = 80)

```

Get the CAM predictions

```{r, fig.width=10}

plot_lab_clust(seu, seu$RNA_snn_res.1.8, seu$cor.labels, filter_out = c( "Unassigned","Mixed"))


 t.lables <- as.data.frame(table(seu$RNA_snn_res.1.8, seu$cor.labels))
  t.lables$Freq <- as.double(t.lables$Freq)
  colnames(t.lables) <- c("Cluster", "Label","Freq")
  top.labs <- t.lables  %>% group_by(Cluster) %>% top_n(5, Freq)
  sort.tops <- top.labs %>% as.data.frame() %>% arrange(desc(Freq))  %>% arrange(Cluster)
  print(sort.tops)

  
  cor.ann <- get_annotation(seu, seu.cluster = seu$RNA_snn_res.1.8, 
                          seu.label = seu$cor.labels, top_n = 1, 
                          filter_out = c("Unknown","unknown","Mixed", 
                                         "unassigned","Unassigned"), 
                          Label = "CAM")
  

```

Random Forest Classifier

```{r}

# Random Forest classifier trained with annotations from subset of 9 MBO cells. 
# read in the saved model
rf <- readRDS("RFM_all9hMOAB.RDS")

# test random forest function takes in the seurat object to have cells predicted and the trained model.  Must be trained with the same antibodies

rfm.pred <- RFM_predict(seu, rf)
head(rfm.pred)
unique(rfm.pred$Prediction)

# add the predictions to the seurat object
seu <- AddMetaData(object=seu, metadata= as.factor(rfm.pred$Prediction), col.name = 'rfm.pred')

celltypes <- unique(seu$rfm.pred)
sorted <- as.character(factor(sort(as.character(celltypes))))

DimPlot(seu, group.by = 'rfm.pred', raster = FALSE, label = TRUE)

```


```{r}
 
plot_lab_clust(seu, seu$RNA_snn_res.1.8, seu$rfm.pred, filter_out = c( "Unassigned","Mixed"))


 t.lables <- as.data.frame(table(seu$RNA_snn_res.1.8, seu$rfm.pred))
  t.lables$Freq <- as.double(t.lables$Freq)
  colnames(t.lables) <- c("Cluster", "Label","Freq")
  top.labs <- t.lables  %>% group_by(Cluster) %>% top_n(5, Freq)
  sort.tops <- top.labs %>% as.data.frame() %>% arrange(desc(Freq))  %>% arrange(Cluster)
  print(sort.tops)

  
  rfm.ann <- get_annotation(seu, seu.cluster = seu$RNA_snn_res.1.8, 
                          seu.label = seu$rfm.pred, top_n = 1, 
                          filter_out = c("Unknown","unknown","Mixed", 
                                         "unassigned","Unassigned"), 
                          Label = "rfm.pred")

  
```
Seurat label transfer

```{r}
# read in the reference matrix
# prepare the reference
seu.r<- readRDS("All9MOannaoteAug.RDS") 

table(seu.r$Celltypes)

Idents(seu.r) <- "Celltypes"
seu.r <- subset(seu.r, downsample = 5000)

seu.r <- ScaleData(seu.r, features = AB)
seu.r <- RunPCA(seu.r, features = AB, verbose = FALSE, approx = FALSE)


```

Run the predictions
```{r}
# notes this function automatically selects the total number of markers as the dim.  So max dim = 1:13 here. 
seu <- seurat_predict(seu, seu.r, ref_id = 'Celltypes', 
                        refdata = seu.r$Celltypes,
                           down.sample = "none", 
                        markers = AB)

DimPlot(seu, group.by = 'seu.pred', raster = FALSE, label = TRUE)
```

```{r}
plot_lab_clust(seu, seu$RNA_snn_res.1.8, seu$seu.pred, filter_out = c( "Unassigned","Mixed"))


 t.lables <- as.data.frame(table(seu$RNA_snn_res.1.8, seu$seu.pred))
  t.lables$Freq <- as.double(t.lables$Freq)
  colnames(t.lables) <- c("Cluster", "Label","Freq")
  top.labs <- t.lables  %>% group_by(Cluster) %>% top_n(5, Freq)
  sort.tops <- top.labs %>% as.data.frame() %>% arrange(desc(Freq))  %>% arrange(Cluster)
  print(sort.tops)

  
  seu.ann <- get_annotation(seu, seu.cluster = seu$RNA_snn_res.1.8, 
                          seu.label = seu$seu.pred, top_n = 1, 
                          filter_out = c("Unknown","unknown","Mixed", 
                                         "unassigned","Unassigned"), 
                          Label = "seu.pred")
```

See the predictions for each method

```{r}
# original cell type
cell.ann <- get_annotation(seu, seu.cluster = seu$RNA_snn_res.1.8, 
                          seu.label = seu$Sample, top_n = 1, 
                          Label = "Cells")
an.list <- list(cor.ann, rfm.ann, seu.ann, cell.ann)
df.an <- annotate_df(an.list)
df.an

```

Add annotations
Used in Figure 5D and E

```{r, fig.width=4}

cluster.ids <- c("NL","GL","GL", "GL","Epi", # 0-4
                 "Neur2","NL","RG1","Neurons1-2","Neur2-oligo", #5-9 
                 "Epi","NPC","Endo","Astro1","RG1", #10-14 
                 "NPC","Neur1","NPC","Endo","Neur1",    #15-19
                 "Astro1","Neur1","Oligo","Neur2","NL", #20-24
                 "NPC","Astro1","Endo","RG3","Astro1",
                 "Neur2","Endo","RG1","NPC", "Oligo"
                 )


seu <- annotate(seu, annotations = cluster.ids, 
                  to_label= "seurat_clusters", 
                  annotation_name = "Annotations1")

# change the names to long forms
cluster.ids <- c("Neural-lineage","Glial-lineage","Epithelial","Neurons2",
                 "RadialGlia1","Neurons1and2","Neurons2andOligo","NPC","Endothelial",
                 "Astrocytes1","Neurons1","Oligodendrocytes","RadialGlia3")

seu <- annotate(seu, annotations = cluster.ids, 
                  to_label= "Annotations1", 
                  annotation_name = "Celltypes")

# reorder the cell types
cell.order <- c("Astrocytes1","RadialGlia1","RadialGlia3","Glial-lineage",
                "Epithelial","Endothelial","Neurons1",
                "Neurons2","Neurons1and2","Neurons2andOligo","NPC",
                "Neural-lineage","Oligodendrocytes")
 
 clust.colours <- c("chocolate2",# Astrocytes 1
                   #"darkorange", # Astrocytes 2
                   #"orange", #Astro
                   "pink", # RG1
                   #"deeppink",# RG1a
                   #"plum1", #RG2
                   "lightpink3",# RG stemlike
                   "mistyrose2", # Glia-lineage
                   "steelblue3",# epithelial
                   "deepskyblue", # endothelial
                   #"blue", # epi-endo
                   "mediumpurple1",# neurons 1
                   "purple",# Neurons2
                   "mediumpurple3", # Neurons1 and 2
                   "grey", # Neurons2 and oligo
                   #"darkgrey",
                   "plum3", # NPC
                   #"tomato3",# stem like
                   "mediumslateblue", # Neural lineage
                   "seagreen3",#Oligo
                   "olivedrab4", # OPC
                   "darkseagreen3",#OPC like
                   "burlywood3" #extra 
                   )

 # need to set the factor to control the order
 
 seu$Celltypes <- factor(seu$Celltypes, levels = cell.order)
 
 DimPlot(seu, cols = clust.colours, shuffle = TRUE, 
        raster=FALSE, pt.size = 0.1, label = TRUE, 
        group.by = "Celltypes")
 

```


UMAP coloured by the sample input cell types
Figure 5D

```{r}
#unique(seu$Sample)

sample.colours  <- c("chocolate2",# Astrocytes 1
                   "steelblue3",# epithelial
                   "deepskyblue", # endothelial
                   "mistyrose2", # Glia-lineage
                   "mediumslateblue", # Neural lineage
                   "mediumpurple1",# neurons 1
                   "purple",# Neurons2
                   "plum3", # NPC
                   "seagreen3",#Oligo
                   "pink" # RG1
                   )



DimPlot(seu, cols = sample.colours, shuffle = TRUE, 
        raster=FALSE, pt.size = 0.1, label = TRUE, 
        group.by = "Sample") +
theme(legend.text = element_text(size=16), axis.title.y = element_text(size=16), 
         axis.title.x = element_text(size=16), axis.text.y = element_text(size =16),
        axis.text.x = element_text(size =16))


pdf("GatedCellPopsUMAP.pdf", width = 9, height = 4.5)
DimPlot(seu, cols = sample.colours, shuffle = TRUE, 
        raster=FALSE, pt.size = 0.1, label = FALSE, 
        group.by = "Sample") +
theme(legend.text = element_text(size=16), axis.title.y = element_text(size=16), 
         axis.title.x = element_text(size=16), axis.text.y = element_text(size =16),
        axis.text.x = element_text(size =16))
dev.off()


```

Figure 5E barchart of proporitons of cell types
```{r}
proportionplots(seu, seu.var = seu$Sample, seu.lable = seu$Celltypes, 
                groups = "Gated samples",  
                my_colours = clust.colours)
```

