---
title: "CelltypeR workflow"
output: html_notebook
---

1. Data preprocessing: put all timepoints and replicates into one folder
a. Read FlowJo files into R.
b. Create a data frame with intensity measurements for each marker for all          samples within the experiment to be analyzed.  
- downsample taking a 2000 cells from each sample to speed up computation.
c. Harmonize data 
d. Create dataframe
e. Create a Seurat object

2. Cluster data
3. Annotate clusters with cell types.
a. Predict cell types by correlation matrix
b. Visualize expression by cluster for manual annotation
c. Add annotations to Seurat object
c. Train RandomForest classifier.

4. Reprocess with all cell in samples
a. preprocess
b. cluster
c. annotate (now seurat label transfer and random forest can be used.)

5. Compare cell type proportions and expression levels across the time course. 



```{r}
# load necessary libraries 
library(Seurat)
library(dplyr) 
library(ggplot2)
library(CelltypeR)
library(flowCore)

```


# Preprocessing

Read in the flow data
This data should be the gated live cells.  
All samples need to be in one folder.
Here we have 

```{r}

input_path <- "/Users/rhalenathomas/Documents/Data/FlowCytometry/PhenoID/TimeCourseAIW/LiveCells"
output_path <- "/Users/rhalenathomas/Documents/Data/FlowCytometry/PhenoID/TimeCourseAIW/Analysis/"

# 1.a Read in FlowJo Files 

# down sample to take cells from each of 4 time points and 4 replicates
flowset <- fsc_to_fs(input_path, downsample = 2000)
# down sample can be a number, 'none' or 'min'

# look at file names and rename with shorter sample names
sampleNames(flowset) # function in flowCore

```


Rename samples with shorter names

```{r}
sampleNames(flowset) <- sampleNames(flowset) <- c("AIW_150_R1","AIW_150_R2","AIW_150_R3","AIW_150_R4",
  "AIW_30_R1","AIW_30_R2","AIW_30_R3","AIW_30_R4",
  "AIW_60_R1","AIW_60_R2","AIW_60_R3","AIW_60_R4",
  "AIW_100_R1","AIW_100_R2","AIW_100_R3","AIW_100_R4"
                                                  )
sampleNames(flowset)
```

Harmonize data to remove batch or technical variation

This requires us to look and see where there are two peaks to align. We need to visualize the peaks of the transformed data before aligning.

```{r}


# we can decided what level of processing to choose with the argument 'processing'
# biexp only applies a biexponential transformation
# align applies biexp transform and then aligns peaks
# retro (default), transforms, aligns and then reverse transforms
flowset_biexp <- harmonize(flowset, processing = 'biexp')
# we can visualize the peaks to see where there are two peaks for alignment

# we need to enter the column index for which peaks to align, the alignment for one or two peaks is not the same. 
#plotdensity_flowset(flowset)

p1 <- plotdensity_flowset(flowset_biexp) # to see the peaks
p1
# if there are two peaks assign as two peaks and one peak as one in the code chunk below

flowset_align <- harmonize(flowset, processing = 'align', 
                           two_peaks = c(9:11,15,17:21,25:42),
                       one_peak = c(1:8,12:14,16,24,43), threshold = 0.01)

flowset_retro <- harmonize(flowset, processing = 'retro', 
                           two_peaks = c(9:11,15,17:21,25:42),
                       one_peak = c(1:8,12:14,16,24,43), threshold = 0.01)
### test again later without running library(flowCore)
df <- flowset_to_csv(flowset_retro)

head(df)
# here we have 3 measures of each channel - we want the area measure and only this should have been selected.
# Select the first value from 

colnames(df)

```

Now we have made all the different processing of the fsc files.  We can visualize the intensity in cell density plots to see the alignment

```{r}
#plotdensity_flowset(flowset)

plotdensity_flowset(flowset_biexp)
plotdensity_flowset(flowset_align)
#plotdensity_flowset(flowset_retro)


```




```{r}



# the function make_seu will take in the df of expression and Antibody/Marker list as a vector and create a seurat object. Values are scaled. Marker expression will be in the "RNA" slot. PCA is calculated using AB vector as the features 
# make sure to always keep the same antibody order or your labels will not be correct

# antibody features in order to appear on the plots
AB <- c("TH","CD24","CD56","CD29","CD15","CD184","CD133","SSEA4","CD44","CD49f","CD140a")

df_select <- df %>% select(c("TH","CD24","CD56","CD29","CD15","CD184","CD133","SSEA4","CD44","CD49f","CD140a","Sample"))

seu <- make_seu(df_select, AB_vector = AB)


```
You can save the data frame and seurat object for later

```{r}


# to save the df for later
write.csv(df, paste(output_path,"df2000Timecourse4.csv", sep = ""))

# save the seurat object
saveRDS(seu, paste(output_path, "seuratObjectTimecourseAIW_4.RDS", sep = ""))



```
Check out the data object

```{r}

RidgePlot(seu, features = AB, log = TRUE)
DimPlot(seu, group.by = "Sample")

```


# Cluster

```{r}
# note the dimensions (number of PC to include needs to be one less than the number in the antibody panel)
explore_param(input = seu, 
                          cluster_method = "louvain", 
                          df_input = df.input, 
                          flow_k = NULL, 
                          pheno_lou_kn = c(80,120,200), 
                          lou_resolution = c(0.6), 
                          pcdim = 1:10,
                          run.plot = TRUE, 
                          run.stats = FALSE, 
                          save_to = NULL)
# this function does not add the clustering to the seurat object it is to explore and run stats.  If you choose save = output_path then you will save a seurat object with the clustering, one for each kn.


```


After checking parameters run cluster function to make add the clustering information into the 

```{r}

seu <- get_clusters(seu, method = "louvain",
                         df_input = df.input,
                         k = 80,
                         resolution = c(1),
                         plots = TRUE,
                         save_plots = FALSE)
# if save_plots is TRUE the feature plots will be save as 13 plots in one file. The UMAP with cluster numbers will also be saved.

```


# Annotate clusters
1. Visualization for manual annotation. - output by clustering function
2. CAM (Correlation assignment model) - requires reference matrix
3. RFM (Random Forest Model) - requires annotated matching flow dataset
4. Seurat label transfer - requires annotated matching flow data in a seurat object

Visualize expression on UMAP and with heat maps

```{r}

AB <- c("TH","CD24","CD56","CD29","CD15","CD184","CD133","SSEA4","CD44","CD49f","CD140a")
# the cluster labels will match the active ident
Idents(seu) <- "RNA_snn_res.1"
# this will let us see one at at time
for (i in AB) {
  print(FeaturePlot(seu, features = i, min.cutoff = 'q1', max.cutoff = 'q97', label = TRUE))
}


```

Visualize your reference matrix

```{r}
df <- read.csv("/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/ReferenceMatrixPanel2.csv")

# we need to remove the marker column and add these are row names
rownames(df) <- df$X
df <- df %>% select(-"X")
# heatmap function take a matix
mat <- as.matrix(df)

heatmap(mat, 
        Rowv = NA,  # Don't cluster rows
        Colv = NA,  # Don't cluster columns
        col = colorRampPalette(c("white", "blue"))(100),  # Define a color palette
        main = "Reference Matrix")



```

Some more visualization of expression values

```{r}

# summary heat map
# use function plotmean
# we need to know how many clusters we have.  They will be index 0 to n-1
# you need to pick the length to annotate

length(unique(seu$RNA_snn_res.1))
# 15
# if we want to plot by cluster we need a vector of from 0 to the n-1 clusters
cluster.num <- c(0:17)

plotmean(plot_type = 'heatmap',seu = seu, group = 'RNA_snn_res.1', markers = AB, 
                     var_names = cluster.num, slot = 'scale.data', xlab = "Cluster",
         ylab = "Antibody")


```

Add annotations based on expression


```{r}
# predict cell type from looking at expression patterns
myann <- c("precursors","neuralblast","neuron1","astrocyte1","stem-like",
           "epithelial","astrocyte-early","astrocyte2","DAneuron","RG-astro",
           "DAneuron2",
           "Neuron_early","RG","Glia","NPC","OPC","neuralstem","stem-like2")

man.ann <- data.frame(Cluster = c(0:17), MyAnn = myann)
man.ann

man.ann$Cluster <- as.factor(man.ann$Cluster)
man.ann$MyAnn <- as.factor(man.ann$MyAnn)
man.ann

```

```{r}
saveRDS(seu, paste(output_path, "Seurat_temp.RDS", sep = ""))

```

Predict cell annotations with CAM (Correlations assignment method)

```{r}

reference_path2 <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/ReferenceMatrixPanel2Timecourse.csv"
reference_data2 <- read.csv(reference_path2)
# the reference matrix need to be in the format cell types as rows and markers as columns
# there is a column X with the cell type names
df1 <- reference_data2
rownames(df1) <- df1$X # add row names (these are the markers)
df1 <- df1 %>% select(-"X") # remove the column with the marker names
df2 <- as.data.frame(t(df1)) # transpose the matrix
df2$X <- rownames(df2) # add the cell type names as a column

# check to make sure the reference matrix is correct
dim(df2)  # should be number of cell types (11) by the markers (13) with the X column of marker names = 14
head(df2) # cell types are columns and antibody/markers are rows

input <- read.csv(paste(output_path,"df2000Timecourse4.csv", sep = ""))
input_df <- input %>% select(c("X",AB,"Sample"))

dim(input_df)
colnames(input_df)

cor <- find_correlation(test = input_df, 
                             reference = df2, 
                             min_corr = 0.2, 
                             min_diff = 0.05)

# creates a dataframe with cor1 cor2 and predicted cell type label

```

Visualize the CAM results

```{r}

plot_corr(cor)

```

Apply correlation predictions to clusters and output a vector for annotation functions

```{r}


# add the correlation predictions to the meta data
seu <- AddMetaData(object=seu, metadata=cor$cell.label, col.name = 'cor.labels')
# see the labels added
unique(seu$cor.labels)

seu.cluster = seu$RNA_snn_res.1
seu.labels = seu$cor.labels

# plot the cluster predictions
plot_lab_clust(seu, seu$RNA_snn_res.1, seu$cor.labels)




```

Run get annotations function to return a vector of annotation in the order of the clusters.

```{r}

cor.ann <- get_annotation(seu, seu.cluster = seu$RNA_snn_res.1, 
                          seu.label = seu$cor.labels, top_n = 3, 
                          filter_out = c("Unknown","unknown","Mixed"), 
                          Label = "CAM")
cor.ann
dim(cor.ann)

unique(cor.ann$CAM)
length(cor.ann$CAM)

```

Use a trained Random Forest model to predict cell types. 
Training of the Random Forest model with an annotated data set is below.

```{r}

# you must have a saved trained model from a data object annotated from the same markers
# see the end of the workbook for how to train the model
# here we don't yet have a trained model, however we can train a model with the known previous annotation using only the overlapping antibodies
# see below in this workbook I have trained a model using the 9 MO samples annotated on the 9000 cells per sample subset. 

# rf <- readRDS("trained_model")


AB <- c("CD24","CD56","CD29","CD15","CD184","CD133","CD44", "CD140a")

rfm.pred <- RFM_predict(seu, rf)
head(rfm.pred)

# add the predictions into the seurat object
# check the column name for the predictions

seu <- AddMetaData(object=seu, metadata=rfm.pred$`predict(rf, df)`, col.name = 'rfm.labels')

# check that the data is added 
table(seu$rfm.labels)

# note these predictions are from the cell type annotatations in the older MOs from 3 genotypes and will only contain those celltypes


```

Get the annotation by cluster for the RFM

```{r}

rfm.ann <- get_annotation(seu, seu$RNA_snn_res.1,seu$rfm.labels, 
               top_n = 3, filter_out = c("unknown","Unknown","Mixed","Mix"), Label = "RFM")
rfm.ann

dim(rfm.ann)


```

Plot RFM predictions

```{r}


plot_lab_clust(seu, seu.cluster = seu$RNA_snn_res.1, seu.labels = seu$rfm.labels, filter_out = c("unknown","Unknown","Mixed"))


```

```{r}
# save the predictions so far
saveRDS(seu, paste(output_path, "SeuratSubTimeline.RDS", sep = ""))

```




Predicting cell types with Seurat label transfer using anchors


```{r}

# takes in a seurat object with the labels added 
# makes a dataframe with the count of predicted labels for each cluster
# input seurat object with the predicted labels in the meta data
# input the clusters meta data slot to be labels

#need reference data object with labels
seu.r<- readRDS("/Users/rhalenathomas/Documents/Data/FlowCytometry/PhenoID/Analysis/PaperFigures/Seu9000annot.08072021.RDS")

# now again we are using the subset of the 9 MO samples with annotations.  We will take only the here we can take the expression values as when we made the reference matrix
# only the proteins in the first panel will exist in this object

proteins <- c("TH","CD24","CD56","CD29","CD15","CD184",
              "CD133","SSEA4","CD44","CD49f","CD140a")

all <- c(proteins,genes)
# the output is a seurat object with the predicted annotations

seu <- seurat_predict(seu, seu.r, ref_id = 'subgroups', down.sample = 500, markers = all)

```

```{r}

# plot the seurat anchor predictions
# get the annotation table for the seurat anchor predictions 

plot_lab_clust(seu, seu$RNA_snn_res.1, seu$seu.pred)

# to not filter anything use c()
seu.ann <- get_annotation(seu, seu$RNA_snn_res.1,seu$seu.pred, 
               top_n = 3, filter_out = c("Unknown","Mixed"), Label = "Seurat")
seu.ann

# some clusters cannot be predicted
# need to filter the seu.ann table

seu.ann.ft <- seu.ann[!duplicated(as.character(seu.ann$Cluster)), ]
rownames(seu.ann.ft) <- NULL


```

Get a consensus of cluster annotations, Add the annotations to the seurat object


```{r}

# make a list of data frames
# all the dataframes need to be as.factor
ann.list <- list(cor.ann,rfm.ann,seu.ann.ft, man.ann)

# annotate the seurat object

seu <- cluster_annotate(seu, ann.list = ann.list, 
                        annotation_name ="CellType", 
                        to_label = "RNA_snn_res.1")

DimPlot(seu, group.by = "CellType")

# unfortunately for the consensus to work we need the cell type names to be identical

```

Use the annotate function

```{r}

new_annotations <- c("earlyDA-NPC","earlyGlia","Neurons","Astrocytes1",
                     "stem-like","epithelial","RG-fetal-astro","Astrocytes2",
                     "DAneurons","RG-astro","DANPC","Neurons_early",
                     "epithelialprecursor","Glia", "NPC","OPC","precursors","stem-like-RG"                   )

seu <- annotate(seu, annotations = new_annotations, to_label = "RNA_snn_res.1", annotation_name = "Celltypes2")

DimPlot(seu, group.by = "Celltypes2", label = TRUE)

combine_types <- c("earlyDA-NPC","earlyGlia","Neurons","Astrocytes",
                     "stem-like","epithelial","earlyRG","Astrocytes",
                     "DAneurons","RG","DA-NPC","earlyNeurons",
                     "epithelial and precursors","Glia","NPC",
                   "OPC","stem-like","stem-like")

seu <- annotate(seu, annotations = combine_types, to_label = "RNA_snn_res.1", annotation_name = "CelltypesMain")

DimPlot(seu, group.by = "CelltypesMain", label = TRUE)

```

# Compare groups

We first need to add the variables into the seurat object that we want to compare.

```{r}
# see the sample names and order

# set the ident to sample names
Idents(seu) <- 'Sample'
levels(seu)

Age <- c("150","150","150","150",
         "30","30","30","30",
         "60","60","60","60",
         "100","100","100","100")
Replicate <- c("R1","R2","R3","R4",
               "R1","R2","R3","R4",
               "R1","R2","R3","R4",
               "R1","R2","R3","R4")


# vector with the new names - you need this vector from me
cluster.ids <- Age

names(cluster.ids) <- levels(seu)    # get the levels
seu <- RenameIdents(seu, cluster.ids) # rename  
seu$Days_in_FD <- Idents(seu)   # add a new dataslot

# Replicate
Idents(seu) <- "Sample"
cluster.ids <- Replicate
# vector with the new names - you need this vector from me

names(cluster.ids) <- levels(seu)    # get the levels
seu <- RenameIdents(seu, cluster.ids) # rename  
seu$Replicate <- Idents(seu)   # add a new dataslot



```

Plots some variables to look for differences between groups

```{r}
# one plot
proportionplots(seu.q,seu.var = seu$Days_in_FD, seu.lable = seu$CelltypesMain, groups = "Days_in_FD")
# add colours
colours <- c("chocolate1","chocolate3","orange",
                   "lightsalmon", "pink","lightpink3",
                   "steelblue3","deepskyblue",
                   "plum3","purple",
                   "seagreen3","tomato4","burlywood3","grey90","brown",
             "royalblue3", "tan4","yellowgreen")

# same thing with custom colours
proportionplots(seu.q,seu.var = seu$Days_in_FD, seu.lable = seu$CelltypesMain, groups = "Days_in_FD", my_colours = colours)

# now check with more distinctive cell types
proportionplots(seu.q,seu.var = seu$Days_in_FD, seu.lable = seu$Celltypes2, groups = "Days_in_FD", my_colours = colours)

```

Make a heat map 

```{r}

# make sure the order is correct and there are only the amount of different cell types.  The order can be found in the dimplot above or check the levels
Idents(seu) <- "CelltypesMain"
levels(seu)
var_names <- c("earlyDA-NPC","earlyGlia","Neurons","Astrocytes","stem-like",
"epithelial","earlyRG","DAneurons", "RG", "DA-NPC","earlyNeurons",
"epithelial and precursors","Glia", "NPC","OPC")

plotmean(plot_type = 'heatmap',seu = seu, group = 'CelltypesMain', markers = AB, 
                     var_names = var_names, slot = 'scale.data', xlab = "Cell Type",
         ylab = "Antibody")




```


```{r}
# dot plot
var_names <- c("earlyDA-NPC","earlyGlia","Neurons","Astrocytes","stem-like",
"epithelial","earlyRG","DAneurons", "RG", "DA-NPC","earlyNeurons",
"epithelial and precursors","Glia", "NPC","OPC")


# make sure the terms are exactly the same and you don't miss any
new.order <- c("stem-like","epithelial and precursors",
               "earlyDA-NPC","earlyGlia","earlyRG","OPC","NPC","DA-NPC",
               "earlyNeurons","Neurons","DAneurons","epithelial","Astrocytes","Glia","RG"
               )
new.order <- rev(new.order)

AB.order <- c("TH","CD24","CD56","CD29","CD15","CD184",
              "CD133","SSEA4","CD44","CD49f","CD140a")

plotmean(plot_type = 'dotplot',seu = seu, group = 'CelltypesMain', markers = AB, 
                     var_names = var_names, slot = 'scale.data', xlab = "Cell Type",
         ylab = "Antibody", var1order = new.order, var2order = AB.order)


```

Look at some things across time

```{r}

table(seu$Celltypes2,seu$Days_in_FD)

```

New cell type names based on if cell types are changing the same over time

```{r}

combine_types <- c("earlyDA-NPC","earlyGlia","Neurons","Astrocytes",
                     "stem-like","epithelial","earlyRG","Astrocytes",
                     "DAneurons","RG-astro","DA-NPC","earlyNeurons",
                     "epithelial-endothelial","Glia","NPC",
                   "OPC","epithelial-maybeOligo","glia-maybeOligo")

seu <- annotate(seu, annotations = combine_types, to_label = "RNA_snn_res.1", annotation_name = "CellTypes")

DimPlot(seu, group.by = "CellTypes", label = TRUE, repel = TRUE)


```

Save the Seurat object

```{r}
saveRDS(seu, paste(output_path, "SeuratSubTimeline.RDS", sep = ""))

```

Train the Random Forest model

```{r}

# use Revised Celltype
# use all Markers

rf <- RFM_train(seurate_object = seu, 
                             AB_list = AB, 
                annotations = seu$CellTypes,
                      split = c(0.8,0.2),
                      downsample = 1000,
                      seed = 222,
                      mytry = c(1:10),
                      maxnodes = c(12: 25),
                      trees = c(250, 500, 1000,2000),
                      start_node = 15)

saveRDS(rf, paste(output_path,"trainedRFMfromTimeCourseSub.Rds",sep =""))

```

Now I will use the full data set
The cell types will be more clear with the full data set.
With the amount of cells in the full time course data this cannot be run in a workbook without down sampling. 

```{r}
input_path <- "/Users/rhalenathomas/Documents/Data/FlowCytometry/PhenoID/TimeCourseAIW/LiveCells"
output_path <- "/Users/rhalenathomas/Documents/Data/FlowCytometry/PhenoID/TimeCourseAIW/Analysis/"

# 1.a Read in FlowJo Files 

# down sample to take cells from each of 4 time points and 4 replicates
flowset <- fsc_to_fs(input_path)
# down sample can be a number, 'none' or 'min'

# look at file names and rename with shorter sample names
sampleNames(flowset) # function in flowCore

```

Rename samples with shorter names

```{r}
sampleNames(flowset) <- sampleNames(flowset) <- c("AIW_150_R1","AIW_150_R2","AIW_150_R3","AIW_150_R4",
  "AIW_30_R1","AIW_30_R2","AIW_30_R3","AIW_30_R4",
  "AIW_60_R1","AIW_60_R2","AIW_60_R3","AIW_60_R4",
  "AIW_100_R1","AIW_100_R2","AIW_100_R3","AIW_100_R4"
                                                  )
sampleNames(flowset)
```

Harmonize data to remove batch or technical variation

This requires us to look and see where there are two peaks to align. We need to visualize the peaks of the transformed data before aligning.

```{r}


# we can decided what level of processing to choose with the argument 'processing'
# biexp only applies a biexponential transformation
# align applies biexp transform and then aligns peaks
# retro (default), transforms, aligns and then reverse transforms
flowset_biexp <- harmonize(flowset, processing = 'biexp')
# we can visualize the peaks to see where there are two peaks for alignment

# we need to enter the column index for which peaks to align, the alignment for one or two peaks is not the same. 
#plotdensity_flowset(flowset)

#p1 <- plotdensity_flowset(flowset_biexp) # to see the peaks
#p1
# if there are two peaks assign as two peaks and one peak as one in the code chunk below

flowset_align <- harmonize(flowset, processing = 'align', 
                           two_peaks = c(9:11,15,17:21,25:31,33:42),
                       one_peak = c(1:8,12:14,16,24,32,43), threshold = 0.01)

flowset_retro <- harmonize(flowset, processing = 'retro', 
                           two_peaks = c(9:11,15,17:21,25:31,33:42),
                       one_peak = c(1:8,12:14,16,24,32,43), threshold = 0.01)
### test again later without running library(flowCore)
df <- flowset_to_csv(flowset_retro)

head(df)
# here we have 3 measures of each channel - we want the area measure and only this should have been selected.
# Select the first value from 

colnames(df)

```

Filter to have only one measurement per Marker/channel (area which is the first entry)

```{r}

df2 <- df[!duplicated(colnames(df))]
head(df2)

write.csv(df2,paste(output_path,"FlowDFretroAllcellsTimecourse.csv"))


```

Read in the saved dataframe

```{r}

df2 <- read.csv(paste(output_path,"FlowDFretroAllcellsTimecourse.csv"))

```


Make the seurat object
```{r}

# antibody features in order to appear on the plots
AB <- c("TH","CD24","CD56","CD29","CD15","CD184","CD133","SSEA4","CD44","CD49f","CD140a")

seu <- make_seu(df2, AB_vector = AB)

```
Have a look at the object

```{r}
#RidgePlot(seu, features = AB, log = TRUE)
DimPlot(seu, group.by = "Sample", reduction = "pca")
#colSums(seu) %>% summary
VlnPlot(seu, features = c("TH","CD24"))

```

```{r}
saveRDS(seu, paste(output_path,"SeuTimeCourseAll.RDS",sep=""))

```



Cluster
```{r}
# running from base R via terminal.  With some many cells this crashes R studio.
# use the parameter from above
seu <- get_clusters(seu, method = "louvain",
                    df_input = df.input,
                    k = 80,
                    resolution = c(1),
                    pcdim = 1:10,
                    plots = TRUE,
                    save_plots = FALSE)

```



```{r}
seu2 <- readRDS(paste(output_path,"SeuTimeCourseAll.RDS",sep=""))

```



# Annotate clusters
1. Visualization for manual annotation. - output by clustering function
2. CAM (Correlation assignment model) - requires reference matrix
3. RFM (Random Forest Model) - requires annotated matching flow dataset
4. Seurat label transfer - requires annotated matching flow data in a seurat object


Predict with CAM
```{r}

reference_path2 <- "/Users/rhalenathomas/GITHUB/CelltypeR/ExampleOuts/ReferenceMatrixPanel2Timecourse.csv"
reference_data2 <- read.csv(reference_path)
# the reference matrix need to be in the format cell types as rows and markers as columns
# there is a column X with the cell type names
df1 <- reference_data2
rownames(df1) <- df1$X
df1 <- df1 %>% select(-"X")
df2 <- as.data.frame(t(df1))
df2$X <- rownames(df2)
head(df2)
input <- read.csv(paste(output_path,"df2000Timecourse4.csv", sep = ""))
input_df <- input %>% select(c("X",AB,"cell","Sample"))
head(input_df)

cor <- find_correlation(test = input_df, 
                        reference = df2, 
                        min_corr = 0.45, 
                        min_diff = 0.05)
dim(input_df) # rows x columns
colnames(input_df)
colnames(df2)

```



# Statistics comparing groups


```{r}

# prepare a dataframe for stats
# this function takes the annotated seurat object with all the variables already existing as metadata slots

# check what meta data slots exist in your object
colnames(seu@meta.data)



```


```{r}
# run the function
# put all things you might want to use as variables in var names, including the cell type groups
var.names <- c("Days_in_FD","Sample","Replicate","CelltypesMain")

df.for.stats <- Prep_for_stats(seu, marker_list = AB, variables = var.names, 
                               marker_name = 'Marker')


# save the csv for later
# write.csv(df.for.stats, paste(output_path,"filename.csv"))

head(df.for.stats)


```


First we have 3 variables: Experimental variable (Days in culture), Cell types, Markers
We want to compare the mean expression per group:
For all antibodies together in each cell type between a given variable (Genotype) (One way anova)
We want to compare each antibody separately for a given variable for each cell type (two way anova)

```{r}


# one way anova (combine all antibodies)
# two way anova variable 1 = antibody/marker variable 2 = group variable

# normally runs a loop across each cell type
# option to select to calculate for all cell types together

# can run the loop across antibodies

# group variables we can run: Genotype, Batch, days in culture, Experiment date

# we run one-way here with different conditions

# compare genotypes for each cell type
test.stats1 <- run_stats(input_df= df.for.stats, group_cols = c("Sample", "CelltypesMain","Marker","Days_in_FD","Replicate"),
                     value_col = "value",
                     stat_type = "ANOVA", id1 = 'Days_in_FD', 
                     id2 = NULL, use_means = TRUE,
                     loop_by = "CelltypesMain")

# see the dataframe results
test.stats1[['ANOVA']]
test.stats1[['TukeyHSD']]



```
Overall effect we can see a significant effect of Days in culture on cell types
In pair wise conparisons only 30 vs 100 days is dfferent.



```{r}
# Compare expression across cell types for each marker
test.stats2 <- run_stats(input_df= df.for.stats, group_cols = c("Sample", "CelltypesMain","Marker","Days_in_FD"),
                     value_col = "value",
                     stat_type = "ANOVA", id1 = 'Days_in_FD', 
                     id2 = NULL, use_means = TRUE,
                     loop_by = "Marker")
test.stats2[['TukeyHSD']]


```


```{r}
# compare genotype for each cell type without taking the group means and using each cell as an n (not good practice)
test.stats4 <- run_stats(input_df= df.for.stats, group_cols = c("Sample", "CellType","Marker","Genotype","Batch"),
                     value_col = "value",
                     stat_type = "ANOVA", id1 = 'Genotype', 
                     id2 = NULL, use_means = FALSE,
                     loop_by = "CellType")

test.stats4[[1]] # ANOVA

```


Test two way ANOVAs

```{r}

test.stats5 <- run_stats(input_df= df.for.stats, group_cols = c("Sample", "CelltypesMain","Days_in_FD","Marker"),
                     value_col = "value",
                     stat_type = "ANOVA2", id1 = 'Days_in_FD', 
                     id2 = "Marker", use_means = TRUE,
                     loop_by = "CellType")

# see the relevant interactions Tukey's results

df <- as.data.frame(test.stats5[["TukeyHSD"]][["Interactions_ Genotype"]])
head(df)

# now filter for significant differences

df_sig <- df %>% filter(p.adj < 0.05)
df_sig



```


```{r}
test.stats6 <- run_stats(input_df= df.for.stats, group_cols = c("Sample", "CellType","Marker","Genotype","Batch"),
                     value_col = "value",
                     stat_type = "ANOVA2", id1 = 'Genotype', 
                     id2 = "CellType", use_means = TRUE,
                     loop_by = "Marker")


# see the interactions Tukey's results where id1 or id2 match

df <- as.data.frame(test.stats6[["TukeyHSD"]][["Interactions_ Genotype"]])
head(df)

# now filter for significant differences

df_sig <- df %>% filter(p.adj < 0.05)
df_sig

df <- as.data.frame(test.stats6[["TukeyHSD"]][["Interactions_ CellType"]])
head(df)
# now filter for significant differences
df_sig <- df %>% filter(p.adj < 0.05)
df_sig


```






Train Random Forest model
Requires a labelled seurat object
More cells will give higher accuracy but increase computation time.
Run in HPC with lots of cells


```{r}

# we need a labelled object 
# here we will train a model with the 3 cell lines samples to apply to the time line samples using only the overlapping antibodies.
seu.r<- readRDS("/Users/rhalenathomas/Documents/Data/FlowCytometry/PhenoID/Analysis/PaperFigures/Seu9000annot.08072021.RDS")

overlapMarkers <- c("CD24","CD56","CD29","CD15","CD184","CD133","CD44", "CD140a")
# check where the labels are in the seurat object 
colnames(seu.r@meta.data)

rf <- RFM_train(seurate_object = seu.r, 
                             AB_list = overlapMarkers, 
                annotations = seu.r$subgroups,
                      split = c(0.8,0.2),
                      downsample = "none",
                      seed = 222,
                      mytry = c(1:10),
                      maxnodes = c(12: 25),
                      trees = c(250, 500, 1000,2000),
                      start_node = 15)

saveRDS(rf, paste(output_path,"trainedRFMfrom9MOsubsetOverlapPanel.Rds",sep =""))


```


